--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -43,7 +_,11 @@
    private ClientInformation f_290851_;
 
    public ServerConfigurationPacketListenerImpl(MinecraftServer p_301415_, Connection p_298106_, CommonListenerCookie p_301309_) {
-      super(p_301415_, p_298106_, p_301309_);
+      this(p_301415_, p_298106_, p_301309_, new ServerPlayer(p_301415_, p_301415_.m_129783_(), p_301309_.f_290628_(), p_301309_.f_290565_()));
+
+   }
+   public ServerConfigurationPacketListenerImpl(MinecraftServer p_301415_, Connection p_298106_, CommonListenerCookie p_301309_, ServerPlayer player) { // CraftBukkit
+      super(p_301415_, p_298106_, p_301309_, player); // CraftBukkit
       this.f_291596_ = p_301309_.f_290628_();
       this.f_290851_ = p_301309_.f_290565_();
    }
@@ -61,12 +_,15 @@
       return this.f_291338_.m_129536_();
    }
 
-   public void m_294295_() {
+   private void vanillaStart() {
       this.m_141995_(new ClientboundCustomPayloadPacket(new BrandPayload(this.f_291389_.getServerModName())));
       LayeredRegistryAccess<RegistryLayer> layeredregistryaccess = this.f_291389_.m_247573_();
       this.m_141995_(new ClientboundUpdateEnabledFeaturesPacket(FeatureFlags.f_244280_.m_245829_(this.f_291389_.m_129910_().m_247623_())));
       this.m_141995_(new ClientboundRegistryDataPacket((new RegistryAccess.ImmutableRegistryAccess(RegistrySynchronization.m_257599_(layeredregistryaccess))).m_203557_()));
       this.m_141995_(new ClientboundUpdateTagsPacket(TagNetworkSerialization.m_245799_(layeredregistryaccess)));
+   }
+
+   public void m_294295_() {
       this.m_295612_();
       this.f_290691_.add(new JoinWorldTask());
       this.m_295676_();
@@ -107,13 +_,14 @@
             return;
          }
 
-         Component component = playerlist.m_6418_(this.f_291338_.m_129523_(), this.f_291596_);
+         Component component = null; // CraftBukkit - login checks already completed
+
          if (component != null) {
             this.m_294716_(component);
             return;
          }
 
-         ServerPlayer serverplayer = playerlist.m_215624_(this.f_291596_, this.f_290851_);
+         ServerPlayer serverplayer = playerlist.getPlayerForLogin(this.f_291596_, this.f_290851_, this.player); // CraftBukkit
          playerlist.m_11261_(this.f_291338_, serverplayer, this.m_295477_(this.f_290851_));
          this.f_291338_.m_294304_();
       } catch (Exception exception) {
