--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -52,6 +_,7 @@
    private final String serverId = "";
    @Nullable
    private ServerPlayer delayedAcceptPlayer;
+   private net.minecraft.server.level.ServerPlayer player; // CraftBukkit
 
    public ServerLoginPacketListenerImpl(MinecraftServer p_10027_, Connection p_10028_) {
       this.server = p_10027_;
@@ -60,7 +_,13 @@
    }
 
    public void tick() {
-      if (this.state == ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT) {
+      if (this.state == State.NEGOTIATING) {
+         // We force the state into "NEGOTIATING" which is otherwise unused. Once we're completed we move the negotiation onto "READY_TO_ACCEPT"
+         // Might want to promote player object creation to here as well..
+         boolean negotiationComplete = net.minecraftforge.network.NetworkHooks.tickNegotiation(this, this.connection, this.delayedAcceptPlayer);
+         if (negotiationComplete)
+            this.state = State.READY_TO_ACCEPT;
+      } else if (this.state == ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT) {
          this.handleAcceptedLogin();
       } else if (this.state == ServerLoginPacketListenerImpl.State.DELAY_ACCEPT) {
          ServerPlayer serverplayer = this.server.getPlayerList().getPlayer(this.gameProfile.getId());
@@ -77,6 +_,13 @@
 
    }
 
+   // CraftBukkit start
+   @Deprecated
+   public void disconnect(String s) {
+      disconnect(Component.literal(s));
+   }
+   // CraftBukkit end
+
    public boolean isAcceptingMessages() {
       return this.connection.isConnected();
    }
@@ -92,14 +_,42 @@
 
    }
 
+   // Spigot start
+   public void initUUID()
+   {
+      UUID uuid;
+      if ( connection.spoofedUUID != null )
+      {
+          uuid = connection.spoofedUUID;
+      } else
+      {
+          uuid = UUIDUtil.createOfflinePlayerUUID( this.gameProfile.getName() );
+      }
+
+      this.gameProfile = new GameProfile( uuid, this.gameProfile.getName() );
+
+      if (connection.spoofedProfile != null)
+      {
+          for ( com.mojang.authlib.properties.Property property : connection.spoofedProfile )
+          {
+              if ( !ServerHandshakePacketListenerImpl.PROP_PATTERN.matcher( property.getName() ).matches() ) continue;
+              this.gameProfile.getProperties().put( property.getName(), property );
+          }
+      }
+   }
+
    public void handleAcceptedLogin() {
-      if (!this.gameProfile.isComplete()) {
-         this.gameProfile = this.createFakeProfile(this.gameProfile);
+      if (!this.server.usesAuthentication()) {
+         // this.gameProfile = this.createFakeProfile(this.gameProfile); // Spigot - Moved to initUUID
+         // Spigot end
       }
 
-      Component component = this.server.getPlayerList().canPlayerLogin(this.connection.getRemoteAddress(), this.gameProfile);
-      if (component != null) {
-         this.disconnect(component);
+      // CraftBukkit start - fire PlayerLoginEvent
+      ServerPlayer s = this.server.getPlayerList().canPlayerLogin(this, this.gameProfile);
+//      this.player = playerlist.canPlayerLogin(this, p_299507_); // CraftBukkit //Ketting 1.20.2 stuffs
+      if (s == null) {
+         // this.disconnect(component);
+         // CraftBukkit end
       } else {
          this.state = ServerLoginPacketListenerImpl.State.ACCEPTED;
          if (this.server.getCompressionThreshold() >= 0 && !this.connection.isMemoryConnection()) {
@@ -112,7 +_,7 @@
          ServerPlayer serverplayer = this.server.getPlayerList().getPlayer(this.gameProfile.getId());
 
          try {
-            ServerPlayer serverplayer1 = this.server.getPlayerList().getPlayerForLogin(this.gameProfile);
+            ServerPlayer serverplayer1 = this.server.getPlayerList().getPlayerForLogin(this.gameProfile, s);
             if (serverplayer != null) {
                this.state = ServerLoginPacketListenerImpl.State.DELAY_ACCEPT;
                this.delayedAcceptPlayer = serverplayer1;
@@ -138,7 +_,8 @@
    }
 
    public String getUserName() {
-      return this.gameProfile != null ? this.gameProfile + " (" + this.connection.getRemoteAddress() + ")" : String.valueOf((Object)this.connection.getRemoteAddress());
+      final String addressString = net.minecraftforge.network.DualStackUtils.getAddressString(this.connection.getRemoteAddress());
+      return this.gameProfile != null ? this.gameProfile + " (" + addressString + ")" : addressString;
    }
 
    public void handleHello(ServerboundHelloPacket p_10047_) {
@@ -147,14 +_,29 @@
       GameProfile gameprofile = this.server.getSingleplayerProfile();
       if (gameprofile != null && p_10047_.name().equalsIgnoreCase(gameprofile.getName())) {
          this.gameProfile = gameprofile;
-         this.state = ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT;
+         this.state = ServerLoginPacketListenerImpl.State.NEGOTIATING; // FORGE: continue NEGOTIATING, we move to READY_TO_ACCEPT after Forge is ready
       } else {
          this.gameProfile = new GameProfile((UUID)null, p_10047_.name());
          if (this.server.usesAuthentication() && !this.connection.isMemoryConnection()) {
             this.state = ServerLoginPacketListenerImpl.State.KEY;
             this.connection.send(new ClientboundHelloPacket("", this.server.getKeyPair().getPublic().getEncoded(), this.challenge));
          } else {
-            this.state = ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT;
+            //this.state = ServerLoginPacketListenerImpl.State.NEGOTIATING; // FORGE: continue NEGOTIATING, we move to READY_TO_ACCEPT after Forge is ready
+            // Spigot start
+            new Thread("User Authenticator #" + ServerLoginPacketListenerImpl.UNIQUE_THREAD_ID.incrementAndGet()) {
+
+               @Override
+               public void run() {
+                  try {
+                     initUUID();
+                     new LoginHandler().fireEvents();
+                  } catch (Exception ex) {
+                     disconnect("Failed to verify username!");
+                     server.server.getLogger().log(java.util.logging.Level.WARNING, "Exception verifying " + gameProfile.getName(), ex);
+                  }
+               }
+            }.start();
+            // Spigot end
          }
 
       }
@@ -186,19 +_,25 @@
          throw new IllegalStateException("Protocol error", cryptexception);
       }
 
-      Thread thread = new Thread("User Authenticator #" + UNIQUE_THREAD_ID.incrementAndGet()) {
+      Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, "User Authenticator #" + UNIQUE_THREAD_ID.incrementAndGet()) {
          public void run() {
             GameProfile gameprofile = ServerLoginPacketListenerImpl.this.gameProfile;
 
             try {
                ServerLoginPacketListenerImpl.this.gameProfile = ServerLoginPacketListenerImpl.this.server.getSessionService().hasJoinedServer(new GameProfile((UUID)null, gameprofile.getName()), s, this.getAddress());
                if (ServerLoginPacketListenerImpl.this.gameProfile != null) {
-                  ServerLoginPacketListenerImpl.LOGGER.info("UUID of player {} is {}", ServerLoginPacketListenerImpl.this.gameProfile.getName(), ServerLoginPacketListenerImpl.this.gameProfile.getId());
-                  ServerLoginPacketListenerImpl.this.state = ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT;
+                  //ServerLoginPacketListenerImpl.this.state = ServerLoginPacketListenerImpl.State.NEGOTIATING; // FORGE: continue NEGOTIATING, we move to READY_TO_ACCEPT after Forge is ready
+
+                  // CraftBukkit start - fire PlayerPreLoginEvent
+                  if (!connection.isConnected()) {
+                     return;
+                  }
+
+                  new LoginHandler().fireEvents();
                } else if (ServerLoginPacketListenerImpl.this.server.isSingleplayer()) {
                   ServerLoginPacketListenerImpl.LOGGER.warn("Failed to verify username but will let them in anyway!");
                   ServerLoginPacketListenerImpl.this.gameProfile = gameprofile;
-                  ServerLoginPacketListenerImpl.this.state = ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT;
+                  ServerLoginPacketListenerImpl.this.state = ServerLoginPacketListenerImpl.State.NEGOTIATING; // FORGE: continue NEGOTIATING, we move to READY_TO_ACCEPT after Forge is ready
                } else {
                   ServerLoginPacketListenerImpl.this.disconnect(Component.translatable("multiplayer.disconnect.unverified_username"));
                   ServerLoginPacketListenerImpl.LOGGER.error("Username '{}' tried to join with an invalid session", (Object)gameprofile.getName());
@@ -207,11 +_,16 @@
                if (ServerLoginPacketListenerImpl.this.server.isSingleplayer()) {
                   ServerLoginPacketListenerImpl.LOGGER.warn("Authentication servers are down but will let them in anyway!");
                   ServerLoginPacketListenerImpl.this.gameProfile = gameprofile;
-                  ServerLoginPacketListenerImpl.this.state = ServerLoginPacketListenerImpl.State.READY_TO_ACCEPT;
+                  ServerLoginPacketListenerImpl.this.state = ServerLoginPacketListenerImpl.State.NEGOTIATING; // FORGE: continue NEGOTIATING, we move to READY_TO_ACCEPT after Forge is ready
                } else {
                   ServerLoginPacketListenerImpl.this.disconnect(Component.translatable("multiplayer.disconnect.authservers_down"));
                   ServerLoginPacketListenerImpl.LOGGER.error("Couldn't verify username because servers are unavailable");
                }
+               // CraftBukkit start - catch all exceptions
+            } catch (Exception exception) {
+               disconnect("Failed to verify username!");
+               server.server.getLogger().log(java.util.logging.Level.WARNING, "Exception verifying " + (gameprofile!=null?gameprofile.getId():"null"), exception);
+               // CraftBukkit end
             }
 
          }
@@ -226,7 +_,50 @@
       thread.start();
    }
 
+   // Spigot start
+   public class LoginHandler {
+
+      public void fireEvents() throws Exception {
+         String playerName = gameProfile.getName();
+         java.net.InetAddress address = ((java.net.InetSocketAddress) connection.getRemoteAddress()).getAddress();
+         java.util.UUID uniqueId = gameProfile.getId();
+         final org.bukkit.craftbukkit.v1_20_R1.CraftServer server = ServerLoginPacketListenerImpl.this.server.server;
+
+         org.bukkit.event.player.AsyncPlayerPreLoginEvent asyncEvent = new org.bukkit.event.player.AsyncPlayerPreLoginEvent(playerName, address, uniqueId);
+         server.getPluginManager().callEvent(asyncEvent);
+
+         if (org.bukkit.event.player.PlayerPreLoginEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            final org.bukkit.event.player.PlayerPreLoginEvent event = new org.bukkit.event.player.PlayerPreLoginEvent(playerName, address, uniqueId);
+            if (asyncEvent.getResult() != org.bukkit.event.player.PlayerPreLoginEvent.Result.ALLOWED) {
+               event.disallow(asyncEvent.getResult(), asyncEvent.getKickMessage());
+            }
+            org.bukkit.craftbukkit.v1_20_R1.util.Waitable<org.bukkit.event.player.PlayerPreLoginEvent.Result> waitable = new org.bukkit.craftbukkit.v1_20_R1.util.Waitable<org.bukkit.event.player.PlayerPreLoginEvent.Result>() {
+               @Override
+               protected org.bukkit.event.player.PlayerPreLoginEvent.Result evaluate() {
+                  server.getPluginManager().callEvent(event);
+                  return event.getResult();
+               }};
+
+            ServerLoginPacketListenerImpl.this.server.processQueue.add(waitable);
+            if (waitable.get() != org.bukkit.event.player.PlayerPreLoginEvent.Result.ALLOWED) {
+               disconnect(event.getKickMessage());
+               return;
+            }
+         } else {
+            if (asyncEvent.getLoginResult() != org.bukkit.event.player.AsyncPlayerPreLoginEvent.Result.ALLOWED) {
+               disconnect(asyncEvent.getKickMessage());
+               return;
+            }
+         }
+         // CraftBukkit end
+         ServerLoginPacketListenerImpl.LOGGER.info("UUID of player {} is {}", gameProfile.getName(), gameProfile.getId());
+         state = State.NEGOTIATING; // FORGE: continue NEGOTIATING, we move to READY_TO_ACCEPT after Forge is ready
+      }
+   }
+   // Spigot end
+
    public void handleCustomQueryPacket(ServerboundCustomQueryPacket p_10045_) {
+      if (!net.minecraftforge.network.NetworkHooks.onCustomPayload(p_10045_, this.connection))
       this.disconnect(Component.translatable("multiplayer.disconnect.unexpected_query_response"));
    }
 
@@ -234,6 +_,39 @@
       UUID uuid = UUIDUtil.createOfflinePlayerUUID(p_10039_.getName());
       return new GameProfile(uuid, p_10039_.getName());
    }
+
+   // Spigot start
+   /* Ketting - 1.20.2 stuffs. 
+   @Nullable
+   public GameProfile getAuthenticatedProfile() {
+       return this.authenticatedProfile;
+   }
+
+   protected GameProfile createOfflineProfileSpigot(String p_297551_) {
+      UUID uuid;
+      if ( connection.spoofedUUID != null )
+      {
+         uuid = connection.spoofedUUID;
+      } else
+      {
+         uuid = UUIDUtil.createOfflinePlayerUUID( p_297551_ );
+      }
+
+      GameProfile gameProfile = new GameProfile( uuid, p_297551_ );
+
+      if (connection.spoofedProfile != null)
+      {
+         for ( com.mojang.authlib.properties.Property property : connection.spoofedProfile )
+         {
+            if ( !ServerHandshakePacketListenerImpl.PROP_PATTERN.matcher( property.getName()).matches() ) continue;
+            gameProfile.getProperties().put( property.getName(), property );
+         }
+      }
+
+      return gameProfile;
+   }
+   */
+   // Spigot end
 
    static enum State {
       HELLO,
