--- a/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/BoatDispenseItemBehavior.java
@@ -33,22 +_,54 @@
       double d2 = p_123375_.y() + (double)((float)direction.getStepY() * 1.125F);
       double d3 = p_123375_.z() + (double)direction.getStepZ() * d0;
       BlockPos blockpos = p_123375_.getPos().relative(direction);
+      Boat boat = (Boat)(this.isChestBoat ? new ChestBoat(level, d0, d1, d2) : new Boat(level, d0, d1, d2));
+      boat.setVariant(this.type);
+      boat.setYRot(direction.toYRot());
       double d4;
-      if (level.getFluidState(blockpos).is(FluidTags.WATER)) {
+      if (boat.canBoatInFluid(level.getFluidState(blockpos))) {
          d4 = 1.0D;
       } else {
-         if (!level.getBlockState(blockpos).isAir() || !level.getFluidState(blockpos.below()).is(FluidTags.WATER)) {
+         if (!level.getBlockState(blockpos).isAir() || !boat.canBoatInFluid(level.getFluidState(blockpos.below()))) {
             return this.defaultDispenseItemBehavior.dispense(p_123375_, p_123376_);
          }
 
          d4 = 0.0D;
       }
 
-      Boat boat = (Boat)(this.isChestBoat ? new ChestBoat(level, d1, d2 + d4, d3) : new Boat(level, d1, d2 + d4, d3));
-      boat.setVariant(this.type);
-      boat.setYRot(direction.toYRot());
-      level.addFreshEntity(boat);
-      p_123376_.shrink(1);
+      // Boat boat = (Boat)(this.isChestBoat ? new ChestBoat(level, d0, d1, d2) : new Boat(level, d0, d1, d2));
+      // CraftBukkit start
+      ItemStack itemstack1 = p_123376_.split(1);
+      org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123375_.getPos());
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+      org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d1, d2 + d4, d3));
+      if (!DispenserBlock.eventFired) {
+         level.getCraftServer().getPluginManager().callEvent(event);
+      }
+
+      if (event.isCancelled()) {
+         p_123376_.grow(1);
+         return p_123376_;
+      }
+
+      if (!event.getItem().equals(craftItem)) {
+         p_123376_.grow(1);
+         // Chain to handler for new item
+         ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+         DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+         if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+            idispensebehavior.dispense(p_123375_, eventStack);
+            return p_123376_;
+         }
+      }
+
+      Boat boat1 = (Boat)(this.isChestBoat ? new ChestBoat(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ()) : new Boat(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ()));
+      // CraftBukkit end
+
+      boat1.setVariant(this.type);
+      boat1.setYRot(direction.toYRot());
+      if (!level.addFreshEntity(boat1)) p_123376_.grow(1); // CraftBukkit
+      // p_123376_.shrink(1); // CraftBukkit - handled during event processing
       return p_123376_;
    }
 
