--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -181,6 +_,33 @@
             Direction direction = p_123523_.getBlockState().getValue(DispenserBlock.FACING);
             EntityType<?> entitytype = ((SpawnEggItem)p_123524_.getItem()).getType(p_123524_.getTag());
 
+            // CraftBukkit start
+            ServerLevel level = p_123523_.getLevel();
+            ItemStack itemstack = p_123524_.split(1);
+            org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123523_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               p_123524_.grow(1);
+               return p_123524_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               p_123524_.grow(1);
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123523_, eventStack);
+                  return p_123524_;
+               }
+            }
+
             try {
                entitytype.spawn(p_123523_.getLevel(), p_123524_, (Player)null, p_123523_.getPos().relative(direction), MobSpawnType.DISPENSER, direction != Direction.UP, false);
             } catch (Exception exception) {
@@ -188,7 +_,8 @@
                return ItemStack.EMPTY;
             }
 
-            p_123524_.shrink(1);
+            // p_123524_.shrink(1); // Handled during event processing
+            // CraftBukkit end
             p_123523_.getLevel().gameEvent((Entity)null, GameEvent.ENTITY_PLACE, p_123523_.getPos());
             return p_123524_;
          }
@@ -203,12 +_,40 @@
             Direction direction = p_123461_.getBlockState().getValue(DispenserBlock.FACING);
             BlockPos blockpos = p_123461_.getPos().relative(direction);
             ServerLevel serverlevel = p_123461_.getLevel();
+
+            // CraftBukkit start
+            ItemStack itemstack = p_123462_.split(1);
+            org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(serverlevel, p_123461_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+            if (!DispenserBlock.eventFired) {
+               serverlevel.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               p_123462_.grow(1);
+               return p_123462_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               p_123462_.grow(1);
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123461_, eventStack);
+                  return p_123462_;
+               }
+            }
+            // CraftBukkit end
+
             Consumer<ArmorStand> consumer = EntityType.appendDefaultStackConfig((p_277236_) -> {
                p_277236_.setYRot(direction.toYRot());
             }, serverlevel, p_123462_, (Player)null);
             ArmorStand armorstand = EntityType.ARMOR_STAND.spawn(serverlevel, p_123462_.getTag(), consumer, blockpos, MobSpawnType.DISPENSER, false, false);
             if (armorstand != null) {
-               p_123462_.shrink(1);
+               // p_123462_.shrink(1); // CraftBukkit - Handled during event processing
             }
 
             return p_123462_;
@@ -225,8 +_,36 @@
                }
             });
             if (!list.isEmpty()) {
+               // CraftBukkit start
+               ItemStack itemstack = p_123530_.split(1);
+               Level level = p_123529_.getLevel();
+               org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123529_.getPos());
+               org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+               org.bukkit.event.block.BlockDispenseArmorEvent event = new org.bukkit.event.block.BlockDispenseArmorEvent(block, craftItem.clone(), (org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) list.get(0).getBukkitEntity());
+               if (!DispenserBlock.eventFired) {
+                  level.getCraftServer().getPluginManager().callEvent(event);
+               }
+
+               if (event.isCancelled()) {
+                  p_123530_.grow(1);
+                  return p_123530_;
+               }
+
+               if (!event.getItem().equals(craftItem)) {
+                  p_123530_.grow(1);
+                  // Chain to handler for new item
+                  ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                  DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+                  if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != ArmorItem.DISPENSE_ITEM_BEHAVIOR) {
+                     idispensebehavior.dispense(p_123529_, eventStack);
+                     return p_123530_;
+                  }
+               }
+               // CraftBukkit end
+
                ((Saddleable)list.get(0)).equipSaddle(SoundSource.BLOCKS);
-               p_123530_.shrink(1);
+               // p_123530_.shrink(1); // CraftBukkit - handled above
                this.setSuccess(true);
                return p_123530_;
             } else {
@@ -242,7 +_,35 @@
                return p_289248_.isAlive() && p_289248_.canWearArmor();
             })) {
                if (abstracthorse.isArmor(p_123536_) && !abstracthorse.isWearingArmor() && abstracthorse.isTamed()) {
-                  abstracthorse.getSlot(401).set(p_123536_.split(1));
+                  // CraftBukkit start
+                  ItemStack itemstack = p_123536_.split(1);
+                  Level level = p_123535_.getLevel();
+                  org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123535_.getPos());
+                  org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+                  org.bukkit.event.block.BlockDispenseArmorEvent event = new org.bukkit.event.block.BlockDispenseArmorEvent(block, craftItem.clone(), (org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) abstracthorse.getBukkitEntity());
+                  if (!DispenserBlock.eventFired) {
+                     level.getCraftServer().getPluginManager().callEvent(event);
+                  }
+
+                  if (event.isCancelled()) {
+                     p_123536_.grow(1);
+                     return p_123536_;
+                  }
+
+                  if (!event.getItem().equals(craftItem)) {
+                     p_123536_.grow(1);
+                     // Chain to handler for new item
+                     ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                     DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+                     if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != ArmorItem.DISPENSE_ITEM_BEHAVIOR) {
+                        idispensebehavior.dispense(p_123535_, eventStack);
+                        return p_123536_;
+                     }
+                  }
+
+                  abstracthorse.getSlot(401).set(org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem()));
+                  // CraftBukkit end
                   this.setSuccess(true);
                   return p_123536_;
                }
@@ -278,8 +_,36 @@
             for(AbstractChestedHorse abstractchestedhorse : p_123541_.getLevel().getEntitiesOfClass(AbstractChestedHorse.class, new AABB(blockpos), (p_289249_) -> {
                return p_289249_.isAlive() && !p_289249_.hasChest();
             })) {
-               if (abstractchestedhorse.isTamed() && abstractchestedhorse.getSlot(499).set(p_123542_)) {
-                  p_123542_.shrink(1);
+               // CraftBukkit start
+               if (abstractchestedhorse.isTamed()) {
+                  ItemStack itemstack = p_123542_.split(1);
+                  Level level = p_123541_.getLevel();
+                  org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123541_.getPos());
+                  org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+                  org.bukkit.event.block.BlockDispenseArmorEvent event = new org.bukkit.event.block.BlockDispenseArmorEvent(block, craftItem.clone(), (org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) abstractchestedhorse.getBukkitEntity());
+                  if (!DispenserBlock.eventFired) {
+                     level.getCraftServer().getPluginManager().callEvent(event);
+                  }
+
+                  if (event.isCancelled()) {
+                     return p_123542_;
+                  }
+
+                  if (!event.getItem().equals(craftItem)) {
+                     // Chain to handler for new item
+                     ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                     DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+                     if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != ArmorItem.DISPENSE_ITEM_BEHAVIOR) {
+                        idispensebehavior.dispense(p_123541_, eventStack);
+                        return p_123542_;
+                     }
+                  }
+
+                  abstractchestedhorse.getSlot(499).set(org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem()));
+                  // CraftBukkit end
+
+                  // p_123542_.shrink(1); // CraftBukkit - handled above
                   this.setSuccess(true);
                   return p_123542_;
                }
@@ -291,11 +_,40 @@
       DispenserBlock.registerBehavior(Items.FIREWORK_ROCKET, new DefaultDispenseItemBehavior() {
          public ItemStack execute(BlockSource p_123547_, ItemStack p_123548_) {
             Direction direction = p_123547_.getBlockState().getValue(DispenserBlock.FACING);
+            // CraftBukkit start
+            ServerLevel worldserver = p_123547_.getLevel();
+            ItemStack itemstack = p_123548_.split(1);
+            org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(worldserver, p_123547_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(direction.getStepX(), direction.getStepY(), direction.getStepZ()));
+            if (!DispenserBlock.eventFired) {
+               worldserver.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               p_123548_.grow(1);
+               return p_123548_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               p_123548_.grow(1);
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123547_, eventStack);
+                  return p_123548_;
+               }
+            }
+
+            itemstack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
             FireworkRocketEntity fireworkrocketentity = new FireworkRocketEntity(p_123547_.getLevel(), p_123548_, p_123547_.x(), p_123547_.y(), p_123547_.x(), true);
             DispenseItemBehavior.setEntityPokingOutOfBlock(p_123547_, fireworkrocketentity, direction);
             fireworkrocketentity.shoot((double)direction.getStepX(), (double)direction.getStepY(), (double)direction.getStepZ(), 0.5F, 1.0F);
             p_123547_.getLevel().addFreshEntity(fireworkrocketentity);
-            p_123548_.shrink(1);
+            // p_123548_.shrink(1); // Handled during event processing
+            // CraftBukkit end
             return p_123548_;
          }
 
@@ -315,11 +_,40 @@
             double d3 = randomsource.triangle((double)direction.getStepX(), 0.11485000000000001D);
             double d4 = randomsource.triangle((double)direction.getStepY(), 0.11485000000000001D);
             double d5 = randomsource.triangle((double)direction.getStepZ(), 0.11485000000000001D);
-            SmallFireball smallfireball = new SmallFireball(level, d0, d1, d2, d3, d4, d5);
-            level.addFreshEntity(Util.make(smallfireball, (p_123552_) -> {
-               p_123552_.setItem(p_123557_);
-            }));
-            p_123557_.shrink(1);
+
+            // CraftBukkit start
+            ItemStack itemstack = p_123557_.split(1);
+            org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123556_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d3, d4, d5));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               p_123557_.grow(1);
+               return p_123557_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               p_123557_.grow(1);
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123556_, eventStack);
+                  return p_123557_;
+               }
+            }
+
+            SmallFireball smallfireball = new SmallFireball(level, d0, d1, d2, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
+            smallfireball.setItem(itemstack);
+            smallfireball.projectileSource = new org.bukkit.craftbukkit.v1_20_R1.projectiles.CraftBlockProjectileSource(p_123556_.getEntity());
+
+            level.addFreshEntity(smallfireball);
+            // p_123557_.shrink(1); // Handled during event processing
+            // CraftBukkit end
             return p_123557_;
          }
 
@@ -352,9 +_,52 @@
             DispensibleContainerItem dispensiblecontaineritem = (DispensibleContainerItem)p_123562_.getItem();
             BlockPos blockpos = p_123561_.getPos().relative(p_123561_.getBlockState().getValue(DispenserBlock.FACING));
             Level level = p_123561_.getLevel();
+
+            // CraftBukkit start
+            int x = blockpos.getX();
+            int y = blockpos.getY();
+            int z = blockpos.getZ();
+            BlockState iblockdata = level.getBlockState(blockpos);
+            if (iblockdata.isAir() || iblockdata.canBeReplaced() || (dispensiblecontaineritem instanceof net.minecraft.world.item.BucketItem && iblockdata.getBlock() instanceof net.minecraft.world.level.block.LiquidBlockContainer && ((net.minecraft.world.level.block.LiquidBlockContainer) iblockdata.getBlock()).canPlaceLiquid(level, blockpos, iblockdata, ((net.minecraft.world.item.BucketItem) dispensiblecontaineritem).getFluid()))) { //Ketting
+               org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123561_.getPos());
+               org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123562_);
+
+               org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(x, y, z));
+               if (!DispenserBlock.eventFired) {
+                  level.getCraftServer().getPluginManager().callEvent(event);
+               }
+
+               if (event.isCancelled()) {
+                  return p_123562_;
+               }
+
+               if (!event.getItem().equals(craftItem)) {
+                  // Chain to handler for new item
+                  ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                  DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+                  if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                     idispensebehavior.dispense(p_123561_, eventStack);
+                     return p_123562_;
+                  }
+               }
+
+               dispensiblecontaineritem = (DispensibleContainerItem) org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem()).getItem();
+            }
+            // CraftBukkit end
+
             if (dispensiblecontaineritem.emptyContents((Player)null, level, blockpos, (BlockHitResult)null)) {
                dispensiblecontaineritem.checkExtraContent((Player)null, level, p_123562_, blockpos);
-               return new ItemStack(Items.BUCKET);
+               // CraftBukkit start - Handle stacked buckets
+               Item item = Items.BUCKET;
+               p_123562_.shrink(1);
+               if (p_123562_.isEmpty()) {
+                  p_123562_.setItem(Items.BUCKET);
+                  p_123562_.setCount(1);
+               } else if (((DispenserBlockEntity)p_123561_.getEntity()).addItem(new ItemStack(item)) < 0) {
+                  this.defaultDispenseItemBehavior.dispense(p_123561_, new ItemStack(item));
+               }
+               return p_123562_;
+               // CraftBukkit end
             } else {
                return this.defaultDispenseItemBehavior.dispense(p_123561_, p_123562_);
             }
@@ -373,17 +_,44 @@
          private final DefaultDispenseItemBehavior defaultDispenseItemBehavior = new DefaultDispenseItemBehavior();
 
          public ItemStack execute(BlockSource p_123566_, ItemStack p_123567_) {
-            LevelAccessor levelaccessor = p_123566_.getLevel();
+            ServerLevel level = p_123566_.getLevel(); //Ketting: LevelAccessor -> ServerLevel
             BlockPos blockpos = p_123566_.getPos().relative(p_123566_.getBlockState().getValue(DispenserBlock.FACING));
-            BlockState blockstate = levelaccessor.getBlockState(blockpos);
+            BlockState blockstate = level.getBlockState(blockpos);
             Block block = blockstate.getBlock();
             if (block instanceof BucketPickup) {
-               ItemStack itemstack = ((BucketPickup)block).pickupBlock(levelaccessor, blockpos, blockstate);
+               ItemStack itemstack = ((BucketPickup) block).pickupBlock(org.bukkit.craftbukkit.v1_20_R1.util.DummyGeneratorAccess.INSTANCE, blockpos, blockstate); // CraftBukkit
                if (itemstack.isEmpty()) {
                   return super.execute(p_123566_, p_123567_);
                } else {
-                  levelaccessor.gameEvent((Entity)null, GameEvent.FLUID_PICKUP, blockpos);
+                  level.gameEvent((Entity)null, GameEvent.FLUID_PICKUP, blockpos);
                   Item item = itemstack.getItem();
+
+                  // CraftBukkit start
+                  org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123566_.getPos());
+                  org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123567_);
+
+                  org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                  if (!DispenserBlock.eventFired) {
+                     level.getCraftServer().getPluginManager().callEvent(event);
+                  }
+
+                  if (event.isCancelled()) {
+                     return p_123567_;
+                  }
+
+                  if (!event.getItem().equals(craftItem)) {
+                     // Chain to handler for new item
+                     ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                     DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+                     if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                        idispensebehavior.dispense(p_123566_, eventStack);
+                        return p_123567_;
+                     }
+                  }
+
+                  itemstack = ((BucketPickup)block).pickupBlock(level, blockpos, blockstate); // From above
+                  // CraftBukkit end
+
                   p_123567_.shrink(1);
                   if (p_123567_.isEmpty()) {
                      return new ItemStack(item);
@@ -403,17 +_,53 @@
       DispenserBlock.registerBehavior(Items.FLINT_AND_STEEL, new OptionalDispenseItemBehavior() {
          protected ItemStack execute(BlockSource p_123412_, ItemStack p_123413_) {
             Level level = p_123412_.getLevel();
+
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123412_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123413_);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               return p_123413_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123412_, eventStack);
+                  return p_123413_;
+               }
+            }
+            // CraftBukkit end
+
             this.setSuccess(true);
             Direction direction = p_123412_.getBlockState().getValue(DispenserBlock.FACING);
             BlockPos blockpos = p_123412_.getPos().relative(direction);
             BlockState blockstate = level.getBlockState(blockpos);
             if (BaseFireBlock.canBePlacedAt(level, blockpos, direction)) {
-               level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(level, blockpos));
-               level.gameEvent((Entity)null, GameEvent.BLOCK_PLACE, blockpos);
+               // CraftBukkit start - Ignition by dispensing flint and steel
+               if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, p_123412_.getPos()).isCancelled()) {
+                  level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(level, blockpos));
+                  level.gameEvent((Entity)null, GameEvent.BLOCK_PLACE, blockpos);
+               }
+               // CraftBukkit end
             } else if (!CampfireBlock.canLight(blockstate) && !CandleBlock.canLight(blockstate) && !CandleCakeBlock.canLight(blockstate)) {
-               if (blockstate.getBlock() instanceof TntBlock) {
-                  TntBlock.explode(level, blockpos);
-                  level.removeBlock(blockpos, false);
+               if (blockstate.isFlammable(level, blockpos, p_123412_.getBlockState().getValue(DispenserBlock.FACING).getOpposite())) {
+                  if (blockstate.getBlock() instanceof TntBlock) {
+                     // CraftBukkit - TNTPrimeEvent
+                     if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callTNTPrimeEvent(level, blockpos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.DISPENSER, null, p_123412_.getPos())) {
+                        blockstate.onCaughtFire(level, blockpos, p_123412_.getBlockState().getValue(DispenserBlock.FACING).getOpposite(), null);
+                        level.removeBlock(blockpos, false);
+                     }
+                  } else {
+                     blockstate.onCaughtFire(level, blockpos, p_123412_.getBlockState().getValue(DispenserBlock.FACING).getOpposite(), null);
+                  }
                } else {
                   this.setSuccess(false);
                }
@@ -434,11 +_,62 @@
             this.setSuccess(true);
             Level level = p_123416_.getLevel();
             BlockPos blockpos = p_123416_.getPos().relative(p_123416_.getBlockState().getValue(DispenserBlock.FACING));
+            // CraftBukkit start
+            org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123416_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123417_);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               return p_123417_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123416_, eventStack);
+                  return p_123417_;
+               }
+            }
+
+            level.captureTreeGeneration = true;
+            // CraftBukkit end
+
             if (!BoneMealItem.growCrop(p_123417_, level, blockpos) && !BoneMealItem.growWaterPlant(p_123417_, level, blockpos, (Direction)null)) {
                this.setSuccess(false);
             } else if (!level.isClientSide) {
                level.levelEvent(1505, blockpos, 0);
             }
+            // CraftBukkit start
+            level.captureTreeGeneration = false;
+            if (level.capturedBlockStates.size() > 0) {
+               org.bukkit.TreeType treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
+               net.minecraft.world.level.block.SaplingBlock.treeType = null;
+               org.bukkit.Location location = org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation.toBukkit(blockpos, level.getWorld());
+               List<org.bukkit.block.BlockState> blocks = new java.util.ArrayList<>(level.capturedBlockStates.values());
+               level.capturedBlockStates.clear();
+               org.bukkit.event.world.StructureGrowEvent structureEvent = null;
+               if (treeType != null) {
+                  structureEvent = new org.bukkit.event.world.StructureGrowEvent(location, treeType, false, null, blocks);
+                  org.bukkit.Bukkit.getPluginManager().callEvent(structureEvent);
+               }
+
+               org.bukkit.event.block.BlockFertilizeEvent fertilizeEvent = new org.bukkit.event.block.BlockFertilizeEvent(location.getBlock(), null, blocks);
+               fertilizeEvent.setCancelled(structureEvent != null && structureEvent.isCancelled());
+               org.bukkit.Bukkit.getPluginManager().callEvent(fertilizeEvent);
+
+               if (!fertilizeEvent.isCancelled()) {
+                  for (org.bukkit.block.BlockState blockstate : blocks) {
+                     blockstate.update(true);
+                  }
+               }
+            }
+            // CraftBukkit end
 
             return p_123417_;
          }
@@ -447,11 +_,41 @@
          protected ItemStack execute(BlockSource p_123425_, ItemStack p_123426_) {
             Level level = p_123425_.getLevel();
             BlockPos blockpos = p_123425_.getPos().relative(p_123425_.getBlockState().getValue(DispenserBlock.FACING));
-            PrimedTnt primedtnt = new PrimedTnt(level, (double)blockpos.getX() + 0.5D, (double)blockpos.getY(), (double)blockpos.getZ() + 0.5D, (LivingEntity)null);
+            // CraftBukkit start
+            // PrimedTnt primedtnt = new PrimedTnt(level, (double)blockpos.getX() + 0.5D, (double)blockpos.getY(), (double)blockpos.getZ() + 0.5D, (LivingEntity)null);
+
+            ItemStack itemstack1 = p_123426_.split(1);
+            org.bukkit.block.Block block = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123425_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector((double) blockpos.getX() + 0.5D, (double) blockpos.getY(), (double) blockpos.getZ() + 0.5D));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               p_123426_.grow(1);
+               return p_123426_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               p_123426_.grow(1);
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123425_, eventStack);
+                  return p_123426_;
+               }
+            }
+
+            PrimedTnt primedtnt = new PrimedTnt(level, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), (LivingEntity) null);
+            // CraftBukkit end
+
             level.addFreshEntity(primedtnt);
             level.playSound((Player)null, primedtnt.getX(), primedtnt.getY(), primedtnt.getZ(), SoundEvents.TNT_PRIMED, SoundSource.BLOCKS, 1.0F, 1.0F);
             level.gameEvent((Entity)null, GameEvent.ENTITY_PLACE, blockpos);
-            p_123426_.shrink(1);
+            // p_123426_.shrink(1); // CraftBukkit - handled above
             return p_123426_;
          }
       });
@@ -472,6 +_,31 @@
             Level level = p_123433_.getLevel();
             Direction direction = p_123433_.getBlockState().getValue(DispenserBlock.FACING);
             BlockPos blockpos = p_123433_.getPos().relative(direction);
+
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123433_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123434_);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               return p_123434_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123433_, eventStack);
+                  return p_123434_;
+               }
+            }
+            // CraftBukkit end
+
             if (level.isEmptyBlock(blockpos) && WitherSkullBlock.canSpawnMob(level, blockpos, p_123434_)) {
                level.setBlock(blockpos, Blocks.WITHER_SKELETON_SKULL.defaultBlockState().setValue(SkullBlock.ROTATION, Integer.valueOf(RotationSegment.convertToSegment(direction))), 3);
                level.gameEvent((Entity)null, GameEvent.BLOCK_PLACE, blockpos);
@@ -494,6 +_,31 @@
             Level level = p_123437_.getLevel();
             BlockPos blockpos = p_123437_.getPos().relative(p_123437_.getBlockState().getValue(DispenserBlock.FACING));
             CarvedPumpkinBlock carvedpumpkinblock = (CarvedPumpkinBlock)Blocks.CARVED_PUMPKIN;
+
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(level, p_123437_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123438_);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+            if (!DispenserBlock.eventFired) {
+               level.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               return p_123438_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123437_, eventStack);
+                  return p_123438_;
+               }
+            }
+            // CraftBukkit end
+
             if (level.isEmptyBlock(blockpos) && carvedpumpkinblock.canSpawnGolem(level, blockpos)) {
                if (!level.isClientSide) {
                   level.setBlock(blockpos, carvedpumpkinblock.defaultBlockState(), 3);
@@ -537,6 +_,31 @@
             ServerLevel serverlevel = p_123444_.getLevel();
             BlockPos blockpos = p_123444_.getPos().relative(p_123444_.getBlockState().getValue(DispenserBlock.FACING));
             BlockState blockstate = serverlevel.getBlockState(blockpos);
+
+            // CraftBukkit start
+            org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(serverlevel, p_123444_.getPos());
+            org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(p_123445_);
+
+            org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+            if (!DispenserBlock.eventFired) {
+               serverlevel.getCraftServer().getPluginManager().callEvent(event);
+            }
+
+            if (event.isCancelled()) {
+               return p_123445_;
+            }
+
+            if (!event.getItem().equals(craftItem)) {
+               // Chain to handler for new item
+               ItemStack eventStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+               DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+               if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+                  idispensebehavior.dispense(p_123444_, eventStack);
+                  return p_123445_;
+               }
+            }
+            // CraftBukkit end
+
             if (blockstate.is(BlockTags.BEEHIVES, (p_123442_) -> {
                return p_123442_.hasProperty(BeehiveBlock.HONEY_LEVEL) && p_123442_.getBlock() instanceof BeehiveBlock;
             }) && blockstate.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
