--- a/net/minecraft/world/level/block/entity/BannerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BannerBlockEntity.java
@@ -26,9 +_,9 @@
    public static final String TAG_COLOR = "Color";
    @Nullable
    private Component name;
-   private DyeColor baseColor;
+   public DyeColor baseColor;
    @Nullable
-   private ListTag itemPatterns;
+   public ListTag itemPatterns;
    @Nullable
    private List<Pair<Holder<BannerPattern>, DyeColor>> patterns;
 
@@ -96,6 +_,7 @@
       }
 
       this.itemPatterns = p_155042_.getList("Patterns", 10);
+      // CraftBukkit start
+      while (this.itemPatterns.size() > 20) {
+         this.itemPatterns.remove(20);
+      }
+      // CraftBukkit end
       this.patterns = null;
    }
 
