--- a/net/minecraft/world/level/block/entity/BannerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BannerBlockEntity.java
@@ -46,6 +_,7 @@
    public static ListTag getItemPatterns(ItemStack p_58488_) {
       ListTag listtag = null;
       CompoundTag compoundtag = BlockItem.getBlockEntityData(p_58488_);
+
       if (compoundtag != null && compoundtag.contains("Patterns", 9)) {
          listtag = compoundtag.getList("Patterns", 10).copy();
       }
@@ -96,6 +_,11 @@
       }
 
       this.itemPatterns = p_155042_.getList("Patterns", 10);
+      // CraftBukkit start
+      while (this.itemPatterns.size() > 20) {
+         this.itemPatterns.remove(20);
+      }
+      // CraftBukkit end
       this.patterns = null;
    }
 
