--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -20,6 +_,24 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
+   // CraftBukkit start - add fields and methods
+   private int maxStack = MAX_STACK;
+
+   @Override
+   public java.util.List<ItemStack> getContents() {
+      return this.items;
+   }
+
+   @Override
+   public int getMaxStackSize() {
+      return maxStack;
+   }
+
+   @Override
+   public void setMaxStackSize(int i) {
+      maxStack = i;
+   }
+   // CraftBukkit end
    private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
    public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
       protected void onOpen(Level p_155062_, BlockPos p_155063_, BlockState p_155064_) {
