--- a/net/minecraft/world/level/block/entity/CommandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CommandBlockEntity.java
@@ -19,6 +_,13 @@
    private boolean auto;
    private boolean conditionMet;
    private final BaseCommandBlock commandBlock = new BaseCommandBlock() {
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return new org.bukkit.craftbukkit.v1_20_R1.command.CraftBlockCommandSender(wrapper, CommandBlockEntity.this);
+      }
+      // CraftBukkit end
+
       public void setCommand(String p_59157_) {
          super.setCommand(p_59157_);
          CommandBlockEntity.this.setChanged();
