--- a/net/minecraft/world/level/block/entity/CrafterBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CrafterBlockEntity.java
@@ -20,6 +_,9 @@
 import net.minecraft.world.level.block.CrafterBlock;
 import net.minecraft.world.level.block.state.BlockState;
 
+// CraftBukkit start
+// CraftBukkit end
+
 public class CrafterBlockEntity extends RandomizableContainerBlockEntity implements CraftingContainer {
    public static final int CONTAINER_WIDTH = 3;
    public static final int CONTAINER_HEIGHT = 3;
@@ -51,6 +_,47 @@
          return 10;
       }
    };
+   // CraftBukkit start - add fields and methods
+   public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+   private int maxStack = 1;
+
+   @Override
+   public java.util.List<ItemStack> getContents() {
+       return this.items;
+   }
+
+   @Override
+   public void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+       transaction.add(who);
+   }
+
+   @Override
+   public void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+       transaction.remove(who);
+   }
+
+   @Override
+   public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+       return transaction;
+   }
+
+   @Override
+   public int getMaxStackSize() {
+       return maxStack;
+   }
+
+   @Override
+   public void setMaxStackSize(int size) {
+       maxStack = size;
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+       if (level == null) return null;
+       return new org.bukkit.Location(level.getWorld(), worldPosition.getX(), worldPosition.getY(), worldPosition.getZ());
+   }
+   // CraftBukkit end
+
 
    public CrafterBlockEntity(BlockPos p_309972_, BlockState p_313058_) {
       super(BlockEntityType.CRAFTER, p_309972_, p_313058_);
