--- a/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
@@ -29,6 +_,6 @@
    @Nullable
    public DecoratedPotBlockEntity.WobbleStyle lastWobbleStyle;
-   private DecoratedPotBlockEntity.Decorations decorations;
+   public DecoratedPotBlockEntity.Decorations decorations;
    private ItemStack item = ItemStack.EMPTY;
    @Nullable
    protected ResourceLocation lootTable;
