--- a/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
@@ -27,8 +_,8 @@
    public static final String LOOT_TABLE_TAG = "LootTable";
    public static final String LOOT_TABLE_SEED_TAG = "LootTableSeed";
    @Nullable
-   protected ResourceLocation lootTable;
-   protected long lootTableSeed;
+   public ResourceLocation lootTable;
+   public long lootTableSeed;
 
    protected RandomizableContainerBlockEntity(BlockEntityType<?> p_155629_, BlockPos p_155630_, BlockState p_155631_) {
       super(p_155629_, p_155630_, p_155631_);
@@ -148,4 +_,12 @@
          return null;
       }
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.Location getLocation() {
+      if (level == null) return null;
+      return new org.bukkit.Location(level.getWorld(), worldPosition.getX(), worldPosition.getY(), worldPosition.getZ());
+   }
+   // CraftBukkit end
 }
