--- a/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -55,7 +_,7 @@
       p_222866_.put(3, SoundEvents.WARDEN_NEARBY_CLOSEST);
       p_222866_.put(4, SoundEvents.WARDEN_LISTENING_ANGRY);
    });
-   private int warningLevel;
+   public int warningLevel;
    private final VibrationSystem.User vibrationUser = new SculkShriekerBlockEntity.VibrationUser();
    private VibrationSystem.Data vibrationData = new VibrationSystem.Data();
    private final VibrationSystem.Listener vibrationListener = new VibrationSystem.Listener(this);
