--- a/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -37,9 +_,9 @@
 
    };
    @Nullable
-   private GameProfile owner;
+   public GameProfile owner;
    @Nullable
-   private ResourceLocation noteBlockSound;
+   public ResourceLocation noteBlockSound;
    private int animationTickCount;
    private boolean isAnimating;
 
@@ -162,6 +_,16 @@
       String s = p_301404_.getString("SkullOwner");
       if (!Util.isBlank(s)) {
          resolveGameProfile(p_301404_, s);
+         // CraftBukkit start
+      } else {
+         net.minecraft.nbt.ListTag textures = p_301404_.getCompound("SkullOwner").getCompound("Properties").getList("textures", 10); // Safe due to method contracts
+         for (int i = 0; i < textures.size(); i++) {
+            if (textures.get(i) instanceof CompoundTag && !((CompoundTag) textures.get(i)).contains("Signature", 8) && ((CompoundTag) textures.get(i)).getString("Value").trim().isEmpty()) {
+               p_301404_.remove("SkullOwner");
+               break;
+            }
+         }
+         // CraftBukkit end
       }
 
    }
@@ -187,7 +_,7 @@
       }, CHECKED_MAIN_THREAD_EXECUTOR);
    }
 
-   private static CompletableFuture<Optional<GameProfile>> fillProfileTextures(GameProfile p_299217_) {
+   public static CompletableFuture<Optional<GameProfile>> fillProfileTextures(GameProfile p_299217_) {
       return hasTextures(p_299217_) ? CompletableFuture.completedFuture(Optional.of(p_299217_)) : CompletableFuture.supplyAsync(() -> {
          MinecraftSessionService minecraftsessionservice = sessionService;
          if (minecraftsessionservice != null) {
