--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -154,7 +_,8 @@
    public static void teleportEntity(Level p_155829_, BlockPos p_155830_, BlockState p_155831_, Entity p_155832_, TheEndGatewayBlockEntity p_155833_) {
       if (p_155829_ instanceof ServerLevel serverlevel && !p_155833_.isCoolingDown()) {
          p_155833_.teleportCooldown = 100;
-         if (p_155833_.exitPortal == null && p_155829_.dimension() == Level.END) {
+
+         if (p_155833_.exitPortal == null && p_155829_.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit - work in alternate worlds
             BlockPos blockpos = findOrCreateValidTeleportPos(serverlevel, p_155830_);
             blockpos = blockpos.above(10);
             LOGGER.debug("Creating portal at {}", (Object)blockpos);
@@ -180,6 +_,27 @@
             } else {
                entity = p_155832_.getRootVehicle();
             }
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (entity instanceof ServerPlayer) {
+               org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer player = (org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer) entity.getBukkitEntity();
+               org.bukkit.Location location = org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation.toBukkit(blockpos1, p_155829_.getWorld()).add(0.5D, 0.5D, 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+
+               org.bukkit.event.player.PlayerTeleportEvent teleEvent = new org.bukkit.event.player.PlayerTeleportEvent(player, player.getLocation(), location, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               org.bukkit.Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+
+               entity.setPortalCooldown();
+               ((ServerPlayer) entity).connection.teleport(teleEvent.getTo());
+               triggerCooldown(p_155829_, p_155830_, p_155831_, p_155833_); // CraftBukkit - call at end of method
+               return;
+
+            }
+            // CraftBukkit end
 
             entity.setPortalCooldown();
             entity.teleportToWithTicket((double)blockpos1.getX() + 0.5D, (double)blockpos1.getY(), (double)blockpos1.getZ() + 0.5D);
