--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -154,7 +_,8 @@
    public static void m_155828_(Level p_155829_, BlockPos p_155830_, BlockState p_155831_, Entity p_155832_, TheEndGatewayBlockEntity p_155833_) {
       if (p_155829_ instanceof ServerLevel serverlevel && !p_155833_.m_59972_()) {
          p_155833_.f_59927_ = 100;
-         if (p_155833_.f_59928_ == null && p_155829_.m_46472_() == Level.f_46430_) {
+
+         if (p_155833_.f_59928_ == null && p_155829_.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.f_63973_) { // CraftBukkit - work in alternate worlds
             BlockPos blockpos = m_155818_(serverlevel, p_155830_);
             blockpos = blockpos.m_6630_(10);
             f_59925_.debug("Creating portal at {}", (Object)blockpos);
@@ -180,6 +_,27 @@
             } else {
                entity = p_155832_.m_20201_();
             }
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (entity instanceof ServerPlayer) {
+               org.bukkit.craftbukkit.v1_20_R2.entity.CraftPlayer player = (org.bukkit.craftbukkit.v1_20_R2.entity.CraftPlayer) entity.getBukkitEntity();
+               org.bukkit.Location location = org.bukkit.craftbukkit.v1_20_R2.util.CraftLocation.toBukkit(blockpos1, p_155829_.getWorld()).add(0.5D, 0.5D, 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+
+               org.bukkit.event.player.PlayerTeleportEvent teleEvent = new org.bukkit.event.player.PlayerTeleportEvent(player, player.getLocation(), location, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               org.bukkit.Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+
+               entity.m_20091_();
+               ((ServerPlayer) entity).f_8906_.teleport(teleEvent.getTo());
+               m_155849_(p_155829_, p_155830_, p_155831_, p_155833_); // CraftBukkit - call at end of method
+               return;
+
+            }
+            // CraftBukkit end
 
             entity.m_20091_();
             entity.m_20324_((double)blockpos1.m_123341_() + 0.5D, (double)blockpos1.m_123342_(), (double)blockpos1.m_123343_() + 0.5D);
