--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -14,8 +_,16 @@
    private static final Component f_11500_ = Component.m_237113_("Rcon");
    private final StringBuffer f_11501_ = new StringBuffer();
    private final MinecraftServer f_11502_;
+   // CraftBukkit start
+   public final java.net.SocketAddress socketAddress;
+   private final org.bukkit.craftbukkit.v1_20_R1.command.CraftRemoteConsoleCommandSender remoteConsole = new org.bukkit.craftbukkit.v1_20_R1.command.CraftRemoteConsoleCommandSender(this);
 
    public RconConsoleSource(MinecraftServer p_11505_) {
+      this(p_11505_, null);
+   }
+   public RconConsoleSource(MinecraftServer p_11505_, java.net.SocketAddress socketAddress) {
+      this.socketAddress = socketAddress;
+      // CraftBukkit end
       this.f_11502_ = p_11505_;
    }
 
@@ -32,8 +_,19 @@
       return new CommandSourceStack(this, Vec3.m_82528_(serverlevel.m_220360_()), Vec2.f_82462_, serverlevel, 4, "Rcon", f_11500_, this.f_11502_, (Entity)null);
    }
 
+   // CraftBukkit start - Send a String
+   public void sendMessage(String message) {
+      this.f_11501_.append(message);
+   }
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return this.remoteConsole;
+   }
+   // CraftBukkit end
+
    public void m_213846_(Component p_215653_) {
-      this.f_11501_.append(p_215653_.getString());
+      this.f_11501_.append(p_215653_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean m_6999_() {
