--- a/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
@@ -45,9 +_,9 @@
       }
    };
    private final RealmsNotificationsScreen.DataFetcherConfiguration onlyNotifications = new RealmsNotificationsScreen.DataFetcherConfiguration() {
-      public DataFetcher.Subscription initDataFetcher(RealmsDataFetcher p_275731_) {
-         DataFetcher.Subscription datafetcher$subscription = p_275731_.dataFetcher.createSubscription();
-         RealmsNotificationsScreen.this.addNotificationsSubscriptions(p_275731_, datafetcher$subscription);
+      public DataFetcher.Subscription initDataFetcher(RealmsDataFetcher p_275318_) {
+         DataFetcher.Subscription datafetcher$subscription = p_275318_.dataFetcher.createSubscription();
+         RealmsNotificationsScreen.this.addNotificationsSubscriptions(p_275318_, datafetcher$subscription);
          return datafetcher$subscription;
       }
 
