--- a/net/minecraft/client/MouseHandler.java
+++ b/net/minecraft/client/MouseHandler.java
@@ -73,6 +_,7 @@
             this.activeButton = -1;
          }
 
+         if (net.minecraftforge.client.ForgeHooksClient.onMouseButtonPre(p_91532_, p_91533_, p_91534_)) return;
          boolean[] aboolean = new boolean[]{false};
          if (this.minecraft.getOverlay() == null) {
             if (this.minecraft.screen == null) {
@@ -86,11 +_,19 @@
                if (flag) {
                   screen.afterMouseAction();
                   Screen.wrapScreenError(() -> {
-                     aboolean[0] = screen.mouseClicked(d0, d1, i);
+                     aboolean[0] = net.minecraftforge.client.ForgeHooksClient.onScreenMouseClickedPre(screen, d0, d1, i);
+                     if (!aboolean[0]) {
+                        aboolean[0] = this.minecraft.screen.mouseClicked(d0, d1, i);
+                        aboolean[0] = net.minecraftforge.client.ForgeHooksClient.onScreenMouseClickedPost(screen, d0, d1, i, aboolean[0]);
+                     }
                   }, "mouseClicked event handler", screen.getClass().getCanonicalName());
                } else {
                   Screen.wrapScreenError(() -> {
-                     aboolean[0] = screen.mouseReleased(d0, d1, i);
+                     aboolean[0] = net.minecraftforge.client.ForgeHooksClient.onScreenMouseReleasedPre(screen, d0, d1, i);
+                     if (!aboolean[0]) {
+                        aboolean[0] = this.minecraft.screen.mouseReleased(d0, d1, i);
+                        aboolean[0] = net.minecraftforge.client.ForgeHooksClient.onScreenMouseReleasedPost(screen, d0, d1, i, aboolean[0]);
+                     }
                   }, "mouseReleased event handler", screen.getClass().getCanonicalName());
                }
             }
@@ -114,19 +_,26 @@
                }
             }
          }
-
+         net.minecraftforge.client.ForgeHooksClient.onMouseButtonPost(p_91532_, p_91533_, p_91534_);
       }
    }
 
    private void onScroll(long p_91527_, double p_91528_, double p_91529_) {
       if (p_91527_ == Minecraft.getInstance().getWindow().getWindow()) {
-         double d0 = (this.minecraft.options.discreteMouseScroll().get() ? Math.signum(p_91529_) : p_91529_) * this.minecraft.options.mouseWheelSensitivity().get();
+         // FORGE: Allows for Horizontal Scroll to be recognized as Vertical Scroll - Fixes MC-121772
+         double offset = p_91529_;
+         if (Minecraft.ON_OSX && p_91529_ == 0) {
+            offset = p_91528_;
+         }
+         double d0 = (this.minecraft.options.discreteMouseScroll().get() ? Math.signum(offset) : offset) * this.minecraft.options.mouseWheelSensitivity().get();
          if (this.minecraft.getOverlay() == null) {
             if (this.minecraft.screen != null) {
                double d1 = this.xpos * (double)this.minecraft.getWindow().getGuiScaledWidth() / (double)this.minecraft.getWindow().getScreenWidth();
                double d2 = this.ypos * (double)this.minecraft.getWindow().getGuiScaledHeight() / (double)this.minecraft.getWindow().getScreenHeight();
-               this.minecraft.screen.mouseScrolled(d1, d2, d0);
                this.minecraft.screen.afterMouseAction();
+               if (net.minecraftforge.client.ForgeHooksClient.onScreenMouseScrollPre(this, this.minecraft.screen, d0)) return;
+               if (this.minecraft.screen.mouseScrolled(d1, d2, d0)) return;
+               net.minecraftforge.client.ForgeHooksClient.onScreenMouseScrollPost(this, this.minecraft.screen, d0);
             } else if (this.minecraft.player != null) {
                if (this.accumulatedScroll != 0.0D && Math.signum(d0) != Math.signum(this.accumulatedScroll)) {
                   this.accumulatedScroll = 0.0D;
@@ -139,6 +_,7 @@
                }
 
                this.accumulatedScroll -= (double)i;
+               if (net.minecraftforge.client.ForgeHooksClient.onMouseScroll(this, d0)) return;
                if (this.minecraft.player.isSpectator()) {
                   if (this.minecraft.gui.getSpectatorGui().isMenuActive()) {
                      this.minecraft.gui.getSpectatorGui().onMouseScrolled(-i);
@@ -207,7 +_,9 @@
                double d2 = (p_91563_ - this.xpos) * (double)this.minecraft.getWindow().getGuiScaledWidth() / (double)this.minecraft.getWindow().getScreenWidth();
                double d3 = (p_91564_ - this.ypos) * (double)this.minecraft.getWindow().getGuiScaledHeight() / (double)this.minecraft.getWindow().getScreenHeight();
                Screen.wrapScreenError(() -> {
-                  screen.mouseDragged(d0, d1, this.activeButton, d2, d3);
+                  if (net.minecraftforge.client.ForgeHooksClient.onScreenMouseDragPre(screen, d0, d1, this.activeButton, d2, d3)) return;
+                  if (screen.mouseDragged(d0, d1, this.activeButton, d2, d3)) return;
+                  net.minecraftforge.client.ForgeHooksClient.onScreenMouseDragPost(screen, d0, d1, this.activeButton, d2, d3);
                }, "mouseDragged event handler", screen.getClass().getCanonicalName());
             }
 
@@ -290,6 +_,14 @@
 
    public double ypos() {
       return this.ypos;
+   }
+
+   public double getXVelocity() {
+      return this.accumulatedDX;
+   }
+
+   public double getYVelocity() {
+      return this.accumulatedDY;
    }
 
    public void setIgnoreFirstMove() {
