--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -90,9 +_,19 @@
       }
    }
 
+   // Ketting start - Keep CraftBukkit method
+   public java.util.concurrent.atomic.AtomicBoolean ketting$parse$overridePermissions = new java.util.concurrent.atomic.AtomicBoolean(false);
+
+   public EntitySelector parse(StringReader p_91451_, boolean overridePermissions) throws CommandSyntaxException {
+      ketting$parse$overridePermissions.set(overridePermissions);
+      return parse(p_91451_);
+   }
+   // Ketting end
+
    public EntitySelector parse(StringReader p_91451_) throws CommandSyntaxException {
       int i = 0;
       EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_91451_);
+      entityselectorparser.ketting$parse$overridePermissions.set(ketting$parse$overridePermissions.getAndSet(false));
       EntitySelector entityselector = entityselectorparser.parse();
       if (entityselector.getMaxResults() > 1 && this.single) {
          if (this.playersOnly) {
@@ -115,7 +_,7 @@
       if (s instanceof SharedSuggestionProvider sharedsuggestionprovider) {
          StringReader stringreader = new StringReader(p_91483_.getInput());
          stringreader.setCursor(p_91483_.getStart());
-         EntitySelectorParser entityselectorparser = new EntitySelectorParser(stringreader, sharedsuggestionprovider.hasPermission(2));
+         EntitySelectorParser entityselectorparser = new EntitySelectorParser(stringreader, net.minecraftforge.common.ForgeHooks.canUseEntitySelectors(sharedsuggestionprovider));
 
          try {
             entityselectorparser.parse();
