--- a/net/minecraft/world/effect/HungerMobEffect.java
+++ b/net/minecraft/world/effect/HungerMobEffect.java
@@ -11,7 +_,7 @@
    public void m_6742_(LivingEntity p_301304_, int p_301079_) {
       super.m_6742_(p_301304_, p_301079_);
       if (p_301304_ instanceof Player player) {
-         player.m_36399_(0.005F * (float)(p_301079_ + 1));
+         player.causeFoodExhaustion(0.005F * (float)(p_301079_ + 1), org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.HUNGER_EFFECT); // CraftBukkit - EntityExhaustionEvent
       }
 
    }
