--- a/net/minecraft/world/item/crafting/RecipeHolder.java
+++ b/net/minecraft/world/item/crafting/RecipeHolder.java
@@ -3,6 +_,11 @@
 import net.minecraft.resources.ResourceLocation;
 
 public record RecipeHolder<T extends Recipe<?>>(ResourceLocation f_291676_, T f_291008_) {
+   public final org.bukkit.inventory.Recipe toBukkitRecipe() {
+      return this.f_291008_.toBukkitRecipe(org.bukkit.craftbukkit.v1_20_R2.util.CraftNamespacedKey.fromMinecraft(this.f_291676_));
+   }
+   // CraftBukkit end
+
    public boolean equals(Object p_298053_) {
       if (this == p_298053_) {
          return true;
