--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -51,10 +_,10 @@
    private BlockState blockState = Blocks.SAND.defaultBlockState();
    public int time;
    public boolean dropItem = true;
-   private boolean cancelDrop;
-   private boolean hurtEntities;
-   private int fallDamageMax = 40;
-   private float fallDamagePerDistance;
+   public boolean cancelDrop;
+   public boolean hurtEntities;
+   public int fallDamageMax = 40;
+   public float fallDamagePerDistance;
    @Nullable
    public CompoundTag blockData;
    protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
@@ -75,11 +_,23 @@
       this.setStartPos(this.blockPosition());
    }
 
+   private static java.util.concurrent.atomic.AtomicReference<org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason> ketting$spawnReason = new java.util.concurrent.atomic.AtomicReference<>(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT); // Ketting
    public static FallingBlockEntity fall(Level p_201972_, BlockPos p_201973_, BlockState p_201974_) {
-      FallingBlockEntity fallingblockentity = new FallingBlockEntity(p_201972_, (double)p_201973_.getX() + 0.5D, (double)p_201973_.getY(), (double)p_201973_.getZ() + 0.5D, p_201974_.hasProperty(BlockStateProperties.WATERLOGGED) ? p_201974_.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(false)) : p_201974_);
+      FallingBlockEntity fallingblockentity = new FallingBlockEntity(p_201972_, (double) p_201973_.getX() + 0.5D, (double) p_201973_.getY(), (double) p_201973_.getZ() + 0.5D, p_201974_.hasProperty(BlockStateProperties.WATERLOGGED) ? (BlockState) p_201974_.setValue(BlockStateProperties.WATERLOGGED, false) : p_201974_);
+      if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(fallingblockentity, p_201973_, p_201974_.getFluidState().createLegacyBlock())) return fallingblockentity; // CraftBukkit
+
       p_201972_.setBlock(p_201973_, p_201974_.getFluidState().createLegacyBlock(), 3);
+      p_201972_.prepareAddFreshEntity(ketting$spawnReason.getAndSet(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT)); // Ketting
       p_201972_.addFreshEntity(fallingblockentity);
       return fallingblockentity;
+      
+   }
+   
+   // CraftBukkit start
+   public static FallingBlockEntity fall(Level p_201972_, BlockPos p_201973_, BlockState p_201974_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason) {
+      // CraftBukkit end
+      ketting$spawnReason.set(spawnReason); // Ketting
+      return fall(p_201972_, p_201973_, p_201974_);
    }
 
    public boolean isAttackable() {
@@ -120,11 +_,11 @@
          if (!this.level().isClientSide) {
             BlockPos blockpos = this.blockPosition();
             boolean flag = this.blockState.getBlock() instanceof ConcretePowderBlock;
-            boolean flag1 = flag && this.level().getFluidState(blockpos).is(FluidTags.WATER);
+            boolean flag1 = flag && this.blockState.canBeHydrated(this.level(), blockpos, this.level().getFluidState(blockpos), blockpos);
             double d0 = this.getDeltaMovement().lengthSqr();
             if (flag && d0 > 1.0D) {
                BlockHitResult blockhitresult = this.level().clip(new ClipContext(new Vec3(this.xo, this.yo, this.zo), this.position(), ClipContext.Block.COLLIDER, ClipContext.Fluid.SOURCE_ONLY, this));
-               if (blockhitresult.getType() != HitResult.Type.MISS && this.level().getFluidState(blockhitresult.getBlockPos()).is(FluidTags.WATER)) {
+               if (blockhitresult.getType() != HitResult.Type.MISS && this.blockState.canBeHydrated(this.level(), blockpos, this.level().getFluidState(blockhitresult.getBlockPos()), blockhitresult.getBlockPos())) {
                   blockpos = blockhitresult.getBlockPos();
                   flag1 = true;
                }
@@ -151,6 +_,13 @@
                            this.blockState = this.blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true));
                         }
 
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, this.blockState)) {
+                           this.discard(); // SPIGOT-6586 called before the event in previous versions
+                           return;
+                        }
+                        // CraftBukkit end
+
                         if (this.level().setBlock(blockpos, this.blockState, 3)) {
                            ((ServerLevel)this.level()).getChunkSource().chunkMap.broadcast(this, new ClientboundBlockUpdatePacket(blockpos, this.level().getBlockState(blockpos)));
                            this.discard();
@@ -228,7 +_,9 @@
             DamageSource damagesource = damagesource1;
             float f = (float)Math.min(Mth.floor((float)i * this.fallDamagePerDistance), this.fallDamageMax);
             this.level().getEntities(this, this.getBoundingBox(), predicate).forEach((p_149649_) -> {
+               org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.entityDamage = this; // CraftBukkit
                p_149649_.hurt(damagesource, f);
+                org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.entityDamage = null; // CraftBukkit
             });
             boolean flag = this.blockState.is(BlockTags.ANVIL);
             if (flag && f > 0.0F && this.random.nextFloat() < 0.05F + (float)i * 0.05F) {
