--- a/net/minecraft/world/entity/monster/Pillager.java
+++ b/net/minecraft/world/entity/monster/Pillager.java
@@ -56,7 +_,7 @@
    private static final int INVENTORY_SIZE = 5;
    private static final int SLOT_OFFSET = 300;
    private static final float CROSSBOW_POWER = 1.6F;
-   private final SimpleContainer inventory = new SimpleContainer(5);
+   public final SimpleContainer inventory = new SimpleContainer(5);
 
    public Pillager(EntityType<? extends Pillager> p_33262_, Level p_33263_) {
       super(p_33262_, p_33263_);
@@ -109,7 +_,7 @@
    public AbstractIllager.IllagerArmPose getArmPose() {
       if (this.isChargingCrossbow()) {
          return AbstractIllager.IllagerArmPose.CROSSBOW_CHARGE;
-      } else if (this.isHolding(Items.CROSSBOW)) {
+      } else if (this.isHolding(is -> is.getItem() instanceof net.minecraft.world.item.CrossbowItem)) {
          return AbstractIllager.IllagerArmPose.CROSSBOW_HOLD;
       } else {
          return this.isAggressive() ? AbstractIllager.IllagerArmPose.ATTACKING : AbstractIllager.IllagerArmPose.NEUTRAL;
