--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -360,6 +_,20 @@
             if (blockpos1.getY() > this.level().getMinBuildHeight() && this.level().isEmptyBlock(blockpos1) && this.level().getWorldBorder().isWithinBounds(blockpos1) && this.level().noCollision(this, (new AABB(blockpos1)).deflate(1.0E-6D))) {
                Direction direction = this.findAttachableSurface(blockpos1);
                if (direction != null) {
+                  net.minecraftforge.event.entity.EntityTeleportEvent.EnderEntity event = net.minecraftforge.event.ForgeEventFactory.onEnderTeleport(this, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ());
+                  if (event.isCanceled()) direction = null;
+                  blockpos1 = BlockPos.containing(event.getTargetX(), event.getTargetY(), event.getTargetZ());
+               }
+
+               if (direction != null) {
+                  // CraftBukkit start
+                  org.bukkit.event.entity.EntityTeleportEvent teleportEvent = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTeleportEvent(this, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ());
+                  if (teleportEvent.isCancelled()) {
+                     return false;
+                  } else {
+                     blockpos1 = org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation.toBlockPosition(teleportEvent.getTo());
+                  }
+                  // CraftBukkit end
                   this.unRide();
                   this.setAttachFace(direction);
                   this.playSound(SoundEvents.SHULKER_TELEPORT, 1.0F, 1.0F);
@@ -423,6 +_,7 @@
             if (shulker != null) {
                shulker.setVariant(this.getVariant());
                shulker.moveTo(vec3);
+               this.level().prepareAddFreshEntity(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING);
                this.level().addFreshEntity(shulker);
             }
 
