--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -83,7 +_,7 @@
    }
 
    protected void m_149840_() {
-      this.m_21406_(EntityType.f_20481_, true);
+      this.convertTo(EntityType.f_20481_, true, org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN); // CraftBukkit - add spawn and transform reasons
       if (!this.m_20067_()) {
          this.m_9236_().m_5898_((Player)null, 1048, this.m_20183_(), 0);
       }
