--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -83,6 +_,9 @@
    }
 
    protected void doFreezeConversion() {
+      // CraftBukkit - add spawn and transform reasons
+      ketting$convertTo$transformReason.set(org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN); //Ketting
+      ketting$convertTo$spawnReason.set(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN); //Ketting
       this.convertTo(EntityType.STRAY, true);
       if (!this.isSilent()) {
          this.level().levelEvent((Player)null, 1048, this.blockPosition(), 0);
