--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -18,7 +_,7 @@
    private static final EntityDataAccessor<Boolean> DATA_STRAY_CONVERSION_ID = SynchedEntityData.defineId(Skeleton.class, EntityDataSerializers.BOOLEAN);
    public static final String CONVERSION_TAG = "StrayConversionTime";
    private int inPowderSnowTime;
-   private int conversionTime;
+   public int conversionTime;
 
    public Skeleton(EntityType<? extends Skeleton> p_33570_, Level p_33571_) {
       super(p_33570_, p_33571_);
@@ -83,7 +_,7 @@
    }
 
    protected void doFreezeConversion() {
-      this.convertTo(EntityType.STRAY, true);
+      this.convertTo(EntityType.STRAY, true, org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN); // CraftBukkit - add spawn and transform reasons
       if (!this.isSilent()) {
          this.level().levelEvent((Player)null, 1048, this.blockPosition(), 0);
       }
