--- a/net/minecraft/world/entity/monster/SpellcasterIllager.java
+++ b/net/minecraft/world/entity/monster/SpellcasterIllager.java
@@ -183,6 +_,11 @@
       public void m_8037_() {
          --this.f_33774_;
          if (this.f_33774_ == 0) {
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleEntitySpellCastEvent(SpellcasterIllager.this, this.m_7269_())) {
+               return;
+            }
+            // CraftBukkit end
             this.m_8130_();
             SpellcasterIllager.this.m_5496_(SpellcasterIllager.this.m_7894_(), 1.0F, 1.0F);
          }
