--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -280,6 +_,8 @@
       }
 
       public void start() {
+         Vex.this.ketting$setTarget$reason.set(org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET);
+         Vex.this.ketting$setTarget$fireEvent.set(true);
          Vex.this.setTarget(Vex.this.owner.getTarget());
          super.start();
       }
