--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -144,6 +_,8 @@
       }).filter((p_289463_) -> {
          return !p_289463_.isAlliedTo(this.getTarget());
       }).forEach((p_289464_) -> {
+         p_289464_.ketting$setTarget$reason.set(org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY);
+         p_289464_.ketting$setTarget$fireEvent.set(true);
          p_289464_.setTarget(this.getTarget());
       });
    }
@@ -152,6 +_,7 @@
       this.playSound(SoundEvents.ZOMBIFIED_PIGLIN_ANGRY, this.getSoundVolume() * 2.0F, this.getVoicePitch() * 1.8F);
    }
 
+   @Override
    public void setTarget(@Nullable LivingEntity p_34478_) {
       if (this.getTarget() == null && p_34478_ != null) {
          this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
@@ -166,7 +_,16 @@
    }
 
    public void startPersistentAngerTimer() {
-      this.setRemainingPersistentAngerTime(PERSISTENT_ANGER_TIME.sample(this.random));
+      // CraftBukkit start
+      net.minecraft.world.entity.Entity entity = ((ServerLevel) this.level()).getEntity(getPersistentAngerTarget());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), ZombifiedPiglin.PERSISTENT_ANGER_TIME.sample(this.random));
+      this.level().getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.setPersistentAngerTarget(null);
+         return;
+      }
+      this.setRemainingPersistentAngerTime(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean checkZombifiedPiglinSpawnRules(EntityType<ZombifiedPiglin> p_219174_, LevelAccessor p_219175_, MobSpawnType p_219176_, BlockPos p_219177_, RandomSource p_219178_) {
