--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -89,7 +_,9 @@
             list.add(DataProvider.saveStable(p_254020_, p_253413_.serializeRecipe(), this.recipePathProvider.json(p_253413_.getId())));
             JsonObject jsonobject = p_253413_.serializeAdvancement();
             if (jsonobject != null) {
-               list.add(DataProvider.saveStable(p_254020_, jsonobject, this.advancementPathProvider.json(p_253413_.getAdvancementId())));
+               var saveAdvancementFuture = saveAdvancement(p_254020_, p_253413_, jsonobject);
+               if (saveAdvancementFuture != null)
+                  list.add(saveAdvancementFuture);
             }
 
          }
@@ -99,13 +_,23 @@
       }));
    }
 
+   /**
+    * Called every time a recipe is saved to also save the advancement JSON if it exists.
+    *
+    * @return A completable future that saves the advancement to disk, or null to cancel saving the advancement.
+    */
+   @org.jetbrains.annotations.Nullable
+   protected CompletableFuture<?> saveAdvancement(CachedOutput output, FinishedRecipe finishedRecipe, JsonObject advancementJson) {
+      return DataProvider.saveStable(output, advancementJson, this.advancementPathProvider.json(finishedRecipe.getAdvancementId()));
+   }
+
    protected CompletableFuture<?> buildAdvancement(CachedOutput p_253674_, ResourceLocation p_254102_, Advancement.Builder p_253712_) {
       return DataProvider.saveStable(p_253674_, p_253712_.serializeToJson(), this.advancementPathProvider.json(p_254102_));
    }
 
    protected abstract void buildRecipes(Consumer<FinishedRecipe> p_251297_);
 
-   protected static void generateForEnabledBlockFamilies(Consumer<FinishedRecipe> p_249188_, FeatureFlagSet p_251836_) {
+   protected void generateForEnabledBlockFamilies(Consumer<FinishedRecipe> p_249188_, FeatureFlagSet p_251836_) {
       BlockFamilies.getAllFamilies().filter((p_248034_) -> {
          return p_248034_.shouldGenerateRecipe(p_251836_);
       }).forEach((p_176624_) -> {
