--- a/net/minecraft/world/level/levelgen/FlatLevelSource.java
+++ b/net/minecraft/world/level/levelgen/FlatLevelSource.java
@@ -33,7 +_,13 @@
    private final FlatLevelGeneratorSettings settings;
 
    public FlatLevelSource(FlatLevelGeneratorSettings p_256337_) {
-      super(new FixedBiomeSource(p_256337_.getBiome()), Util.memoize(p_256337_::adjustGenerationSettings));
+      // CraftBukkit start
+      this(p_256337_, new FixedBiomeSource(p_256337_.getBiome()));
+   }
+
+   public FlatLevelSource(FlatLevelGeneratorSettings p_256337_, net.minecraft.world.level.biome.BiomeSource manager) {
+      super(manager, Util.memoize(p_256337_::adjustGenerationSettings));
+      // CraftBukkit end
       this.settings = p_256337_;
    }
 
@@ -43,6 +_,7 @@
             return p_255579_;
          });
       });
+      ChunkGeneratorStructureState.ketting$createFor$conf.set(ketting$createState$conf.getAndSet(null)); // Ketting
       return ChunkGeneratorStructureState.createForFlat(p_255830_, p_256355_, this.biomeSource, stream);
    }
 
