--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -82,6 +_,11 @@
       return false;
    }
 
+   // CraftBukkit start
+   public void setStrengthPublic(int i) {
+      this.m_30840_(i);
+   }
+   // CraftBukkit end
    private void m_30840_(int p_30841_) {
       this.f_19804_.m_135381_(f_30745_, Math.max(1, Math.min(5, p_30841_)));
    }
