--- a/net/minecraft/world/scores/criteria/ObjectiveCriteria.java
+++ b/net/minecraft/world/scores/criteria/ObjectiveCriteria.java
@@ -13,7 +_,7 @@
 
 public class ObjectiveCriteria {
    private static final Map<String, ObjectiveCriteria> CUSTOM_CRITERIA = Maps.newHashMap();
-   private static final Map<String, ObjectiveCriteria> CRITERIA_CACHE = Maps.newHashMap();
+   public static final Map<String, ObjectiveCriteria> CRITERIA_CACHE = Maps.newHashMap();
    public static final ObjectiveCriteria DUMMY = registerCustom("dummy");
    public static final ObjectiveCriteria TRIGGER = registerCustom("trigger");
    public static final ObjectiveCriteria DEATH_COUNT = registerCustom("deathCount");
