--- a/net/minecraft/world/entity/player/Abilities.java
+++ b/net/minecraft/world/entity/player/Abilities.java
@@ -8,8 +_,8 @@
    public boolean mayfly;
    public boolean instabuild;
    public boolean mayBuild = true;
-   private float flyingSpeed = 0.05F;
-   private float walkingSpeed = 0.1F;
+   public float flyingSpeed = 0.05F;
+   public float walkingSpeed = 0.1F;
 
    public void addSaveData(CompoundTag p_35946_) {
       CompoundTag compoundtag = new CompoundTag();
