--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -54,7 +_,7 @@
       }).toArray((p_11070_) -> {
          return new String[p_11070_];
       });
-      if (p_11087_.usesAuthentication()) {
+      if (p_11087_.usesAuthentication() || org.spigotmc.SpigotConfig.bungee) { // Spigot: bungee = online mode, for now.
          p_11087_.getProfileRepository().findProfilesByNames(astring, Agent.MINECRAFT, p_11089_);
       } else {
          for(String s : astring) {
@@ -73,7 +_,7 @@
             try {
                userbanlist.load();
             } catch (IOException ioexception1) {
-               LOGGER.warn("Could not load existing file {}", userbanlist.getFile().getName(), ioexception1);
+               LOGGER.warn("Could not load existing file {}", userbanlist.getFile().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -126,7 +_,7 @@
             try {
                ipbanlist.load();
             } catch (IOException ioexception1) {
-               LOGGER.warn("Could not load existing file {}", ipbanlist.getFile().getName(), ioexception1);
+               LOGGER.warn("Could not load existing file {}", ipbanlist.getFile().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -162,7 +_,7 @@
             try {
                serveroplist.load();
             } catch (IOException ioexception1) {
-               LOGGER.warn("Could not load existing file {}", serveroplist.getFile().getName(), ioexception1);
+               LOGGER.warn("Could not load existing file {}", serveroplist.getFile().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -204,7 +_,7 @@
             try {
                userwhitelist.load();
             } catch (IOException ioexception1) {
-               LOGGER.warn("Could not load existing file {}", userwhitelist.getFile().getName(), ioexception1);
+               LOGGER.warn("Could not load existing file {}", userwhitelist.getFile().getName()); // CraftBukkit - don't print stacktrace
             }
          }
 
@@ -315,6 +_,31 @@
                private void movePlayerFile(File p_11168_, String p_11169_, String p_11170_) {
                   File file5 = new File(file1, p_11169_ + ".dat");
                   File file6 = new File(p_11168_, p_11170_ + ".dat");
+
+                  // CraftBukkit start - Use old file name to seed lastKnownName
+                  net.minecraft.nbt.CompoundTag root = null;
+
+                  try {
+                     root = net.minecraft.nbt.NbtIo.readCompressed(new java.io.FileInputStream(file5));
+                  } catch (Exception exception) {
+                     exception.printStackTrace();
+                  }
+
+                  if (root != null) {
+                     if (!root.contains("bukkit")) {
+                        root.put("bukkit", new net.minecraft.nbt.CompoundTag());
+                     }
+                     net.minecraft.nbt.CompoundTag data = root.getCompound("bukkit");
+                     data.putString("lastKnownName", p_11169_);
+
+                     try {
+                        net.minecraft.nbt.NbtIo.writeCompressed(root, new java.io.FileOutputStream(file2));
+                     } catch (Exception exception) {
+                        exception.printStackTrace();
+                     }
+                  }
+                  // CraftBukkit end
+
                   OldUsersConverter.ensureDirectoryExists(p_11168_);
                   if (!file5.renameTo(file6)) {
                      throw new OldUsersConverter.ConversionError("Could not convert file for " + p_11169_);
