--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -480,7 +_,8 @@
 
    protected void pickUpItem(ItemEntity p_28514_) {
       ItemStack itemstack = p_28514_.getItem();
-      if (this.canHoldItem(itemstack)) {
+      if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityPickupItemEvent(this, p_28514_, itemstack.getCount() - 1, !this.canHoldItem(itemstack)).isCancelled()) { // CraftBukkit - call EntityPickupItemEvent
+         itemstack = p_28514_.getItem(); // CraftBukkit - update ItemStack from event
          int i = itemstack.getCount();
          if (i > 1) {
             this.dropItemStack(itemstack.split(i - 1));
@@ -653,13 +_,16 @@
    }
 
    protected void dropAllDeathLoot(DamageSource p_28536_) {
+      super.dropAllDeathLoot(p_28536_);
+   }
+   protected void dropEquipment() { // Forge: move extra drops to dropEquipment to allow them to be captured by LivingDropsEvent
+      super.dropEquipment();
       ItemStack itemstack = this.getItemBySlot(EquipmentSlot.MAINHAND);
       if (!itemstack.isEmpty()) {
          this.spawnAtLocation(itemstack);
          this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
       }
 
-      super.dropAllDeathLoot(p_28536_);
    }
 
    public static boolean isPathClear(Fox p_28472_, LivingEntity p_28473_) {
@@ -809,6 +_,17 @@
       protected void breed() {
          ServerLevel serverlevel = (ServerLevel)this.level;
          Fox fox = (Fox)this.animal.getBreedOffspring(serverlevel, this.partner);
+         final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(animal, partner, fox);
+         final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         fox = (Fox) event.getChild();
+         if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.animal.setAge(6000);
+            this.partner.setAge(6000);
+            this.animal.resetLove();
+            this.partner.resetLove();
+            return;
+         }
          if (fox != null) {
             ServerPlayer serverplayer = this.animal.getLoveCause();
             ServerPlayer serverplayer1 = this.partner.getLoveCause();
@@ -819,6 +_,17 @@
                serverplayer2 = serverplayer1;
             }
 
+            // CraftBukkit start - call EntityBreedEvent
+            fox.setAge(-24000);
+            fox.moveTo(this.animal.getX(), this.animal.getY(), this.animal.getZ(), 0.0F, 0.0F);
+            int experience = this.animal.getRandom().nextInt(7) + 1;
+            org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityBreedEvent(fox, animal, partner, serverplayer, this.animal.breedItem, experience);
+            if (entityBreedEvent.isCancelled()) {
+               return;
+            }
+            experience = entityBreedEvent.getExperience();
+            // CraftBukkit end
+
             if (serverplayer1 != null && serverplayer != serverplayer1) {
                fox.addTrustedUUID(serverplayer1.getUUID());
             }
@@ -832,12 +_,15 @@
             this.partner.setAge(6000);
             this.animal.resetLove();
             this.partner.resetLove();
-            fox.setAge(-24000);
-            fox.moveTo(this.animal.getX(), this.animal.getY(), this.animal.getZ(), 0.0F, 0.0F);
-            serverlevel.addFreshEntityWithPassengers(fox);
+            serverlevel.addFreshEntityWithPassengers(fox, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
             this.level.broadcastEntityEvent(this.animal, (byte)18);
             if (this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
                this.level.addFreshEntity(new ExperienceOrb(this.level, this.animal.getX(), this.animal.getY(), this.animal.getZ(), this.animal.getRandom().nextInt(7) + 1));
+               // CraftBukkit start - use event experience
+               if (experience > 0) {
+                  this.level.addFreshEntity(new ExperienceOrb(this.level, this.animal.getX(), this.animal.getY(), this.animal.getZ(), experience));
+               }
+               // CraftBukkit end
             }
 
          }
@@ -880,7 +_,7 @@
       }
 
       protected void onReachedTarget() {
-         if (Fox.this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+         if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(Fox.this.level(), Fox.this)) {
             BlockState blockstate = Fox.this.level().getBlockState(this.blockPos);
             if (blockstate.is(Blocks.SWEET_BERRY_BUSH)) {
                this.pickSweetBerries(blockstate);
@@ -897,7 +_,12 @@
 
       private void pickSweetBerries(BlockState p_148929_) {
          int i = p_148929_.getValue(SweetBerryBushBlock.AGE);
-         p_148929_.setValue(SweetBerryBushBlock.AGE, Integer.valueOf(1));
+         p_148929_.setValue(SweetBerryBushBlock.AGE, 1);
+         // CraftBukkit start - call EntityChangeBlockEvent
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(Fox.this, this.blockPos, p_148929_.setValue(SweetBerryBushBlock.AGE, 1))) {
+            return;
+         }
+         // CraftBukkit end
          int j = 1 + Fox.this.level().random.nextInt(2) + (i == 3 ? 1 : 0);
          ItemStack itemstack = Fox.this.getItemBySlot(EquipmentSlot.MAINHAND);
          if (itemstack.isEmpty()) {
@@ -935,7 +_,7 @@
       }
 
       public boolean canUse() {
-         return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25D || Fox.this.isInLava();
+         return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25D || Fox.this.isInLava() || Fox.this.isInFluidType((fluidType, height) -> Fox.this.canSwimInFluidType(fluidType) && height > 0.25D);
       }
    }
 
