--- a/net/minecraft/world/entity/animal/Ocelot.java
+++ b/net/minecraft/world/entity/animal/Ocelot.java
@@ -160,7 +_,7 @@
       if ((this.temptGoal == null || this.temptGoal.isRunning()) && !this.isTrusting() && this.isFood(itemstack) && p_29021_.distanceToSqr(this) < 9.0D) {
          this.usePlayerItem(p_29021_, p_29022_, itemstack);
          if (!this.level().isClientSide) {
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_29021_) && !org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTameEvent(this, p_29021_).isCancelled()) { // CraftBukkit - added event call and isCancelled check
                this.setTrusting(true);
                this.spawnTrustingParticles(true);
                this.level().broadcastEntityEvent(this, (byte)41);
