--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -202,13 +_,13 @@
       }
    }
 
-   public static LegacyStructureDataHandler m_71331_(ResourceKey<Level> p_71332_, @Nullable DimensionDataStorage p_71333_) {
-      if (p_71332_ == Level.f_46428_) {
+   public static LegacyStructureDataHandler m_71331_(ResourceKey<net.minecraft.world.level.dimension.LevelStem> p_71332_, @Nullable DimensionDataStorage p_71333_) { // CraftBukkit
+      if (p_71332_ == net.minecraft.world.level.dimension.LevelStem.f_63971_) { // CraftBukkit
          return new LegacyStructureDataHandler(p_71333_, ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument"));
-      } else if (p_71332_ == Level.f_46429_) {
+      } else if (p_71332_ == net.minecraft.world.level.dimension.LevelStem.f_63972_) { // CraftBukkit
          List<String> list1 = ImmutableList.of("Fortress");
          return new LegacyStructureDataHandler(p_71333_, list1, list1);
-      } else if (p_71332_ == Level.f_46430_) {
+      } else if (p_71332_ == net.minecraft.world.level.dimension.LevelStem.f_63973_) { // CraftBukkit
          List<String> list = ImmutableList.of("EndCity");
          return new LegacyStructureDataHandler(p_71333_, list, list);
       } else {
