--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -202,13 +_,16 @@
       }
    }
 
-   public static LegacyStructureDataHandler getLegacyStructureHandler(ResourceKey<Level> p_71332_, @Nullable DimensionDataStorage p_71333_) {
-      if (p_71332_ == Level.OVERWORLD) {
+   //Ketting - Todo: yes, this is a potential mixin inject error.
+   //                Patching this out seems to be hard, though.
+   //                For now we'll just hope nobody needs the mixin into the original method.
+   public static LegacyStructureDataHandler getLegacyStructureHandler(ResourceKey<net.minecraft.world.level.dimension.LevelStem> p_71332_, @Nullable DimensionDataStorage p_71333_) { // CraftBukkit
+      if (p_71332_ == net.minecraft.world.level.dimension.LevelStem.OVERWORLD) { // CraftBukkit
          return new LegacyStructureDataHandler(p_71333_, ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument"));
-      } else if (p_71332_ == Level.NETHER) {
+      } else if (p_71332_ == net.minecraft.world.level.dimension.LevelStem.NETHER) { // CraftBukkit
          List<String> list1 = ImmutableList.of("Fortress");
          return new LegacyStructureDataHandler(p_71333_, list1, list1);
-      } else if (p_71332_ == Level.END) {
+      } else if (p_71332_ == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit
          List<String> list = ImmutableList.of("EndCity");
          return new LegacyStructureDataHandler(p_71333_, list, list);
       } else {
