--- a/net/minecraft/world/level/levelgen/structure/StructureCheck.java
+++ b/net/minecraft/world/level/levelgen/structure/StructureCheck.java
@@ -111,7 +_,7 @@
          if (i <= 1493) {
             return StructureCheckResult.CHUNK_LOAD_NEEDED;
          } else {
-            ChunkStorage.m_196918_(compoundtag, this.f_197241_, this.f_197242_.m_187743_());
+            ChunkStorage.m_196918_(compoundtag, this.typeKey, this.f_197242_.m_187743_());
 
             CompoundTag compoundtag1;
             try {
@@ -211,4 +_,12 @@
          return p_226746_;
       });
    }
+
+   //Ketting start - type key
+   private ResourceKey<net.minecraft.world.level.dimension.LevelStem> typeKey;
+
+   public void setTypeKey(ResourceKey<net.minecraft.world.level.dimension.LevelStem> typeKey) {
+      this.typeKey = typeKey;
+   }
+   //Ketting end
 }
