--- a/net/minecraft/stats/RecipeBook.java
+++ b/net/minecraft/stats/RecipeBook.java
@@ -9,7 +_,7 @@
 import net.minecraft.world.item.crafting.RecipeHolder;
 
 public class RecipeBook {
-   protected final Set<ResourceLocation> known = Sets.newHashSet();
+   public final Set<ResourceLocation> known = Sets.newHashSet();
    protected final Set<ResourceLocation> highlight = Sets.newHashSet();
    private final RecipeBookSettings bookSettings = new RecipeBookSettings();
 
