--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -29,7 +_,7 @@
 
       for(Recipe<?> recipe : p_12792_) {
          ResourceLocation resourcelocation = recipe.getId();
-         if (!this.known.contains(resourcelocation) && !recipe.isSpecial()) {
+         if (!this.known.contains(resourcelocation) && !recipe.isSpecial() && org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handlePlayerRecipeListUpdateEvent(p_12793_, resourcelocation)) { // CraftBukkit
             this.add(resourcelocation);
             this.addHighlight(resourcelocation);
             list.add(resourcelocation);
@@ -63,6 +_,7 @@
    }
 
    private void sendRecipes(ClientboundRecipePacket.State p_12802_, ServerPlayer p_12803_, List<ResourceLocation> p_12804_) {
+      if (p_12803_.connection == null) return; // SPIGOT-4478 during PlayerLoginEvent
       p_12803_.connection.send(new ClientboundRecipePacket(p_12802_, p_12804_, Collections.emptyList(), this.getBookSettings()));
    }
 
