--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -125,13 +_,13 @@
                this.yHeadRot = (float)this.random.nextInt(360);
             }
 
-            if (this.level().getNearestPlayer(BAT_RESTING_TARGETING, this) != null) {
+            if (this.level().getNearestPlayer(Bat.BAT_RESTING_TARGETING, this) != null && org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
                this.setResting(false);
                if (!flag) {
                   this.level().levelEvent((Player)null, 1025, blockpos, 0);
                }
             }
-         } else {
+         } else if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
             this.setResting(false);
             if (!flag) {
                this.level().levelEvent((Player)null, 1025, blockpos, 0);
@@ -156,7 +_,7 @@
          float f1 = Mth.wrapDegrees(f - this.getYRot());
          this.zza = 0.5F;
          this.setYRot(this.getYRot() + f1);
-         if (this.random.nextInt(100) == 0 && this.level().getBlockState(blockpos1).isRedstoneConductor(this.level(), blockpos1)) {
+         if (this.random.nextInt(100) == 0 && this.level().getBlockState(blockpos1).isRedstoneConductor(this.level(), blockpos1) && org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBatToggleSleepEvent(this, false)) { // CraftBukkit - Call BatToggleSleepEvent
             this.setResting(true);
          }
       }
@@ -178,7 +_,7 @@
       if (this.isInvulnerableTo(p_27424_)) {
          return false;
       } else {
-         if (!this.level().isClientSide && this.isResting()) {
+         if (!this.level().isClientSide && this.isResting() && org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) { // CraftBukkit - Call BatToggleSleepEvent
             this.setResting(false);
          }
 
