--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -89,7 +_,7 @@
          this.timeInOverworld = 0;
       }
 
-      if (this.timeInOverworld > 300) {
+      if (this.timeInOverworld > 300 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> this.timeInOverworld = timer)) {
          this.playConvertedSound();
          this.finishConversion((ServerLevel)this.level());
       }
@@ -101,9 +_,12 @@
    }
 
    protected void finishConversion(ServerLevel p_34663_) {
+      ketting$convertTo$transformReason.set(org.bukkit.event.entity.EntityTransformEvent.TransformReason.PIGLIN_ZOMBIFIED); // Ketting
+      ketting$convertTo$spawnReason.set(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PIGLIN_ZOMBIFIED); // Ketting
       ZombifiedPiglin zombifiedpiglin = this.convertTo(EntityType.ZOMBIFIED_PIGLIN, true);
       if (zombifiedpiglin != null) {
          zombifiedpiglin.addEffect(new MobEffectInstance(MobEffects.CONFUSION, 200, 0));
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombifiedpiglin);
       }
 
    }
