--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -20,6 +_,8 @@
    private double pRotY;
    @Nullable
    protected Player player;
+   @Nullable
+   protected LivingEntity bukkitPlayer; //Ketting
    private int calmDown;
    private boolean isRunning;
    private final Ingredient items;
@@ -40,7 +_,19 @@
          return false;
       } else {
          this.player = this.mob.level().getNearestPlayer(this.targetingConditions, this.mob);
-         return this.player != null;
+         //Ketting start - bukkitPlayer
+         this.bukkitPlayer = this.player;
+         // CraftBukkit start
+         if (this.bukkitPlayer != null) {
+            org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTargetLivingEvent(this.mob, this.bukkitPlayer, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TEMPT);
+            if (event.isCancelled()) {
+               return false;
+            }
+            this.bukkitPlayer = (event.getTarget() == null) ? null : ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) event.getTarget()).getHandle();
+         }
+         // CraftBukkit end
+         return this.bukkitPlayer != null;
+         //Ketting end
       }
    }
 
@@ -84,6 +_,7 @@
 
    public void stop() {
       this.player = null;
+      this.bukkitPlayer = null; //Ketting
       this.mob.getNavigation().stop();
       this.calmDown = reducedTickDelay(100);
       this.isRunning = false;
