--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -19,7 +_,7 @@
    private double f_25931_;
    private double f_25932_;
    @Nullable
-   protected Player f_25925_;
+   protected LivingEntity f_25925_; // CraftBukkit
    private int f_25933_;
    private boolean f_25934_;
    private final Ingredient f_25935_;
@@ -40,6 +_,15 @@
          return false;
       } else {
          this.f_25925_ = this.f_25924_.m_9236_().m_45946_(this.f_148137_, this.f_25924_);
+         // CraftBukkit start
+         if (this.f_25925_ != null) {
+            org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTargetLivingEvent(this.f_25924_, this.f_25925_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TEMPT);
+            if (event.isCancelled()) {
+               return false;
+            }
+            this.f_25925_ = (event.getTarget() == null) ? null : ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) event.getTarget()).getHandle();
+         }
+         // CraftBukkit end
          return this.f_25925_ != null;
       }
    }
