--- a/net/minecraft/world/CompoundContainer.java
+++ b/net/minecraft/world/CompoundContainer.java
@@ -4,8 +_,50 @@
 import net.minecraft.world.item.ItemStack;
 
 public class CompoundContainer implements Container {
-   private final Container container1;
-   private final Container container2;
+   public final Container container1;
+   public final Container container2;
+
+   // CraftBukkit start - add fields and methods
+   public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+
+   public java.util.List<ItemStack> getContents() {
+      java.util.List<ItemStack> result = new java.util.ArrayList<ItemStack>(this.getContainerSize());
+      for (int i = 0; i < this.getContainerSize(); i++) {
+         result.add(this.getItem(i));
+      }
+      return result;
+   }
+
+   public void onOpen(org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity who) {
+      this.container1.onOpen(who);
+      this.container2.onOpen(who);
+      transaction.add(who);
+   }
+
+   public void onClose(org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity who) {
+      this.container1.onClose(who);
+      this.container2.onClose(who);
+      transaction.remove(who);
+   }
+
+   public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return null; // This method won't be called since CraftInventoryDoubleChest doesn't defer to here
+   }
+
+   public void setMaxStackSize(int size) {
+      this.container1.setMaxStackSize(size);
+      this.container2.setMaxStackSize(size);
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return container1.getLocation(); // TODO: right?
+   }
+   // CraftBukkit end
 
    public CompoundContainer(Container p_18913_, Container p_18914_) {
       this.container1 = p_18913_;
@@ -46,7 +_,7 @@
    }
 
    public int getMaxStackSize() {
-      return this.container1.getMaxStackSize();
+      return Math.min(this.container1.getMaxStackSize(), this.container2.getMaxStackSize()); // CraftBukkit - check both sides
    }
 
    public void setChanged() {
