--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -78,10 +_,14 @@
    }
 
    static boolean stillValidBlockEntity(BlockEntity p_273154_, Player p_273222_) {
-      return stillValidBlockEntity(p_273154_, p_273222_, 8);
+      return stillValidBlockEntity(p_273154_, p_273222_, p_273222_.getAttributeValue(net.minecraftforge.common.ForgeMod.BLOCK_REACH.get()) + 3.5); // Vanilla uses 8, default reach is 4.5 so add 3.5 padding
    }
 
    static boolean stillValidBlockEntity(BlockEntity p_272877_, Player p_272670_, int p_273411_) {
+      return stillValidBlockEntity(p_272877_, p_272670_, (double)p_273411_);
+   }
+
+   static boolean stillValidBlockEntity(BlockEntity p_272877_, Player p_272670_, double p_273411_) {
       Level level = p_272877_.getLevel();
       BlockPos blockpos = p_272877_.getBlockPos();
       if (level == null) {
@@ -92,4 +_,43 @@
          return p_272670_.distanceToSqr((double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.5D, (double)blockpos.getZ() + 0.5D) <= (double)(p_273411_ * p_273411_);
       }
    }
+
+   // CraftBukkit start
+   default java.util.List<ItemStack> getContents() {
+      return java.util.Collections.emptyList();
+   }
+
+   //Ketting start - transactions
+   default void onOpen(org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity who){
+      org.kettingpowered.ketting.utils.InventoryViewHelper.getContainerTransactions(this).add(who);
+   }
+
+   default void onClose(org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity who){
+      org.kettingpowered.ketting.utils.InventoryViewHelper.removeContainerTransaction(this, who);
+   }
+
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers(){
+      return org.kettingpowered.ketting.utils.InventoryViewHelper.getContainerTransactions(this);
+   }
+   //Ketting end
+
+   default @javax.annotation.Nullable org.bukkit.inventory.InventoryHolder getOwner(){
+      return null;
+   }
+
+   default void setMaxStackSize(int size){}
+
+   default @javax.annotation.Nullable org.bukkit.Location getLocation(){
+      return null;
+   }
+
+   default @javax.annotation.Nullable net.minecraft.world.item.crafting.Recipe<?> getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(net.minecraft.world.item.crafting.Recipe<?> recipe) {
+   }
+
+   int MAX_STACK = 64;
+   // CraftBukkit end
 }
