--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -92,4 +_,42 @@
          return p_272670_.m_20275_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.5D, (double)blockpos.m_123343_() + 0.5D) <= (double)(p_273411_ * p_273411_);
       }
    }
+
+   // CraftBukkit start
+   java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>(); //Ketting - adding a new field like this is not ideal, but this provides the most correctness
+   default java.util.List<ItemStack> getContents() {
+      return java.util.Collections.emptyList();
+   }
+
+   default void onOpen(org.bukkit.craftbukkit.v1_20_R2.entity.CraftHumanEntity who){
+      transaction.add(who);
+   }
+
+   default void onClose(org.bukkit.craftbukkit.v1_20_R2.entity.CraftHumanEntity who){
+      transaction.remove(who);
+   }
+
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers(){
+      return transaction;
+   }
+
+   default @javax.annotation.Nullable org.bukkit.inventory.InventoryHolder getOwner(){
+      return null;
+   }
+
+   default void setMaxStackSize(int size){}
+
+   default @javax.annotation.Nullable org.bukkit.Location getLocation(){
+      return null;
+   }
+
+   default @javax.annotation.Nullable net.minecraft.world.item.crafting.RecipeHolder<?> getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(net.minecraft.world.item.crafting.RecipeHolder<?> recipe) {
+   }
+
+   int MAX_STACK = 64;
+   // CraftBukkit end
 }
