--- a/net/minecraft/nbt/ByteArrayTag.java
+++ b/net/minecraft/nbt/ByteArrayTag.java
@@ -21,8 +_,11 @@
       private static byte[] readAccounted(DataInput p_301772_, NbtAccounter p_301697_) throws IOException {
          p_301697_.accountBytes(24L);
          int i = p_301772_.readInt();
+         com.google.common.base.Preconditions.checkArgument( i < 1 << 24); // Spigot
+
          p_301697_.accountBytes(1L, (long)i);
          byte[] abyte = new byte[i];
+
          p_301772_.readFully(abyte);
          return abyte;
       }
