--- a/net/minecraft/nbt/IntArrayTag.java
+++ b/net/minecraft/nbt/IntArrayTag.java
@@ -21,6 +_,8 @@
       private static int[] readAccounted(DataInput p_301738_, NbtAccounter p_301754_) throws IOException {
          p_301754_.accountBytes(24L);
          int i = p_301738_.readInt();
+         com.google.common.base.Preconditions.checkArgument( i < 1 << 24); // Spigot
+
          p_301754_.accountBytes(4L, (long)i);
          int[] aint = new int[i];
 
