--- a/net/minecraft/nbt/TagParser.java
+++ b/net/minecraft/nbt/TagParser.java
@@ -93,7 +_,7 @@
       }
    }
 
-   private Tag type(String p_129369_) {
+   public Tag type(String p_129369_) {
       try {
          if (FLOAT_PATTERN.matcher(p_129369_).matches()) {
             return FloatTag.valueOf(Float.parseFloat(p_129369_.substring(0, p_129369_.length() - 1)));
@@ -217,7 +_,7 @@
       }
    }
 
-   private Tag readArrayTag() throws CommandSyntaxException {
+   public Tag readArrayTag() throws CommandSyntaxException {
       this.expect('[');
       int i = this.reader.getCursor();
       char c0 = this.reader.read();
