--- a/net/minecraft/server/packs/repository/Pack.java
+++ b/net/minecraft/server/packs/repository/Pack.java
@@ -27,6 +_,7 @@
    private final Pack.Position defaultPosition;
    private final boolean required;
    private final boolean fixedPosition;
+   private final boolean hidden; // Forge: Allow packs to be hidden from the UI entirely
    private final PackSource packSource;
 
    @Nullable
@@ -50,6 +_,7 @@
       this.defaultPosition = p_251298_;
       this.fixedPosition = p_249753_;
       this.packSource = p_251608_;
+      this.hidden = p_250162_.hidden();
    }
 
    @Nullable
@@ -62,7 +_,8 @@
          } else {
             FeatureFlagsMetadataSection featureflagsmetadatasection = packresources.getMetadataSection(FeatureFlagsMetadataSection.TYPE);
             FeatureFlagSet featureflagset = featureflagsmetadatasection != null ? featureflagsmetadatasection.flags() : FeatureFlagSet.of();
-            return new Pack.Info(packmetadatasection.getDescription(), packmetadatasection.getPackFormat(), featureflagset);
+            // Forge: Allow separate pack formats for server data and client resources and setting isHidden
+            return new Pack.Info(packmetadatasection.getDescription(), packmetadatasection.getPackFormat(PackType.SERVER_DATA), packmetadatasection.getPackFormat(PackType.CLIENT_RESOURCES), featureflagset, packresources.isHidden());
          }
       } catch (Exception exception) {
          LOGGER.warn("Failed to read pack metadata", (Throwable)exception);
@@ -116,6 +_,8 @@
       return this.packSource;
    }
 
+   public boolean isHidden() { return hidden; }
+
    public boolean equals(Object p_10448_) {
       if (this == p_10448_) {
          return true;
@@ -131,9 +_,17 @@
       return this.id.hashCode();
    }
 
-   public static record Info(Component description, int format, FeatureFlagSet requestedFeatures) {
+   public static record Info(Component description, int dataFormat, int resourceFormat, FeatureFlagSet requestedFeatures, boolean hidden) {
+      public Info(Component description, int format, FeatureFlagSet requestedFeatures) {
+         this(description, format, format, requestedFeatures, false);
+      }
+
+      public int getFormat(PackType type) {
+         return type == PackType.SERVER_DATA ? this.dataFormat : this.resourceFormat;
+      }
+
       public PackCompatibility compatibility(PackType p_249204_) {
-         return PackCompatibility.forFormat(this.format, p_249204_);
+         return PackCompatibility.forFormat(getFormat(p_249204_), p_249204_);
       }
    }
 
