--- a/net/minecraft/world/item/ChorusFruitItem.java
+++ b/net/minecraft/world/item/ChorusFruitItem.java
@@ -29,7 +_,22 @@
             }
 
             Vec3 vec3 = p_40714_.position();
-            if (p_40714_.randomTeleport(d0, d1, d2, true)) {
+            var event = net.minecraftforge.event.ForgeEventFactory.onChorusFruitTeleport(p_40714_, d0, d1, d2);
+            if (event.isCanceled()) return itemstack;
+            // CraftBukkit start - handle canceled status of teleport event
+            // Ketting start - Use vanilla method for fix mixin issues
+            p_40714_.ketting$teleportCause.set(org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.CHORUS_FRUIT);
+            p_40714_.randomTeleport(d0, d1, d2, true);
+            java.util.Optional<Boolean> status = p_40714_.ketting$randomTeleportCB;
+            // Ketting end
+
+            if (!status.isPresent()) {
+               // teleport event was canceled, no more tries
+               break;
+            }
+
+            if (status.get()) {
+               // CraftBukkit end
                p_40713_.gameEvent(GameEvent.TELEPORT, vec3, GameEvent.Context.of(p_40714_));
                SoundSource soundsource;
                SoundEvent soundevent;
