--- a/net/minecraft/world/item/ItemCooldowns.java
+++ b/net/minecraft/world/item/ItemCooldowns.java
@@ -6,8 +_,8 @@
 import net.minecraft.util.Mth;
 
 public class ItemCooldowns {
-   private final Map<Item, ItemCooldowns.CooldownInstance> cooldowns = Maps.newHashMap();
-   private int tickCount;
+   public final Map<Item, ItemCooldowns.CooldownInstance> cooldowns = Maps.newHashMap();
+   public int tickCount;
 
    public boolean isOnCooldown(Item p_41520_) {
       return this.getCooldownPercent(p_41520_, 0.0F) > 0.0F;
@@ -56,9 +_,9 @@
    protected void onCooldownEnded(Item p_41531_) {
    }
 
-   static class CooldownInstance {
+   public static class CooldownInstance {
       final int startTime;
-      final int endTime;
+      public final int endTime;
 
       CooldownInstance(int p_186358_, int p_186359_) {
          this.startTime = p_186358_;
