--- a/net/minecraft/world/item/SuspiciousStewItem.java
+++ b/net/minecraft/world/item/SuspiciousStewItem.java
@@ -29,6 +_,7 @@
       ListTag listtag = compoundtag.getList("Effects", 9);
       CompoundTag compoundtag1 = new CompoundTag();
       compoundtag1.putInt("EffectId", MobEffect.getId(p_43260_));
+      net.minecraftforge.common.ForgeHooks.saveMobEffect(compoundtag1, "forge:effect_id", p_43260_);
       compoundtag1.putInt("EffectDuration", p_43261_);
       listtag.add(compoundtag1);
       compoundtag.put("Effects", listtag);
@@ -49,6 +_,7 @@
             }
 
             MobEffect mobeffect = MobEffect.byId(compoundtag1.getInt("EffectId"));
+            mobeffect = net.minecraftforge.common.ForgeHooks.loadMobEffect(compoundtag1, "forge:effect_id", mobeffect);
             if (mobeffect != null) {
                p_259500_.accept(new MobEffectInstance(mobeffect, j));
             }
