--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -75,6 +_,18 @@
       int i = this.m_6693_(p_152145_, p_152146_);
       boolean flag = p_152148_ > 0;
       boolean flag1 = i > 0;
+      // CraftBukkit start - Interact Pressure Plate
+      org.bukkit.World bworld = p_152145_.getWorld();
+      org.bukkit.plugin.PluginManager manager = p_152145_.getCraftServer().getPluginManager();
+
+      if (flag != flag1) {
+         org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(bworld.getBlockAt(p_152146_.m_123341_(), p_152146_.m_123342_(), p_152146_.m_123343_()), p_152148_, i);
+         manager.callEvent(eventRedstone);
+
+         flag1 = eventRedstone.getNewCurrent() > 0;
+         i = eventRedstone.getNewCurrent();
+      }
+      // CraftBukkit end
       if (p_152148_ != i) {
          BlockState blockstate = this.m_7422_(p_152147_, i);
          p_152145_.m_7731_(p_152146_, blockstate, 2);
@@ -124,9 +_,15 @@
    }
 
    protected static int m_289607_(Level p_289656_, AABB p_289647_, Class<? extends Entity> p_289686_) {
-      return p_289656_.m_6443_(p_289686_, p_289647_, EntitySelector.f_20408_.and((p_289691_) -> {
-         return !p_289691_.m_6090_();
-      })).size();
+      // CraftBukkit start
+      return getEntities(p_289656_, p_289647_, p_289686_).size();
+   }
+
+   protected static <T extends Entity> java.util.List<T> getEntities(Level world, AABB axisalignedbb, Class<T> oclass) {
+      // CraftBukkit end
+      return world.m_6443_(oclass, axisalignedbb, EntitySelector.f_20408_.and((entity) -> {
+         return !entity.m_6090_();
+      })); // CraftBukkit
    }
 
    protected abstract int m_6693_(Level p_49336_, BlockPos p_49337_);
