--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -127,6 +_,12 @@
             p_152192_ = p_152190_.getBlockState(p_152191_).getValue(FACING);
          }
 
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBellRingEvent(p_152190_, p_152191_, p_152192_, p_152189_)) {
+            return false;
+         }
+         // CraftBukkit end
+
          ((BellBlockEntity)blockentity).onHit(p_152192_);
          p_152190_.playSound((Player)null, p_152191_, SoundEvents.BELL_BLOCK, SoundSource.BLOCKS, 2.0F, 1.0F);
          p_152190_.gameEvent(p_152189_, GameEvent.BLOCK_CHANGE, p_152191_);
@@ -247,4 +_,16 @@
    public boolean isPathfindable(BlockState p_49717_, BlockGetter p_49718_, BlockPos p_49719_, PathComputationType p_49720_) {
       return false;
    }
+
+   // CraftBukkit start - fix MC-253819
+   @Override
+   public BlockState rotate(BlockState blockstate, Rotation rotation) {
+      return blockstate.setValue(FACING, rotation.rotate(blockstate.getValue(FACING)));
+   }
+
+   @Override
+   public BlockState mirror(BlockState blockstate, Mirror mirror) {
+      return blockstate.rotate(mirror.getRotation(blockstate.getValue(FACING)));
+   }
+   // CraftBukkit end
 }
