--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -136,6 +_,12 @@
             p_152192_ = p_152190_.getBlockState(p_152191_).getValue(FACING);
          }
 
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBellRingEvent(p_152190_, p_152191_, p_152192_, p_152189_)) {
+            return false;
+         }
+         // CraftBukkit end
+
          ((BellBlockEntity)blockentity).onHit(p_152192_);
          p_152190_.playSound((Player)null, p_152191_, SoundEvents.BELL_BLOCK, SoundSource.BLOCKS, 2.0F, 1.0F);
          p_152190_.gameEvent(p_152189_, GameEvent.BLOCK_CHANGE, p_152191_);
