--- a/net/minecraft/world/level/block/CaveVines.java
+++ b/net/minecraft/world/level/block/CaveVines.java
@@ -24,10 +_,28 @@
 
    static InteractionResult m_269473_(@Nullable Entity p_270738_, BlockState p_270772_, Level p_270721_, BlockPos p_270587_) {
       if (p_270772_.m_61143_(f_152949_)) {
-         Block.m_49840_(p_270721_, p_270587_, new ItemStack(Items.f_151079_, 1));
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(p_270738_, p_270587_, (BlockState) p_270772_.m_61124_(CaveVines.f_152949_, false))) {
+            return InteractionResult.SUCCESS;
+         }
+
+         if (p_270738_ instanceof Player) {
+            org.bukkit.event.player.PlayerHarvestBlockEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerHarvestBlockEvent(p_270721_, p_270587_, (Player) p_270738_, net.minecraft.world.InteractionHand.MAIN_HAND, java.util.Collections.singletonList(new ItemStack(Items.f_151079_, 1)));
+            if (event.isCancelled()) {
+               return InteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+            }
+            for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+               Block.m_49840_(p_270721_, p_270587_, org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asNMSCopy(itemStack));
+            }
+         } else {
+            Block.m_49840_(p_270721_, p_270587_, new ItemStack(Items.f_151079_, 1));
+         }
+         // CraftBukkit end
+
          float f = Mth.m_216283_(p_270721_.f_46441_, 0.8F, 1.2F);
          p_270721_.m_5594_((Player)null, p_270587_, SoundEvents.f_144088_, SoundSource.BLOCKS, 1.0F, f);
-         BlockState blockstate = p_270772_.m_61124_(f_152949_, Boolean.valueOf(false));
+         BlockState blockstate = p_270772_.m_61124_(f_152949_, false);
+
          p_270721_.m_7731_(p_270587_, blockstate, 2);
          p_270721_.m_220407_(GameEvent.f_157792_, p_270587_, GameEvent.Context.m_223719_(p_270738_, blockstate));
          return InteractionResult.m_19078_(p_270721_.f_46443_);
