--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -315,7 +_,7 @@
       }
    }
 
-   static class EmptyContainer extends SimpleContainer implements WorldlyContainer {
+   public static class EmptyContainer extends SimpleContainer implements WorldlyContainer {
       public EmptyContainer() {
          super(0);
       }
@@ -333,7 +_,7 @@
       }
    }
 
-   static class InputContainer extends SimpleContainer implements WorldlyContainer {
+   public static class InputContainer extends SimpleContainer implements WorldlyContainer {
       private final BlockState state;
       private final LevelAccessor level;
       private final BlockPos pos;
@@ -374,7 +_,7 @@
       }
    }
 
-   static class OutputContainer extends SimpleContainer implements WorldlyContainer {
+   public static class OutputContainer extends SimpleContainer implements WorldlyContainer {
       private final BlockState state;
       private final LevelAccessor level;
       private final BlockPos pos;
