--- a/net/minecraft/world/level/block/DaylightDetectorBlock.java
+++ b/net/minecraft/world/level/block/DaylightDetectorBlock.java
@@ -61,7 +_,8 @@
 
       i = Mth.clamp(i, 0, 15);
       if (p_52411_.getValue(POWER) != i) {
-         p_52412_.setBlock(p_52413_, p_52411_.setValue(POWER, Integer.valueOf(i)), 3);
+         i = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callRedstoneChange(p_52412_, p_52413_, ((Integer) p_52411_.getValue(POWER)), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
+         p_52412_.setBlock(p_52413_, p_52411_.setValue(POWER, i), 3);
       }
 
    }
