--- a/net/minecraft/world/level/block/InfestedBlock.java
+++ b/net/minecraft/world/level/block/InfestedBlock.java
@@ -39,7 +_,7 @@
       Silverfish silverfish = EntityType.SILVERFISH.create(p_54181_);
       if (silverfish != null) {
          silverfish.moveTo((double)p_54182_.getX() + 0.5D, (double)p_54182_.getY(), (double)p_54182_.getZ() + 0.5D, 0.0F, 0.0F);
-         p_54181_.addFreshEntity(silverfish);
+         p_54181_.addFreshEntity(silverfish, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SILVERFISH_BLOCK); // CraftBukkit - add SpawnReason
          silverfish.spawnAnim();
       }
 
