--- a/net/minecraft/world/level/block/PressurePlateBlock.java
+++ b/net/minecraft/world/level/block/PressurePlateBlock.java
@@ -30,20 +_,46 @@
    }
 
    protected int m_6693_(Level p_55264_, BlockPos p_55265_) {
-      Class<? extends Entity> oclass1;
+      Class<? extends Entity> oclass; // CraftBukkit
+
       switch (this.f_55250_) {
          case EVERYTHING:
-            oclass1 = Entity.class;
+            oclass = Entity.class;
             break;
          case MOBS:
-            oclass1 = LivingEntity.class;
+            oclass = LivingEntity.class;
             break;
          default:
             throw new IncompatibleClassChangeError();
       }
 
-      Class oclass = oclass1;
-      return m_289607_(p_55264_, f_49287_.m_82338_(p_55265_), oclass) > 0 ? 15 : 0;
+      Class<? extends Entity> oclass1 = oclass;
+
+      // CraftBukkit start - Call interact event when turning on a pressure plate
+      for (Entity entity : getEntities(p_55264_, PressurePlateBlock.f_49287_.m_82338_(p_55265_), oclass)) {
+         if (this.m_6016_(p_55264_.m_8055_(p_55265_)) == 0) {
+            org.bukkit.World bworld = p_55264_.getWorld();
+            org.bukkit.plugin.PluginManager manager = p_55264_.getCraftServer().getPluginManager();
+            org.bukkit.event.Cancellable cancellable;
+
+            if (entity instanceof net.minecraft.world.entity.player.Player) {
+               cancellable = org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callPlayerInteractEvent((net.minecraft.world.entity.player.Player) entity, org.bukkit.event.block.Action.PHYSICAL, p_55265_, null, null, null);
+            } else {
+               cancellable = new org.bukkit.event.entity.EntityInteractEvent(entity.getBukkitEntity(), bworld.getBlockAt(p_55265_.m_123341_(), p_55265_.m_123342_(), p_55265_.m_123343_()));
+               manager.callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+            }
+
+            // We only want to block turning the plate on if all events are cancelled
+            if (cancellable.isCancelled()) {
+               continue;
+            }
+         }
+
+         return 15;
+      }
+
+      return 0;
+      // CraftBukkit end
    }
 
    protected void m_7926_(StateDefinition.Builder<Block, BlockState> p_55262_) {
