--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -88,6 +_,18 @@
 
    public void stepOn(Level p_222132_, BlockPos p_222133_, BlockState p_222134_, Entity p_222135_) {
       if (!p_222132_.isClientSide() && canActivate(p_222134_) && p_222135_.getType() != EntityType.WARDEN) {
+         // CraftBukkit start
+         org.bukkit.event.Cancellable cancellable;
+         if (p_222135_ instanceof Player) {
+            cancellable = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerInteractEvent((Player) p_222135_, org.bukkit.event.block.Action.PHYSICAL, p_222133_, null, null, null);
+         } else {
+            cancellable = new org.bukkit.event.entity.EntityInteractEvent(p_222135_.getBukkitEntity(), p_222132_.getWorld().getBlockAt(p_222133_.getX(), p_222133_.getY(), p_222133_.getZ()));
+            p_222132_.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+         }
+         if (cancellable.isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          BlockEntity blockentity = p_222132_.getBlockEntity(p_222133_);
          if (blockentity instanceof SculkSensorBlockEntity) {
             SculkSensorBlockEntity sculksensorblockentity = (SculkSensorBlockEntity)blockentity;
@@ -177,6 +_,15 @@
    }
 
    public static void deactivate(Level p_154408_, BlockPos p_154409_, BlockState p_154410_) {
+      // CraftBukkit start
+      org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(p_154408_, p_154409_), p_154410_.getValue(SculkSensorBlock.POWER), 0);
+      p_154408_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+      if (eventRedstone.getNewCurrent() > 0) {
+         p_154408_.setBlock(p_154409_, p_154410_.setValue(SculkSensorBlock.POWER, eventRedstone.getNewCurrent()), 3);
+         return;
+      }
+      // CraftBukkit end
       p_154408_.setBlock(p_154409_, p_154410_.setValue(PHASE, SculkSensorPhase.COOLDOWN).setValue(POWER, Integer.valueOf(0)), 3);
       p_154408_.scheduleTick(p_154409_, p_154410_.getBlock(), 10);
       updateNeighbours(p_154408_, p_154409_, p_154410_);
@@ -188,6 +_,15 @@
    }
 
    public void activate(@Nullable Entity p_277529_, Level p_277340_, BlockPos p_277386_, BlockState p_277799_, int p_277993_, int p_278003_) {
+      // CraftBukkit start
+      org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(p_277340_, p_277386_), p_277799_.getValue(SculkSensorBlock.POWER), p_277993_);
+      p_277340_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+      if (eventRedstone.getNewCurrent() <= 0) {
+         return;
+      }
+      p_277993_ = eventRedstone.getNewCurrent();
+      // CraftBukkit end
       p_277340_.setBlock(p_277386_, p_277799_.setValue(PHASE, SculkSensorPhase.ACTIVE).setValue(POWER, Integer.valueOf(p_277993_)), 3);
       p_277340_.scheduleTick(p_277386_, p_277799_.getBlock(), this.getActiveTicks());
       updateNeighbours(p_277340_, p_277386_, p_277799_);
@@ -252,9 +_,21 @@
 
    public void spawnAfterBreak(BlockState p_222142_, ServerLevel p_222143_, BlockPos p_222144_, ItemStack p_222145_, boolean p_222146_) {
       super.spawnAfterBreak(p_222142_, p_222143_, p_222144_, p_222145_, p_222146_);
-      if (p_222146_) {
-         this.tryDropExperience(p_222143_, p_222144_, p_222145_, ConstantInt.of(5));
+      // CraftBukkit start - Delegate to getExpDrop
+   }
+
+   @Override
+   public int getExpDrop(BlockState iblockdata, ServerLevel worldserver, BlockPos blockposition, ItemStack itemstack, boolean flag) {
+      if (flag) {
+         return this.tryDropExperienceInt(worldserver, blockposition, itemstack, ConstantInt.of(5));
       }
 
+      return 0;
+      // CraftBukkit end
+   }
+
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader level, RandomSource randomSource, BlockPos pos, int fortuneLevel, int silkTouchLevel) {
+      return silkTouchLevel == 0 ? 5 : 0;
    }
 }
