--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -76,7 +_,7 @@
    public static final EnumProperty<Direction> FACING = DirectionalBlock.FACING;
    public static final ResourceLocation CONTENTS = new ResourceLocation("contents");
    @Nullable
-   private final DyeColor color;
+   public final DyeColor color;
 
    public MapCodec<ShulkerBoxBlock> codec() {
       return CODEC;
