--- a/net/minecraft/world/level/block/SoundType.java
+++ b/net/minecraft/world/level/block/SoundType.java
@@ -111,12 +_,13 @@
    public static final SoundType WET_SPONGE = new SoundType(1.0F, 1.0F, SoundEvents.WET_SPONGE_BREAK, SoundEvents.WET_SPONGE_STEP, SoundEvents.WET_SPONGE_PLACE, SoundEvents.WET_SPONGE_HIT, SoundEvents.WET_SPONGE_FALL);
    public final float volume;
    public final float pitch;
-   private final SoundEvent breakSound;
+   public final SoundEvent breakSound;
    private final SoundEvent stepSound;
    private final SoundEvent placeSound;
-   private final SoundEvent hitSound;
+   public final SoundEvent hitSound;
    private final SoundEvent fallSound;
 
+   @Deprecated // Forge: Use {@link net.minecraftforge.common.util.ForgeSoundType} instead for suppliers
    public SoundType(float p_56766_, float p_56767_, SoundEvent p_56768_, SoundEvent p_56769_, SoundEvent p_56770_, SoundEvent p_56771_, SoundEvent p_56772_) {
       this.volume = p_56766_;
       this.pitch = p_56767_;
