--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -38,43 +_,83 @@
    }
 
    private boolean removeWaterBreadthFirstSearch(Level p_56808_, BlockPos p_56809_) {
-      return BlockPos.breadthFirstTraversal(p_56809_, 6, 65, (p_277519_, p_277492_) -> {
+      BlockState spongeState = p_56808_.getBlockState(p_56809_);
+      org.bukkit.craftbukkit.v1_20_R1.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_20_R1.util.BlockStateListPopulator(p_56808_); // CraftBukkit - Use BlockStateListPopulator
+      BlockPos.breadthFirstTraversal(p_56809_, 6, 65, (blockposition1, consumer) -> {
          for(Direction direction : ALL_DIRECTIONS) {
-            p_277492_.accept(p_277519_.relative(direction));
+            consumer.accept(blockposition1.relative(direction));
          }
 
       }, (p_279054_) -> {
          if (p_279054_.equals(p_56809_)) {
             return true;
          } else {
-            BlockState blockstate = p_56808_.getBlockState(p_279054_);
-            FluidState fluidstate = p_56808_.getFluidState(p_279054_);
-            if (!fluidstate.is(FluidTags.WATER)) {
+            // CraftBukkit start
+            BlockState blockstate = blockList.getBlockState(p_279054_);
+            FluidState fluidstate = blockList.getFluidState(p_279054_);
+            // CraftBukkit end
+            if (!spongeState.canBeHydrated(p_56808_, p_56809_, fluidstate, p_279054_)) {
                return false;
             } else {
                Block block = blockstate.getBlock();
                if (block instanceof BucketPickup) {
                   BucketPickup bucketpickup = (BucketPickup)block;
-                  if (!bucketpickup.pickupBlock(p_56808_, p_279054_, blockstate).isEmpty()) {
+                  if (!bucketpickup.pickupBlock(blockList, p_279054_, blockstate).isEmpty()) { // CraftBukkit
                      return true;
                   }
                }
 
                if (blockstate.getBlock() instanceof LiquidBlock) {
-                  p_56808_.setBlock(p_279054_, Blocks.AIR.defaultBlockState(), 3);
+                  blockList.setBlock(p_279054_, Blocks.AIR.defaultBlockState(), 3);
                } else {
                   if (!blockstate.is(Blocks.KELP) && !blockstate.is(Blocks.KELP_PLANT) && !blockstate.is(Blocks.SEAGRASS) && !blockstate.is(Blocks.TALL_SEAGRASS)) {
                      return false;
                   }
 
-                  BlockEntity blockentity = blockstate.hasBlockEntity() ? p_56808_.getBlockEntity(p_279054_) : null;
-                  dropResources(blockstate, p_56808_, p_279054_, blockentity);
-                  p_56808_.setBlock(p_279054_, Blocks.AIR.defaultBlockState(), 3);
+                  // CraftBukkit start
+                  // BlockEntity blockentity = blockstate.hasBlockEntity() ? p_56808_.getBlockEntity(p_279054_) : null;
+                  blockList.setBlock(p_279054_, Blocks.AIR.defaultBlockState(), 3);
+                  // CraftBukkit end
                }
 
                return true;
             }
          }
-      }) > 1;
+      });
+      // CraftBukkit start
+      java.util.List<org.bukkit.craftbukkit.v1_20_R1.block.CraftBlockState> blocks = blockList.getList(); // Is a clone
+      if (!blocks.isEmpty()) {
+         final org.bukkit.block.Block bblock = p_56808_.getWorld().getBlockAt(p_56809_.getX(), p_56809_.getY(), p_56809_.getZ());
+
+         org.bukkit.event.block.SpongeAbsorbEvent event = new org.bukkit.event.block.SpongeAbsorbEvent(bblock, (java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks);
+         p_56808_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (event.isCancelled()) {
+            return false;
+         }
+
+         for (org.bukkit.craftbukkit.v1_20_R1.block.CraftBlockState block : blocks) {
+            BlockPos blockposition1 = block.getPosition();
+            BlockState iblockdata = p_56808_.getBlockState(blockposition1);
+            FluidState fluid = p_56808_.getFluidState(blockposition1);
+
+            if (fluid.is(FluidTags.WATER)) {
+               if (iblockdata.getBlock() instanceof BucketPickup && !((BucketPickup) iblockdata.getBlock()).pickupBlock(blockList, blockposition1, iblockdata).isEmpty()) {
+                  // NOP
+               } else if (iblockdata.getBlock() instanceof LiquidBlock) {
+                  // NOP
+               } else if (iblockdata.is(Blocks.KELP) || iblockdata.is(Blocks.KELP_PLANT) || iblockdata.is(Blocks.SEAGRASS) || iblockdata.is(Blocks.TALL_SEAGRASS)) {
+                  BlockEntity tileentity = iblockdata.hasBlockEntity() ? p_56808_.getBlockEntity(blockposition1) : null;
+
+                  dropResources(iblockdata, p_56808_, blockposition1, tileentity);
+               }
+            }
+            p_56808_.setBlock(blockposition1, block.getHandle(), block.getFlag());
+         }
+
+         return true;
+      }
+      return false;
+      // CraftBukkit end
    }
 }
