--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -22,7 +_,28 @@
    }
 
    protected int getSignalStrength(Level p_58213_, BlockPos p_58214_) {
-      int i = Math.min(getEntityCount(p_58213_, TOUCH_AABB.move(p_58214_), Entity.class), this.maxWeight);
+      // CraftBukkit start
+      // int i = Math.min(getEntityCount(p_58213_, TOUCH_AABB.move(p_58214_), Entity.class), this.maxWeight);
+      int i = 0;
+      for (Entity entity : getEntities(p_58213_, WeightedPressurePlateBlock.TOUCH_AABB.move(p_58214_), Entity.class)) {
+         org.bukkit.event.Cancellable cancellable;
+
+         if (entity instanceof net.minecraft.world.entity.player.Player) {
+            cancellable = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerInteractEvent((net.minecraft.world.entity.player.Player) entity, org.bukkit.event.block.Action.PHYSICAL, p_58214_, null, null, null);
+         } else {
+            cancellable = new org.bukkit.event.entity.EntityInteractEvent(entity.getBukkitEntity(), p_58213_.getWorld().getBlockAt(p_58214_.getX(), p_58214_.getY(), p_58214_.getZ()));
+            p_58213_.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+         }
+
+         // We only want to block turning the plate on if all events are cancelled
+         if (!cancellable.isCancelled()) {
+            i++;
+         }
+      }
+
+      i = Math.min(i, this.maxWeight);
+      // CraftBukkit end
+
       if (i > 0) {
          float f = (float)Math.min(this.maxWeight, i) / (float)this.maxWeight;
          return Mth.ceil(f * 15.0F);
