--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -76,12 +_,16 @@
 
       protected void m_213704_(String p_227582_, BlockPos p_227583_, ServerLevelAccessor p_227584_, RandomSource p_227585_, BoundingBox p_227586_) {
          if ("chest".equals(p_227582_)) {
-            p_227584_.m_7731_(p_227583_, Blocks.f_50016_.m_49966_(), 3);
-            BlockEntity blockentity = p_227584_.m_7702_(p_227583_.m_7495_());
+            // CraftBukkit start
+            /*
+            p_227584_.setBlock(p_227583_, Blocks.AIR.defaultBlockState(), 3);
+            BlockEntity blockentity = p_227584_.getBlockEntity(p_227583_.below());
             if (blockentity instanceof ChestBlockEntity) {
-               ((ChestBlockEntity)blockentity).m_59626_(BuiltInLootTables.f_78688_, p_227585_.m_188505_());
+               ((ChestBlockEntity)blockentity).setLootTable(BuiltInLootTables.IGLOO_CHEST, p_227585_.nextLong());
             }
-
+            */
+            setCraftLootTable(p_227584_, p_227583_.m_7495_(), p_227585_, BuiltInLootTables.f_78688_);
+            // CraftBukkit end
          }
       }
 
