--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -270,7 +_,7 @@
                if (this.spiderCorridor && !this.hasPlacedSpider) {
                   int l1 = 1;
                   int i2 = k1 - 1 + p_227746_.nextInt(3);
-                  BlockPos blockpos = this.getWorldPos(1, 0, i2);
+                  BlockPos.MutableBlockPos blockpos = this.getWorldPos(1, 0, i2);
                   if (p_227747_.isInside(blockpos) && this.isInterior(p_227743_, 1, 0, i2, p_227747_)) {
                      this.hasPlacedSpider = true;
                      p_227743_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
