--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -871,15 +_,20 @@
          this.m_73434_(p_228362_, blockstate, 3, 8, 8, p_228366_);
          this.m_73434_(p_228362_, Blocks.f_50198_.m_49966_().m_61124_(FenceBlock.f_52312_, Boolean.valueOf(true)), 4, 8, 8, p_228366_);
          if (!this.f_228354_) {
-            BlockPos blockpos = this.m_163582_(3, 5, 5);
+            BlockPos.MutableBlockPos blockpos = this.m_163582_(3, 5, 5);
             if (p_228366_.m_71051_(blockpos)) {
                this.f_228354_ = true;
-               p_228362_.m_7731_(blockpos, Blocks.f_50085_.m_49966_(), 2);
-               BlockEntity blockentity = p_228362_.m_7702_(blockpos);
+               // CraftBukkit start
+               /*
+               p_228362_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
+               BlockEntity blockentity = p_228362_.getBlockEntity(blockpos);
                if (blockentity instanceof SpawnerBlockEntity) {
                   SpawnerBlockEntity spawnerblockentity = (SpawnerBlockEntity)blockentity;
-                  spawnerblockentity.m_252803_(EntityType.f_20551_, p_228365_);
+                  spawnerblockentity.setEntityId(EntityType.BLAZE, p_228365_);
                }
+               */
+               placeCraftSpawner(p_228362_, blockpos, org.bukkit.entity.EntityType.BLAZE, 2);
+               // CraftBukkit end
             }
          }
 
