--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -871,7 +_,7 @@
          this.placeBlock(p_228362_, blockstate, 3, 8, 8, p_228366_);
          this.placeBlock(p_228362_, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, Boolean.valueOf(true)), 4, 8, 8, p_228366_);
          if (!this.hasPlacedSpawner) {
-            BlockPos blockpos = this.getWorldPos(3, 5, 5);
+            BlockPos.MutableBlockPos blockpos = this.getWorldPos(3, 5, 5);
             if (p_228366_.isInside(blockpos)) {
                this.hasPlacedSpawner = true;
                p_228362_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
