--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -703,7 +_,7 @@
          }
 
          if (!this.hasPlacedSpawner) {
-            BlockPos blockpos = this.getWorldPos(5, 3, 6);
+            BlockPos.MutableBlockPos blockpos = this.getWorldPos(5, 3, 6);
             if (p_229639_.isInside(blockpos)) {
                this.hasPlacedSpawner = true;
                p_229635_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
