--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -14,11 +_,14 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory;
 
 public abstract class AbstractHurtingProjectile extends Projectile {
    public double xPower;
    public double yPower;
    public double zPower;
+   public float bukkitYield = 1; // CraftBukkit
+   public boolean isIncendiary = true; // CraftBukkit
 
    protected AbstractHurtingProjectile(EntityType<? extends AbstractHurtingProjectile> p_36833_, Level p_36834_) {
       super(p_36833_, p_36834_);
@@ -65,8 +_,14 @@
          }
 
          HitResult hitresult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
-         if (hitresult.getType() != HitResult.Type.MISS) {
-            this.onHit(hitresult);
+         if (hitresult.getType() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
+            this.preOnHit(hitresult); // CraftBukkit - projectile hit event
+
+            // CraftBukkit start - Fire ProjectileHitEvent
+            if (this.isRemoved()) {
+               CraftEventFactory.callProjectileHitEvent(this, hitresult);
+            }
+            // CraftBukkit end
          }
 
          this.checkInsideBlocks();
@@ -143,6 +_,11 @@
          Entity entity = p_36839_.getEntity();
          if (entity != null) {
             if (!this.level().isClientSide) {
+               // CraftBukkit start
+               if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_36839_, p_36840_, false)) {
+                  return false;
+               }
+               // CraftBukkit end
                Vec3 vec3 = entity.getLookAngle();
                this.setDeltaMovement(vec3);
                this.xPower = vec3.x * 0.1D;
