--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -42,6 +_,11 @@
    public Arrow(Level p_36861_, LivingEntity p_310439_, ItemStack p_310691_) {
       super(EntityType.ARROW, p_310439_, p_36861_, p_310691_);
    }
+   //Ketting start
+   public Arrow(Level p_36861_, LivingEntity p_310439_) {
+      super(EntityType.ARROW, p_310439_, p_36861_, DEFAULT_ARROW_STACK);
+   }
+   //Ketting end
 
    public void setEffectsFromItem(ItemStack p_36879_) {
       if (p_36879_.is(Items.TIPPED_ARROW)) {
@@ -179,15 +_,17 @@
       Entity entity = this.getEffectSource();
 
       for(MobEffectInstance mobeffectinstance : this.potion.getEffects()) {
+         p_36873_.prepareAddEffect(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); //Ketting
          p_36873_.addEffect(new MobEffectInstance(mobeffectinstance.getEffect(), Math.max(mobeffectinstance.mapDuration((p_268168_) -> {
             return p_268168_ / 8;
          }), 1), mobeffectinstance.getAmplifier(), mobeffectinstance.isAmbient(), mobeffectinstance.isVisible()), entity);
-      }
 
       if (!this.effects.isEmpty()) {
          for(MobEffectInstance mobeffectinstance1 : this.effects) {
+            p_36873_.prepareAddEffect(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); //Ketting
             p_36873_.addEffect(mobeffectinstance1, entity);
          }
+      }
       }
 
    }
