--- a/net/minecraft/world/entity/projectile/Fireball.java
+++ b/net/minecraft/world/entity/projectile/Fireball.java
@@ -57,6 +_,6 @@
    public void readAdditionalSaveData(CompoundTag p_37009_) {
       super.readAdditionalSaveData(p_37009_);
       ItemStack itemstack = ItemStack.of(p_37009_.getCompound("Item"));
-      this.setItem(itemstack);
+      if (!itemstack.isEmpty()) this.setItem(itemstack); // CraftBukkit - SPIGOT-5474 probably came from bugged earlier versions
    }
 }
