--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -12,7 +_,7 @@
 
 public class SpectralArrow extends AbstractArrow {
    private static final ItemStack DEFAULT_ARROW_STACK = new ItemStack(Items.SPECTRAL_ARROW);
-   private int duration = 200;
+   public int duration = 200;
 
    public SpectralArrow(EntityType<? extends SpectralArrow> p_37411_, Level p_37412_) {
       super(p_37411_, p_37412_, DEFAULT_ARROW_STACK);
@@ -22,6 +_,12 @@
       super(EntityType.SPECTRAL_ARROW, p_311904_, p_37414_, p_309799_);
    }
 
+   //Ketting start
+   public SpectralArrow(Level p_37414_, LivingEntity p_311904_) {
+      super(EntityType.SPECTRAL_ARROW, p_311904_, p_37414_, DEFAULT_ARROW_STACK);
+   }
+   //Ketting end
+
    public SpectralArrow(Level p_37419_, double p_309976_, double p_310894_, double p_309922_, ItemStack p_311719_) {
       super(EntityType.SPECTRAL_ARROW, p_309976_, p_310894_, p_309922_, p_37419_, p_311719_);
    }
@@ -37,7 +_,7 @@
    protected void doPostHurtEffects(LivingEntity p_37422_) {
       super.doPostHurtEffects(p_37422_);
       MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
-      p_37422_.addEffect(mobeffectinstance, this.getEffectSource());
+      p_37422_.addEffect(mobeffectinstance, this.getEffectSource(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
    }
 
    public void readAdditionalSaveData(CompoundTag p_37424_) {
