--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -22,6 +_,12 @@
       super(EntityType.SPECTRAL_ARROW, p_311904_, p_37414_, p_309799_);
    }
 
+   //Ketting start
+   public SpectralArrow(Level p_37414_, LivingEntity p_311904_) {
+      super(EntityType.SPECTRAL_ARROW, p_311904_, p_37414_, DEFAULT_ARROW_STACK);
+   }
+   //Ketting end
+
    public SpectralArrow(Level p_37419_, double p_309976_, double p_310894_, double p_309922_, ItemStack p_311719_) {
       super(EntityType.SPECTRAL_ARROW, p_309976_, p_310894_, p_309922_, p_37419_, p_311719_);
    }
@@ -37,6 +_,7 @@
    protected void doPostHurtEffects(LivingEntity p_37422_) {
       super.doPostHurtEffects(p_37422_);
       MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
+      p_37422_.prepareAddEffect(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); //Ketting
       p_37422_.addEffect(mobeffectinstance, this.getEffectSource());
    }
 
