--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -46,22 +_,40 @@
          if (entity instanceof ServerPlayer) {
             ServerPlayer serverplayer = (ServerPlayer)entity;
             if (serverplayer.connection.isAcceptingMessages() && serverplayer.level() == this.level() && !serverplayer.isSleeping()) {
-               if (this.random.nextFloat() < 0.05F && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
-                  Endermite endermite = EntityType.ENDERMITE.create(this.level());
-                  if (endermite != null) {
-                     endermite.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
-                     this.level().addFreshEntity(endermite);
+               net.minecraftforge.event.entity.EntityTeleportEvent.EnderPearl event = net.minecraftforge.event.ForgeEventFactory.onEnderPearlLand(serverplayer, this.getX(), this.getY(), this.getZ(), this, 5.0F, p_37504_);
+               if (!event.isCanceled()) { // Don't indent to lower patch size
+                  // CraftBukkit start - Fire PlayerTeleportEvent
+                  org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer player = serverplayer.getBukkitEntity();
+                  org.bukkit.Location location = getBukkitEntity().getLocation();
+                  location.setPitch(player.getLocation().getPitch());
+                  location.setYaw(player.getLocation().getYaw());
+
+                  org.bukkit.event.player.PlayerTeleportEvent teleEvent = new org.bukkit.event.player.PlayerTeleportEvent(player, player.getLocation(), location, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL);
+                  org.bukkit.Bukkit.getPluginManager().callEvent(teleEvent);
+
+                  if (!teleEvent.isCancelled() && serverplayer.connection.isAcceptingMessages()) {
+                     if (this.random.nextFloat() < 0.05F && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+                        Endermite entityendermite = (Endermite) EntityType.ENDERMITE.create(this.level());
+
+                        if (entityendermite != null) {
+                           entityendermite.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
+                           this.level().prepareAddFreshEntity(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.ENDER_PEARL);
+                           this.level().addFreshEntity(entityendermite);
+                        }
+                     }
+
+                     if (entity.isPassenger()) {
+                        entity.stopRiding();
+                     }
+
+                     serverplayer.connection.teleport(teleEvent.getTo());
+                     entity.resetFallDistance();
+                     org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.entityDamage = this;
+                     entity.hurt(this.damageSources().fall(), 5.0F);
+                     org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.entityDamage = null;
                   }
-               }
-
-               if (entity.isPassenger()) {
-                  serverplayer.dismountTo(this.getX(), this.getY(), this.getZ());
-               } else {
-                  entity.teleportTo(this.getX(), this.getY(), this.getZ());
-               }
-
-               entity.resetFallDistance();
-               entity.hurt(this.damageSources().fall(), 5.0F);
+                  // CraftBukkit end
+               } //Forge: End
             }
          } else if (entity != null) {
             entity.teleportTo(this.getX(), this.getY(), this.getZ());
@@ -84,12 +_,12 @@
    }
 
    @Nullable
-   public Entity changeDimension(ServerLevel p_37506_) {
+   public Entity changeDimension(ServerLevel p_37506_, net.minecraftforge.common.util.ITeleporter teleporter) {
       Entity entity = this.getOwner();
-      if (entity != null && entity.level().dimension() != p_37506_.dimension()) {
+      if (entity != null && p_37506_ != null && entity.level().dimension() != p_37506_.dimension()) { // CraftBukkit - SPIGOT-6113
          this.setOwner((Entity)null);
       }
 
-      return super.changeDimension(p_37506_);
+      return super.changeDimension(p_37506_, teleporter);
    }
 }
