--- a/net/minecraft/world/entity/projectile/ThrownPotion.java
+++ b/net/minecraft/world/entity/projectile/ThrownPotion.java
@@ -88,7 +_,7 @@
          boolean flag = potion == Potions.WATER && list.isEmpty();
          if (flag) {
             this.applyWater();
-         } else if (!list.isEmpty()) {
+         } else if (true || !list.isEmpty()) { // CraftBukkit - Call event even if no effects to apply
             if (this.isLingering()) {
                this.makeAreaOfEffectCloud(itemstack, potion);
             } else {
@@ -127,39 +_,66 @@
    private void applySplash(List<MobEffectInstance> p_37548_, @Nullable Entity p_37549_) {
       AABB aabb = this.getBoundingBox().inflate(4.0D, 2.0D, 4.0D);
       List<LivingEntity> list = this.level().getEntitiesOfClass(LivingEntity.class, aabb);
+      java.util.Map<org.bukkit.entity.LivingEntity, Double> affected = new java.util.HashMap<org.bukkit.entity.LivingEntity, Double>(); // CraftBukkit
       if (!list.isEmpty()) {
          Entity entity = this.getEffectSource();
 
-         for(LivingEntity livingentity : list) {
+         for (LivingEntity livingentity : list) {
             if (livingentity.isAffectedByPotions()) {
                double d0 = this.distanceToSqr(livingentity);
                if (d0 < 16.0D) {
-                  double d1;
+                  double d1 = 1.0D - Math.sqrt(d0) / 4.0D;
                   if (livingentity == p_37549_) {
                      d1 = 1.0D;
-                  } else {
-                     d1 = 1.0D - Math.sqrt(d0) / 4.0D;
-                  }
-
-                  for(MobEffectInstance mobeffectinstance : p_37548_) {
-                     MobEffect mobeffect = mobeffectinstance.getEffect();
-                     if (mobeffect.isInstantenous()) {
-                        mobeffect.applyInstantenousEffect(this, this.getOwner(), livingentity, mobeffectinstance.getAmplifier(), d1);
-                     } else {
-                        int i = mobeffectinstance.mapDuration((p_267930_) -> {
-                           return (int)(d1 * (double)p_267930_ + 0.5D);
-                        });
-                        MobEffectInstance mobeffectinstance1 = new MobEffectInstance(mobeffect, i, mobeffectinstance.getAmplifier(), mobeffectinstance.isAmbient(), mobeffectinstance.isVisible());
-                        if (!mobeffectinstance1.endsWithin(20)) {
-                           livingentity.addEffect(mobeffectinstance1, entity);
-                        }
-                     }
-                  }
-               }
-            }
-         }
-      }
-
+                  }
+
+                  // CraftBukkit start
+                  affected.put((org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity(), d1);
+               }
+            }
+         }
+      }
+
+
+      org.bukkit.event.entity.PotionSplashEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPotionSplashEvent(this, affected);
+      if (!event.isCancelled() && p_37548_ != null && !p_37548_.isEmpty()) { // do not process effects if there are no effects to process
+         Entity entity1 = this.getEffectSource();
+         for (org.bukkit.entity.LivingEntity victim : event.getAffectedEntities()) {
+            if (!(victim instanceof org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity)) {
+               continue;
+            }
+
+            LivingEntity livingentity = ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) victim).getHandle();
+            double intensity = event.getIntensity(victim);
+            // CraftBukkit end
+
+            for (MobEffectInstance effectinstance : p_37548_) {
+               MobEffect mobeffectlist = effectinstance.getEffect();
+               // CraftBukkit start - Abide by PVP settings - for players only!
+               if (!this.level().pvpMode && this.getOwner() instanceof net.minecraft.server.level.ServerPlayer && livingentity instanceof net.minecraft.server.level.ServerPlayer && livingentity != this.getOwner()) {
+                  if (mobeffectlist == net.minecraft.world.effect.MobEffects.MOVEMENT_SLOWDOWN || mobeffectlist == net.minecraft.world.effect.MobEffects.DIG_SLOWDOWN || mobeffectlist == net.minecraft.world.effect.MobEffects.HARM || mobeffectlist == net.minecraft.world.effect.MobEffects.BLINDNESS
+                          || mobeffectlist == net.minecraft.world.effect.MobEffects.HUNGER || mobeffectlist == net.minecraft.world.effect.MobEffects.WEAKNESS || mobeffectlist == net.minecraft.world.effect.MobEffects.POISON) {
+                     continue;
+                  }
+               }
+               // CraftBukkit end
+
+               if (mobeffectlist.isInstantenous()) {
+                  mobeffectlist.applyInstantenousEffect(this, this.getOwner(), livingentity, effectinstance.getAmplifier(), intensity);
+               } else {
+                  int i = effectinstance.mapDuration((j) -> {
+                     return (int) (intensity * (double) j + 0.5D);
+                  });
+                  MobEffectInstance mobeffect1 = new MobEffectInstance(mobeffectlist, i, effectinstance.getAmplifier(), effectinstance.isAmbient(), effectinstance.isVisible());
+
+                  if (!mobeffect1.endsWithin(20)) {
+                     livingentity.prepareAddEffect(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_SPLASH); //Ketting
+                     livingentity.addEffect(mobeffect1, entity1);
+                  }
+               }
+            }
+         }
+      }
    }
 
    private void makeAreaOfEffectCloud(ItemStack p_37538_, Potion p_37539_) {
@@ -184,7 +_,14 @@
          areaeffectcloud.setFixedColor(compoundtag.getInt("CustomPotionColor"));
       }
 
-      this.level().addFreshEntity(areaeffectcloud);
+      // CraftBukkit start
+      org.bukkit.event.entity.LingeringPotionSplashEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callLingeringPotionSplashEvent(this, areaeffectcloud);
+      if (!(event.isCancelled() || areaeffectcloud.isRemoved())) {
+         this.level().addFreshEntity(areaeffectcloud);
+      } else {
+         areaeffectcloud.discard();
+      }
+      // CraftBukkit end
    }
 
    private boolean isLingering() {
@@ -194,13 +_,25 @@
    private void dowseFire(BlockPos p_150193_) {
       BlockState blockstate = this.level().getBlockState(p_150193_);
       if (blockstate.is(BlockTags.FIRE)) {
-         this.level().removeBlock(p_150193_, false);
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, p_150193_, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState())) {
+            this.level().removeBlock(p_150193_, false);
+         }
+         // CraftBukkit end
       } else if (AbstractCandleBlock.isLit(blockstate)) {
-         AbstractCandleBlock.extinguish((Player)null, blockstate, this.level(), p_150193_);
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, p_150193_, blockstate.setValue(AbstractCandleBlock.LIT, false))) {
+            AbstractCandleBlock.extinguish((Player) null, blockstate, this.level(), p_150193_);
+         }
+         // CraftBukkit end
       } else if (CampfireBlock.isLitCampfire(blockstate)) {
-         this.level().levelEvent((Player)null, 1009, p_150193_, 0);
-         CampfireBlock.dowse(this.getOwner(), this.level(), p_150193_, blockstate);
-         this.level().setBlockAndUpdate(p_150193_, blockstate.setValue(CampfireBlock.LIT, Boolean.valueOf(false)));
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, p_150193_, blockstate.setValue(CampfireBlock.LIT, false))) {
+            this.level().levelEvent((Player) null, 1009, p_150193_, 0);
+            CampfireBlock.dowse(this.getOwner(), this.level(), p_150193_, blockstate);
+            this.level().setBlockAndUpdate(p_150193_, (BlockState) blockstate.setValue(CampfireBlock.LIT, false));
+         }
+         // CraftBukkit end
       }
 
    }
