--- a/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -40,7 +_,7 @@
    }
 
    public float getBlockExplosionResistance(Explosion p_37619_, BlockGetter p_37620_, BlockPos p_37621_, BlockState p_37622_, FluidState p_37623_, float p_37624_) {
-      return this.isDangerous() && WitherBoss.canDestroy(p_37622_) ? Math.min(0.8F, p_37624_) : p_37624_;
+      return this.isDangerous() && p_37622_.canEntityDestroy(p_37620_, p_37621_, this) ? Math.min(0.8F, p_37624_) : p_37624_;
    }
 
    protected void onHitEntity(EntityHitResult p_37626_) {
@@ -56,7 +_,7 @@
                if (entity.isAlive()) {
                   this.doEnchantDamageEffects(livingentity, entity);
                } else {
-                  livingentity.heal(5.0F);
+                  livingentity.heal(5.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.WITHER); // CraftBukkit
                }
             }
          } else {
@@ -73,7 +_,8 @@
             }
 
             if (i > 0) {
-               livingentity1.addEffect(new MobEffectInstance(MobEffects.WITHER, 20 * i, 1), this.getEffectSource());
+               livingentity1.prepareAddEffect(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Ketting
+               livingentity1.addEffect(new MobEffectInstance(MobEffects.WITHER, 20 * i, 1), this.getEffectSource());;
             }
          }
 
@@ -83,7 +_,15 @@
    protected void onHit(HitResult p_37628_) {
       super.onHit(p_37628_);
       if (!this.level().isClientSide) {
-         this.level().explode(this, this.getX(), this.getY(), this.getZ(), 1.0F, false, Level.ExplosionInteraction.MOB);
+         // CraftBukkit start
+         // this.level().explode(this, this.getX(), this.getY(), this.getZ(), 1.0F, false, World.a.MOB);
+         org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), 1.0F, false);
+         this.level().getCraftServer().getPluginManager().callEvent(event);
+
+         if (!event.isCancelled()) {
+            this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB);
+         }
+         // CraftBukkit end
          this.discard();
       }
 
