--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -39,7 +_,22 @@
 
    static void m_230763_(Level p_230764_, BlockState p_230765_, BlockPos p_230766_, Block p_230767_, BlockPos p_230768_, boolean p_230769_) {
       try {
+         // CraftBukkit start
+         org.bukkit.craftbukkit.v1_20_R1.CraftWorld cworld = p_230764_.getWorld();
+         if (cworld != null) {
+            org.bukkit.event.block.BlockPhysicsEvent event = new org.bukkit.event.block.BlockPhysicsEvent(org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(p_230764_, p_230766_), org.bukkit.craftbukkit.v1_20_R1.block.data.CraftBlockData.fromData(p_230765_), org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock.at(p_230764_, p_230768_));
+            p_230764_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               return;
+            }
+         }
+         // CraftBukkit end
          p_230765_.m_60690_(p_230764_, p_230766_, p_230767_, p_230768_, p_230769_);
+         // Spigot Start
+      } catch (StackOverflowError ex) {
+         p_230764_.lastPhysicsProblem = new BlockPos(p_230766_);
+         // Spigot End
       } catch (Throwable throwable) {
          CrashReport crashreport = CrashReport.m_127521_(throwable, "Exception while updating neighbours");
          CrashReportCategory crashreportcategory = crashreport.m_127514_("Block being updated");
