--- a/net/minecraft/world/level/levelgen/structure/placement/StructurePlacement.java
+++ b/net/minecraft/world/level/levelgen/structure/placement/StructurePlacement.java
@@ -21,11 +_,11 @@
 public abstract class StructurePlacement {
    public static final Codec<StructurePlacement> f_205036_ = BuiltInRegistries.f_256950_.m_194605_().dispatch(StructurePlacement::m_203443_, StructurePlacementType::m_205049_);
    private static final int f_227021_ = 10387320;
-   private final Vec3i f_227022_;
-   private final StructurePlacement.FrequencyReductionMethod f_227023_;
-   private final float f_227024_;
-   private final int f_227025_;
-   private final Optional<StructurePlacement.ExclusionZone> f_227026_;
+   public final Vec3i f_227022_;
+   public final StructurePlacement.FrequencyReductionMethod f_227023_;
+   public final float f_227024_;
+   public final int f_227025_;
+   public final Optional<StructurePlacement.ExclusionZone> f_227026_;
 
    protected static <S extends StructurePlacement> Products.P5<RecordCodecBuilder.Mu<S>, Vec3i, StructurePlacement.FrequencyReductionMethod, Float, Integer, Optional<StructurePlacement.ExclusionZone>> m_227041_(RecordCodecBuilder.Instance<S> p_227042_) {
       return p_227042_.group(Vec3i.m_194650_(16).optionalFieldOf("locate_offset", Vec3i.f_123288_).forGetter(StructurePlacement::m_227072_), StructurePlacement.FrequencyReductionMethod.f_227108_.optionalFieldOf("frequency_reduction_method", StructurePlacement.FrequencyReductionMethod.DEFAULT).forGetter(StructurePlacement::m_227073_), Codec.floatRange(0.0F, 1.0F).optionalFieldOf("frequency", 1.0F).forGetter(StructurePlacement::m_227074_), ExtraCodecs.f_144628_.fieldOf("salt").forGetter(StructurePlacement::m_227075_), StructurePlacement.ExclusionZone.f_227077_.optionalFieldOf("exclusion_zone").forGetter(StructurePlacement::m_227076_));
