--- a/net/minecraft/server/level/ServerBossEvent.java
+++ b/net/minecraft/server/level/ServerBossEvent.java
@@ -15,7 +_,7 @@
 public class ServerBossEvent extends BossEvent {
    private final Set<ServerPlayer> players = Sets.newHashSet();
    private final Set<ServerPlayer> unmodifiablePlayers = Collections.unmodifiableSet(this.players);
-   private boolean visible = true;
+   public boolean visible = true;
 
    public ServerBossEvent(Component p_8300_, BossEvent.BossBarColor p_8301_, BossEvent.BossBarOverlay p_8302_) {
       super(Mth.createInsecureUUID(), p_8300_, p_8301_, p_8302_);
@@ -80,7 +_,7 @@
 
    }
 
-   private void broadcast(Function<BossEvent, ClientboundBossEventPacket> p_143225_) {
+   public void broadcast(Function<BossEvent, ClientboundBossEventPacket> p_143225_) {
       if (this.visible) {
          ClientboundBossEventPacket clientboundbosseventpacket = p_143225_.apply(this);
 
