--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -9,7 +_,7 @@
 public class TicketType<T> {
    private final String name;
    private final Comparator<T> comparator;
-   private final long timeout;
+   public long timeout;
    public static final TicketType<Unit> START = create("start", (p_9471_, p_9472_) -> {
       return 0;
    });
@@ -22,6 +_,8 @@
    public static final TicketType<BlockPos> PORTAL = create("portal", Vec3i::compareTo, 300);
    public static final TicketType<Integer> POST_TELEPORT = create("post_teleport", Integer::compareTo, 5);
    public static final TicketType<ChunkPos> UNKNOWN = create("unknown", Comparator.comparingLong(ChunkPos::toLong), 1);
+   public static final TicketType<Unit> PLUGIN = create("plugin", (a, b) -> 0); // CraftBukkit
+   public static final TicketType<org.bukkit.plugin.Plugin> PLUGIN_TICKET = create("plugin_ticket", (plugin1, plugin2) -> plugin1.getClass().getName().compareTo(plugin2.getClass().getName())); // CraftBukkit
 
    public static <T> TicketType<T> create(String p_9463_, Comparator<T> p_9464_) {
       return new TicketType<>(p_9463_, p_9464_, 0L);
