--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -190,7 +_,7 @@
       if (blockstate.isAir()) {
          return false;
       } else {
-         if (p_9551_) {
+         if (false) { // CraftBukkit - SPIGOT-6833: Do not drop during world generation
             BlockEntity blockentity = blockstate.hasBlockEntity() ? this.getBlockEntity(p_9550_) : null;
             Block.dropResources(blockstate, this.level, p_9550_, blockentity, p_9552_, ItemStack.EMPTY);
          }
@@ -296,7 +_,38 @@
       this.getChunk(p_9592_).markPosForPostprocessing(p_9592_);
    }
 
+   //Ketting start
+   private org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason addFreshEntity$spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT;
+   @Override
+   public void prepareAddFreshEntity(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      this.addFreshEntity$spawnReason = reason;
+   }
+
+   private java.util.List<Entity> disabledEntitySpawns = new java.util.ArrayList<>();
+   @Override
+   public void disableSpawnFor(Entity entity) {
+      disabledEntitySpawns.add(entity);
+   }
+
    public boolean addFreshEntity(Entity p_9580_) {
+      // CraftBukkit start
+      try {
+         if (disabledEntitySpawns.contains(p_9580_)) {
+            disabledEntitySpawns.remove(p_9580_);
+            System.out.println("Tried to spawn a disabled entity: " + p_9580_.getClass().getName());
+            return false;
+         }
+         return addFreshEntity(p_9580_, addFreshEntity$spawnReason);
+      } finally {
+         addFreshEntity$spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT;
+      }
+   }
+   //Ketting end
+
+   public boolean addFreshEntity(Entity p_9580_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      // CraftBukkit end
+      if (p_9580_ instanceof net.minecraft.world.entity.Mob mob && mob.isSpawnCancelled()) return false;
+
       int i = SectionPos.blockToSectionCoord(p_9580_.getBlockX());
       int j = SectionPos.blockToSectionCoord(p_9580_.getBlockZ());
       this.getChunk(i, j).addEntity(p_9580_);
