--- a/net/minecraft/world/item/enchantment/Enchantments.java
+++ b/net/minecraft/world/item/enchantment/Enchantments.java
@@ -47,6 +_,10 @@
    public static final Enchantment VANISHING_CURSE = register("vanishing_curse", new VanishingCurseEnchantment(Enchantment.Rarity.VERY_RARE, EquipmentSlot.values()));
 
    private static Enchantment register(String p_44993_, Enchantment p_44994_) {
-      return Registry.register(BuiltInRegistries.ENCHANTMENT, p_44993_, p_44994_);
+      // CraftBukkit start
+      p_44994_ = (Enchantment) Registry.register(BuiltInRegistries.ENCHANTMENT, p_44993_, p_44994_);
+      org.bukkit.enchantments.Enchantment.registerEnchantment(new org.bukkit.craftbukkit.v1_20_R1.enchantments.CraftEnchantment(p_44994_));
+      return p_44994_;
+      // CraftBukkit end
    }
 }
