--- a/net/minecraft/world/item/crafting/BlastingRecipe.java
+++ b/net/minecraft/world/item/crafting/BlastingRecipe.java
@@ -15,4 +_,17 @@
    public RecipeSerializer<?> getSerializer() {
       return RecipeSerializer.BLASTING_RECIPE;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+      org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.result);
+
+      org.bukkit.craftbukkit.inventory.CraftBlastingRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftBlastingRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toBukkit(this.ingredient), this.experience, this.cookingTime);
+      recipe.setGroup(this.group);
+      recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
