--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -3,6 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
 
 public class CampfireCookingRecipe extends AbstractCookingRecipe {
    public CampfireCookingRecipe(String p_250140_, CookingBookCategory p_251808_, Ingredient p_249826_, ItemStack p_251839_, float p_251432_, int p_251471_) {
       super(RecipeType.CAMPFIRE_COOKING, p_250140_, p_251808_, p_249826_, p_251839_, p_251432_, p_251471_);
@@ -15,4 +_,17 @@
    public RecipeSerializer<?> getSerializer() {
       return RecipeSerializer.CAMPFIRE_COOKING_RECIPE;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+      org.bukkit.craftbukkit.v1_20_R2.inventory.CraftItemStack result = org.bukkit.craftbukkit.v1_20_R2.inventory.CraftItemStack.asCraftMirror(this.result);
+
+      org.bukkit.craftbukkit.v1_20_R2.inventory.CraftCampfireRecipe recipe = new org.bukkit.craftbukkit.v1_20_R2.inventory.CraftCampfireRecipe(id, result, org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe.toBukkit(this.ingredient), this.experience, this.cookingTime);
+      recipe.setGroup(this.group);
+      recipe.setCategory(org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe.getCategory(this.category()));
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
