--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -13,6 +_,13 @@
       this.f_244184_ = p_249010_;
    }
 
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      return new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftComplexRecipe(this);
+   }
+   // CraftBukkit end
+
    public ResourceLocation m_6423_() {
       return this.f_43831_;
    }
