--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -10,6 +_,13 @@
       this.f_244184_ = p_249010_;
    }
 
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+      return new org.bukkit.craftbukkit.v1_20_R2.inventory.CraftComplexRecipe(id, this);
+   }
+   // CraftBukkit end
+
    public boolean m_5598_() {
       return true;
    }
