--- a/net/minecraft/world/item/crafting/RecipeSerializer.java
+++ b/net/minecraft/world/item/crafting/RecipeSerializer.java
@@ -6,7 +_,7 @@
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.resources.ResourceLocation;
 
-public interface RecipeSerializer<T extends Recipe<?>> {
+public interface RecipeSerializer<T extends Recipe<?>> extends net.minecraftforge.common.extensions.IForgeRecipeSerializer<T> {
    RecipeSerializer<ShapedRecipe> SHAPED_RECIPE = register("crafting_shaped", new ShapedRecipe.Serializer());
    RecipeSerializer<ShapelessRecipe> SHAPELESS_RECIPE = register("crafting_shapeless", new ShapelessRecipe.Serializer());
    RecipeSerializer<ArmorDyeRecipe> ARMOR_DYE = register("crafting_special_armordye", new SimpleCraftingRecipeSerializer<>(ArmorDyeRecipe::new));
@@ -31,8 +_,10 @@
    RecipeSerializer<SmithingTrimRecipe> SMITHING_TRIM = register("smithing_trim", new SmithingTrimRecipe.Serializer());
    RecipeSerializer<DecoratedPotRecipe> DECORATED_POT_RECIPE = register("crafting_decorated_pot", new SimpleCraftingRecipeSerializer<>(DecoratedPotRecipe::new));
 
+   // Forge: use fromJson with IContext if you need the context
    T fromJson(ResourceLocation p_44103_, JsonObject p_44104_);
 
+   @org.jetbrains.annotations.Nullable
    T fromNetwork(ResourceLocation p_44105_, FriendlyByteBuf p_44106_);
 
    void toNetwork(FriendlyByteBuf p_44101_, T p_44102_);
