--- a/net/minecraft/world/item/crafting/SmeltingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmeltingRecipe.java
@@ -3,6 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
 
 public class SmeltingRecipe extends AbstractCookingRecipe {
    public SmeltingRecipe(String p_250200_, CookingBookCategory p_251114_, Ingredient p_250340_, ItemStack p_250306_, float p_249577_, int p_250030_) {
       super(RecipeType.SMELTING, p_250200_, p_251114_, p_250340_, p_250306_, p_249577_, p_250030_);
@@ -15,4 +_,17 @@
    public RecipeSerializer<?> getSerializer() {
       return RecipeSerializer.SMELTING_RECIPE;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+      org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.result);
+
+      org.bukkit.craftbukkit.inventory.CraftFurnaceRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftFurnaceRecipe(id, result, org.bukkit.craftbukkit.inventory.CraftRecipe.toBukkit(this.ingredient), this.experience, this.cookingTime);
+      recipe.setGroup(this.group);
+      recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
