--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -94,8 +_,15 @@
    }
 
    public boolean m_142505_() {
-      return Stream.of(this.f_265958_, this.f_266040_, this.f_266053_).anyMatch(Ingredient::m_43947_);
-   }
+      return Stream.of(this.f_265958_, this.f_266040_, this.f_266053_).anyMatch(net.minecraftforge.common.ForgeHooks::hasNoElements);
+   }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+      return new org.bukkit.craftbukkit.v1_20_R2.inventory.CraftSmithingTrimRecipe(id, org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe.toBukkit(this.f_265958_), org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe.toBukkit(this.f_266040_), org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe.toBukkit(this.f_266053_));
+   }
+   // CraftBukkit end
 
    public static class Serializer implements RecipeSerializer<SmithingTrimRecipe> {
       public SmithingTrimRecipe m_6729_(ResourceLocation p_267037_, JsonObject p_267004_) {
