--- a/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -17,4 +_,16 @@
    public ItemStack getToastSymbol() {
       return new ItemStack(Blocks.STONECUTTER);
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+      org.bukkit.craftbukkit.v1_20_R3.inventory.CraftItemStack result = org.bukkit.craftbukkit.v1_20_R3.inventory.CraftItemStack.asCraftMirror(this.result);
+
+      org.bukkit.craftbukkit.v1_20_R3.inventory.CraftStonecuttingRecipe recipe = new org.bukkit.craftbukkit.v1_20_R3.inventory.CraftStonecuttingRecipe(id, result, org.bukkit.craftbukkit.v1_20_R3.inventory.CraftRecipe.toBukkit(this.ingredient));
+      recipe.setGroup(this.group);
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
