--- a/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -18,4 +_,16 @@
    public ItemStack getToastSymbol() {
       return new ItemStack(Blocks.STONECUTTER);
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack result = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(this.result);
+
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftStonecuttingRecipe recipe = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftStonecuttingRecipe(org.bukkit.craftbukkit.v1_20_R1.util.CraftNamespacedKey.fromMinecraft(this.id), result, org.bukkit.craftbukkit.v1_20_R1.inventory.CraftRecipe.toBukkit(this.ingredient));
+      recipe.setGroup(this.group);
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
