--- a/net/minecraft/client/server/IntegratedServer.java
+++ b/net/minecraft/client/server/IntegratedServer.java
@@ -15,6 +_,7 @@
 import net.minecraft.SystemReport;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.multiplayer.ClientPacketListener;
+import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.Services;
 import net.minecraft.server.WorldStem;
@@ -28,6 +_,7 @@
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -59,11 +_,12 @@
       this.setPvpAllowed(true);
       this.setFlightAllowed(true);
       this.initializeKeyPair();
-      this.loadLevel();
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
+      this.loadLevel(storageSource.getLevelId()); // CraftBukkit
       GameProfile gameprofile = this.getSingleplayerProfile();
       String s = this.getWorldData().getLevelName();
       this.setMotd(gameprofile != null ? gameprofile.getName() + " - " + s : s);
-      return true;
+      return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void tickServer(BooleanSupplier p_120049_) {
@@ -121,6 +_,13 @@
       return true;
    }
 
+   // Ketting start
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return null;
+   }
+   // Ketting end
+
    public File getServerDirectory() {
       return this.minecraft.gameDirectory;
    }
@@ -196,6 +_,7 @@
    }
 
    public void halt(boolean p_120053_) {
+      if (isRunning())
       this.executeBlocking(() -> {
          for(ServerPlayer serverplayer : Lists.newArrayList(this.getPlayerList().getPlayers())) {
             if (!serverplayer.getUUID().equals(this.uuid)) {
