--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -29,10 +_,10 @@
    private static final Logger LOGGER = LogUtils.getLogger();
    private static final String ENTITIES_TAG = "Entities";
    private static final String POSITION_TAG = "Position";
-   private final ServerLevel level;
+   public final ServerLevel level;
    private final IOWorker worker;
    private final LongSet emptyChunks = new LongOpenHashSet();
-   private final ProcessorMailbox<Runnable> entityDeserializerQueue;
+   public final ProcessorMailbox<Runnable> entityDeserializerQueue;
    protected final DataFixer fixerUpper;
 
    public EntityStorage(ServerLevel p_196924_, Path p_196925_, DataFixer p_196926_, boolean p_196927_, Executor p_196928_) {
@@ -89,8 +_,12 @@
          ListTag listtag = new ListTag();
          p_156559_.getEntities().forEach((p_156567_) -> {
             CompoundTag compoundtag1 = new CompoundTag();
+            try {
             if (p_156567_.save(compoundtag1)) {
                listtag.add(compoundtag1);
+            }
+            } catch (Exception e) {
+               LOGGER.error("An Entity type {} has thrown an exception trying to write state. It will not persist. Report this to the mod author", p_156567_.getType(), e);
             }
 
          });
