--- a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -25,6 +_,12 @@
       this.sync = p_196955_;
    }
 
+   public boolean ketting$existingOnly; // CraftBukkit // Ketting
+   private RegionFile getRegionFile(ChunkPos p_63712_, boolean existingOnly) throws IOException { // CraftBukkit
+      this.ketting$existingOnly = existingOnly;
+      return this.getRegionFile(p_63712_);
+   }
+   
    private RegionFile getRegionFile(ChunkPos p_63712_) throws IOException {
       long i = ChunkPos.asLong(p_63712_.getRegionX(), p_63712_.getRegionZ());
       RegionFile regionfile = this.regionCache.getAndMoveToFirst(i);
@@ -37,6 +_,7 @@
 
          FileUtil.createDirectoriesSafe(this.folder);
          Path path = this.folder.resolve("r." + p_63712_.getRegionX() + "." + p_63712_.getRegionZ() + ".mca");
+         if (ketting$existingOnly && !java.nio.file.Files.exists(path)) return null; // CraftBukkit
          RegionFile regionfile1 = new RegionFile(path, this.folder, this.sync);
          this.regionCache.putAndMoveToFirst(i, regionfile1);
          return regionfile1;
@@ -45,7 +_,13 @@
 
    @Nullable
    public CompoundTag read(ChunkPos p_63707_) throws IOException {
+      // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
+      ketting$existingOnly = true; // Ketting
       RegionFile regionfile = this.getRegionFile(p_63707_);
+      if (regionfile == null) {
+         return null;
+      }
+      // CraftBukkit end
 
       try (DataInputStream datainputstream = regionfile.getChunkDataInputStream(p_63707_)) {
          return datainputstream == null ? null : NbtIo.read(datainputstream);
@@ -53,7 +_,13 @@
    }
 
    public void scanChunk(ChunkPos p_196957_, StreamTagVisitor p_196958_) throws IOException {
+      // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
+      ketting$existingOnly = true; // Ketting
       RegionFile regionfile = this.getRegionFile(p_196957_);
+      if (regionfile == null) {
+         return;
+      }
+      // CraftBukkit end
 
       try (DataInputStream datainputstream = regionfile.getChunkDataInputStream(p_196957_)) {
          if (datainputstream != null) {
@@ -64,6 +_,7 @@
    }
 
    protected void write(ChunkPos p_63709_, @Nullable CompoundTag p_63710_) throws IOException {
+      ketting$existingOnly = false; // CraftBukkit // Ketting
       RegionFile regionfile = this.getRegionFile(p_63709_);
       if (p_63710_ == null) {
          regionfile.clear(p_63709_);
