--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -46,6 +_,7 @@
    static final Logger f_214805_ = LogUtils.getLogger();
    private static final Pattern f_214806_ = Pattern.compile("^[a-fA-F0-9]{40}$");
    private static final Splitter f_243988_ = Splitter.on(',').trimResults();
+   public final boolean debug = this.m_139836_("debug", false); // CraftBukkit
    public final boolean f_139728_ = this.m_139836_("online-mode", true);
    public final boolean f_139729_ = this.m_139836_("prevent-proxy-connections", false);
    public final String f_139730_ = this.m_139811_("server-ip", "");
@@ -104,8 +_,10 @@
    private final DedicatedServerProperties.WorldDimensionData f_244504_;
    public final WorldOptions f_243824_;
 
-   public DedicatedServerProperties(Properties p_180926_) {
-      super(p_180926_);
+   // CraftBukkit start
+   public DedicatedServerProperties(Properties properties, joptsimple.OptionSet optionset) {
+      super(properties, optionset);
+      // CraftBukkit end
       String s = this.m_139811_("level-seed", "");
       boolean flag = this.m_139836_("generate-structures", true);
       long i = WorldOptions.m_261063_(s).orElse(WorldOptions.m_255240_());
@@ -119,12 +_,14 @@
       this.f_243997_ = m_247013_(this.m_139811_("initial-enabled-packs", String.join(",", WorldDataConfiguration.f_244649_.f_244096_().m_45850_())), this.m_139811_("initial-disabled-packs", String.join(",", WorldDataConfiguration.f_244649_.f_244096_().m_45855_())));
    }
 
-   public static DedicatedServerProperties m_180929_(Path p_180930_) {
-      return new DedicatedServerProperties(m_139839_(p_180930_));
+   // CraftBukkit start
+   public static DedicatedServerProperties fromFile(Path p_180930_, joptsimple.OptionSet optionset) {
+      return new DedicatedServerProperties(m_139839_(p_180930_), optionset);
    }
 
-   protected DedicatedServerProperties m_6764_(RegistryAccess p_139761_, Properties p_139762_) {
-      return new DedicatedServerProperties(p_139762_);
+   protected DedicatedServerProperties reload(RegistryAccess p_139761_, Properties p_139762_, joptsimple.OptionSet optionset) {
+      return new DedicatedServerProperties(f_139798_, optionset);
+      // CraftBukkit end
    }
 
    @Nullable
