--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -60,7 +_,7 @@
    public final List<ConsoleInput> f_139600_ = Collections.synchronizedList(Lists.newArrayList());
    @Nullable
    private QueryThreadGs4 f_139601_;
-   private final RconConsoleSource f_139602_;
+   // private final RconConsoleSource rconConsoleSource; // CraftBukkit - remove field
    @Nullable
    private RconThread f_139603_;
    public DedicatedServerSettings f_139604_;
@@ -68,23 +_,60 @@
    private MinecraftServerGui f_139605_;
    @Nullable
    private final TextFilterClient f_139606_;
+   @Nullable
+   public org.kettingpowered.ketting.common.betterservergui.BetterServerGUI betterGui; //Ketting
+   @Nullable
+   private net.minecraft.client.server.LanServerPinger dediLanPinger;
 
    public DedicatedServer(Thread p_214789_, LevelStorageSource.LevelStorageAccess p_214790_, PackRepository p_214791_, WorldStem p_214792_, DedicatedServerSettings p_214793_, DataFixer p_214794_, Services p_214795_, ChunkProgressListenerFactory p_214796_) {
       super(p_214789_, p_214790_, p_214791_, p_214792_, Proxy.NO_PROXY, p_214794_, p_214795_, p_214796_);
       this.f_139604_ = p_214793_;
-      this.f_139602_ = new RconConsoleSource(this);
+      // this.rconConsoleSource = new RconConsoleSource(this); // CraftBukkit - remove field
       this.f_139606_ = TextFilterClient.m_143736_(p_214793_.m_139777_().f_139724_);
    }
 
    public boolean m_7038_() throws IOException {
       Thread thread = new Thread("Server console handler") {
          public void run() {
-            BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
+            if (net.minecraftforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
+
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.Main.useConsole) {
+               return;
+            }
+            jline.console.ConsoleReader bufferedreader = reader;
+
+            // MC-33041, SPIGOT-5538: if System.in is not valid due to javaw, then return
+            try {
+               System.in.available();
+            } catch (IOException ex) {
+               return;
+            }
+            // CraftBukkit end
 
             String s1;
             try {
-               while(!DedicatedServer.this.m_129918_() && DedicatedServer.this.m_130010_() && (s1 = bufferedreader.readLine()) != null) {
-                  DedicatedServer.this.m_139645_(s1, DedicatedServer.this.m_129893_());
+               // CraftBukkit start - JLine disabling compatibility
+               while (!DedicatedServer.this.m_129918_() && DedicatedServer.this.m_130010_()) {
+                  if (org.bukkit.craftbukkit.Main.useJline) {
+                     s1 = bufferedreader.readLine(">", null);
+                  } else {
+                     s1 = bufferedreader.readLine();
+                  }
+
+                  // SPIGOT-5220: Throttle if EOF (ctrl^d) or stdin is /dev/null
+                  if (s1 == null) {
+                     try {
+                        Thread.sleep(50L);
+                     } catch (InterruptedException ex) {
+                        Thread.currentThread().interrupt();
+                     }
+                     continue;
+                  }
+                  if (s1.trim().length() > 0) { // Trim to filter lines which are just spaces
+                     DedicatedServer.this.m_139645_(s1, DedicatedServer.this.m_129893_());
+                  }
+                  // CraftBukkit end
                }
             } catch (IOException ioexception1) {
                DedicatedServer.f_139598_.error("Exception handling console input", (Throwable)ioexception1);
@@ -92,6 +_,28 @@
 
          }
       };
+
+      // CraftBukkit start - TODO: handle command-line logging arguments
+      java.util.logging.Logger global = java.util.logging.Logger.getLogger("");
+      global.setUseParentHandlers(false);
+      for (java.util.logging.Handler handler : global.getHandlers()) {
+         global.removeHandler(handler);
+      }
+      global.addHandler(new org.bukkit.craftbukkit.util.ForwardLogHandler());
+
+      final org.apache.logging.log4j.core.Logger logger = ((org.apache.logging.log4j.core.Logger) org.apache.logging.log4j.LogManager.getRootLogger());
+      for (org.apache.logging.log4j.core.Appender appender : logger.getAppenders().values()) {
+         if (appender instanceof org.apache.logging.log4j.core.appender.ConsoleAppender) {
+            logger.removeAppender(appender);
+         }
+      }
+
+      // new org.bukkit.craftbukkit.util.TerminalConsoleWriterThread(System.out, this.reader).start(); //Ketting - disable for now
+
+      System.setOut(org.apache.logging.log4j.io.IoBuilder.forLogger(logger).setLevel(org.apache.logging.log4j.Level.INFO).buildPrintStream());
+      System.setErr(org.apache.logging.log4j.io.IoBuilder.forLogger(logger).setLevel(org.apache.logging.log4j.Level.WARN).buildPrintStream());
+      // CraftBukkit end
+
       thread.setDaemon(true);
       thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(f_139598_));
       thread.start();
@@ -109,13 +_,21 @@
          this.m_129993_(dedicatedserverproperties.f_139729_);
          this.m_129913_(dedicatedserverproperties.f_139730_);
       }
+      // Spigot start
+      this.m_129823_(new DedicatedPlayerList(this, this.m_247573_(), this.f_129745_));
+      org.spigotmc.SpigotConfig.init((java.io.File) options.valueOf("spigot-settings"));
+      org.spigotmc.SpigotConfig.registerCommands();
+      // Spigot end
+
+      org.kettingpowered.ketting.inject.ForgeInject.doInjectingMagic(); //Ketting
+      org.kettingpowered.ketting.command.commands.KettingCommands.registerCommands(this); //Ketting
 
       this.m_129997_(dedicatedserverproperties.f_139733_);
       this.m_129999_(dedicatedserverproperties.f_139734_);
       this.m_129989_(dedicatedserverproperties.f_139736_);
       super.m_7196_(dedicatedserverproperties.f_139725_.get());
       this.m_130004_(dedicatedserverproperties.f_139738_);
-      this.f_129749_.m_5458_(dedicatedserverproperties.f_139740_);
+      // this.worldData.setGameType(dedicatedserverproperties.gamemode); // CraftBukkit - moved to world loading
       f_139598_.info("Default game type: {}", (Object)dedicatedserverproperties.f_139740_);
       InetAddress inetaddress = null;
       if (!this.m_130009_().isEmpty()) {
@@ -138,9 +_,23 @@
          return false;
       }
 
+      // CraftBukkit start
+      // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage)); // Spigot - moved up
+      server.loadPlugins();
+      server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+      // CraftBukkit end
+
       if (!this.m_129797_()) {
          f_139598_.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
          f_139598_.warn("The server will make no attempt to authenticate usernames. Beware.");
+         // Spigot start
+         if (org.spigotmc.SpigotConfig.bungee) {
+            DedicatedServer.f_139598_.warn("Whilst this makes it possible to use BungeeCord, unless access to your server is properly restricted, it also opens up the ability for hackers to connect with any username they choose.");
+            DedicatedServer.f_139598_.warn("Please see http://www.spigotmc.org/wiki/firewall-guide/ for further information.");
+         } else {
+            DedicatedServer.f_139598_.warn("While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose.");
+         }
+         // Spigot end
          f_139598_.warn("While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose.");
          f_139598_.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
       }
@@ -152,15 +_,18 @@
       if (!OldUsersConverter.m_11106_(this)) {
          return false;
       } else {
-         this.m_129823_(new DedicatedPlayerList(this, this.m_247573_(), this.f_129745_));
+         // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage)); // CraftBukkit - moved up
          long i = Util.m_137569_();
          SkullBlockEntity.m_222885_(this.f_236721_, this);
          GameProfileCache.m_11004_(this.m_129797_());
+         if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
          f_139598_.info("Preparing level \"{}\"", (Object)this.m_7123_());
+         this.prepareLoadLevel(f_129744_.m_78277_()); //Ketting
          this.m_130006_();
          long j = Util.m_137569_() - i;
          String s = String.format(Locale.ROOT, "%.3fs", (double)j / 1.0E9D);
          f_139598_.info("Done ({})! For help, type \"help\"", (Object)s);
+         this.f_129726_ = Util.m_137550_(); //Forge: Update server time to prevent watchdog/spaming during long load.
          if (dedicatedserverproperties.f_139744_ != null) {
             this.m_129900_().m_46170_(GameRules.f_46153_).m_46246_(dedicatedserverproperties.f_139744_, this);
          }
@@ -175,7 +_,7 @@
             this.f_139603_ = RconThread.m_11615_(this);
          }
 
-         if (this.m_139669_() > 0L) {
+         if (false && this.m_139669_() > 0L) { // Spigot - disable
             Thread thread1 = new Thread(new ServerWatchdog(this));
             thread1.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(f_139598_));
             thread1.setName("Server Watchdog");
@@ -188,7 +_,14 @@
             f_139598_.info("JMX monitoring enabled");
          }
 
-         return true;
+         if (net.minecraftforge.common.ForgeConfig.SERVER.advertiseDedicatedServerToLan.get()) {
+            this.dediLanPinger = new net.minecraft.client.server.LanServerPinger(this.m_129916_(), String.valueOf(this.m_7448_()));
+            this.dediLanPinger.start();
+         }
+
+         org.kettingpowered.ketting.metrics.MetricManager.init(this); //Ketting
+
+         return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
       }
    }
 
@@ -251,9 +_,16 @@
          this.f_139606_.close();
       }
 
-      if (this.f_139605_ != null) {
-         this.f_139605_.m_139925_();
-      }
+      //Ketting start
+      /*
+      if (this.gui != null) {
+         this.gui.close();
+      }
+      */
+      if (this.betterGui != null) {
+         this.betterGui.close();
+      }
+      //Ketting end
 
       if (this.f_139603_ != null) {
          this.f_139603_.m_7530_();
@@ -263,6 +_,10 @@
          this.f_139601_.m_7530_();
       }
 
+      if (this.dediLanPinger != null) {
+         this.dediLanPinger.interrupt();
+         this.dediLanPinger = null;
+      }
    }
 
    public void m_5703_(BooleanSupplier p_139661_) {
@@ -279,11 +_,22 @@
    }
 
    public void m_139665_() {
+      org.bukkit.craftbukkit.SpigotTimings.serverCommandTimer.startTiming(); // Spigot
       while(!this.f_139600_.isEmpty()) {
          ConsoleInput consoleinput = this.f_139600_.remove(0);
-         this.m_129892_().m_230957_(consoleinput.f_135929_, consoleinput.f_135928_);
+
+         // CraftBukkit start - ServerCommand for preprocessing
+         org.bukkit.event.server.ServerCommandEvent event = new org.bukkit.event.server.ServerCommandEvent(console, consoleinput.f_135928_);
+         server.getPluginManager().callEvent(event);
+         if (event.isCancelled()) continue;
+         consoleinput = new ConsoleInput(event.getCommand(), consoleinput.f_135929_);
+
+         // this.getCommands().performPrefixedCommand(consoleinput.source, consoleinput.msg); // Called in dispatchServerCommand
+         server.dispatchServerCommand(console, consoleinput);
+         // CraftBukkit end
       }
 
+      org.bukkit.craftbukkit.SpigotTimings.serverCommandTimer.stopTiming(); // Spigot
    }
 
    public boolean m_6982_() {
@@ -319,14 +_,26 @@
    }
 
    public void m_139667_() {
-      if (this.f_139605_ == null) {
-         this.f_139605_ = MinecraftServerGui.m_139921_(this);
-      }
-
+      //Ketting start
+      /*
+      if (this.gui == null) {
+         this.gui = MinecraftServerGui.showFrameFor(this);
+      }
+      */
+      if (this.betterGui == null) {
+         final org.kettingpowered.ketting.adapters.BetterServerGUIAdapterImpl adapter = new org.kettingpowered.ketting.adapters.BetterServerGUIAdapterImpl(this);
+         this.betterGui = new org.kettingpowered.ketting.common.betterservergui.BetterServerGUI.Builder()
+                 .setAdapter(adapter)
+                 .build();
+         adapter.startAppender(this.betterGui);
+         this.m_129946_(betterGui::tick);
+      }
+      //Ketting end
    }
 
    public boolean m_6370_() {
-      return this.f_139605_ != null;
+      //return this.gui != null; //Ketting
+      return this.betterGui != null;
    }
 
    public boolean m_6993_() {
@@ -474,15 +_,51 @@
    }
 
    public String m_7138_() {
-      return "";
+      // CraftBukkit start - Whole method
+      StringBuilder result = new StringBuilder();
+      org.bukkit.plugin.Plugin[] plugins = server.getPluginManager().getPlugins();
+
+      result.append(server.getName());
+      result.append(" on Bukkit ");
+      result.append(server.getBukkitVersion());
+
+      if (plugins.length > 0 && server.getQueryPlugins()) {
+         result.append(": ");
+
+         for (int i = 0; i < plugins.length; i++) {
+            if (i > 0) {
+               result.append("; ");
+            }
+
+            result.append(plugins[i].getDescription().getName());
+            result.append(" ");
+            result.append(plugins[i].getDescription().getVersion().replaceAll(";", ","));
+         }
+      }
+
+      return result.toString();
+      // CraftBukkit end
    }
 
    public String m_7261_(String p_139644_) {
-      this.f_139602_.m_11512_();
+      // CraftBukkit start - fire RemoteServerCommandEvent
+      throw new UnsupportedOperationException("Not supported - remote source required.");
+   }
+
+   public String runCommand(RconConsoleSource rconConsoleSource, String s) {
+      rconConsoleSource.m_11512_();
       this.m_18709_(() -> {
-         this.m_129892_().m_230957_(this.f_139602_.m_11514_(), p_139644_);
+         CommandSourceStack wrapper = rconConsoleSource.m_11514_();
+         org.bukkit.event.server.RemoteServerCommandEvent event = new org.bukkit.event.server.RemoteServerCommandEvent(rconConsoleSource.getBukkitSender(wrapper), s);
+         server.getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return;
+         }
+         ConsoleInput consoleInput = new ConsoleInput(event.getCommand(), wrapper);
+         server.dispatchServerCommand(event.getSender(), consoleInput);
       });
-      return this.f_139602_.m_11513_();
+      return rconConsoleSource.m_11513_();
+      // CraftBukkit end
    }
 
    public void m_139688_(boolean p_139689_) {
@@ -492,7 +_,12 @@
    }
 
    public void m_7041_() {
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.GameShuttingDownEvent());
       super.m_7041_();
+      if (this.dediLanPinger != null) {
+         this.dediLanPinger.interrupt();
+         this.dediLanPinger = null;
+      }
       Util.m_137580_();
       SkullBlockEntity.m_196704_();
    }
@@ -525,4 +_,49 @@
    public Optional<MinecraftServer.ServerResourcePackInfo> m_214042_() {
       return this.f_139604_.m_139777_().f_214802_;
    }
+
+   // CraftBukkit start
+   public boolean isDebugging() {
+      return this.m_7913_().debug;
+   }
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return console;
+   }
+
+   public void initWorld(ServerLevel worldserver, net.minecraft.world.level.storage.PrimaryLevelData iworlddataserver, net.minecraft.world.level.storage.PrimaryLevelData saveData, net.minecraft.world.level.levelgen.WorldOptions worldoptions) {
+      boolean flag = saveData.m_7513_();
+      // CraftBukkit start
+      if (worldserver.generator != null) {
+         worldserver.getWorld().getPopulators().addAll(worldserver.generator.getDefaultPopulators(worldserver.getWorld()));
+      }
+      net.minecraft.world.level.border.WorldBorder worldborder = worldserver.m_6857_();
+      worldborder.m_61931_(iworlddataserver.m_5813_()); // CraftBukkit - move up so that net.minecraft.world.level.border.WorldBorder is set during WorldInitEvent
+      this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(worldserver.getWorld())); // CraftBukkit - SPIGOT-5569: Call WorldInitEvent before any chunks are generated
+
+      if (!iworlddataserver.m_6535_()) {
+         try {
+            m_177896_(worldserver, iworlddataserver, worldoptions.m_245100_(), flag);
+            iworlddataserver.m_5555_(true);
+            if (flag) {
+               this.m_129847_(this.f_129749_);
+            }
+         } catch (Throwable throwable) {
+            net.minecraft.CrashReport crashreport = net.minecraft.CrashReport.m_127521_(throwable, "Exception initializing level");
+
+            try {
+               worldserver.m_6026_(crashreport);
+            } catch (Throwable throwable1) {
+               ;
+            }
+
+            throw new net.minecraft.ReportedException(crashreport);
+         }
+
+         iworlddataserver.m_5555_(true);
+      }
+
+   }
+   // CraftBukkit end
 }
