--- a/net/minecraft/server/dedicated/DedicatedServerSettings.java
+++ b/net/minecraft/server/dedicated/DedicatedServerSettings.java
@@ -7,9 +_,11 @@
    private final Path f_139772_;
    private DedicatedServerProperties f_139773_;
 
-   public DedicatedServerSettings(Path p_180932_) {
-      this.f_139772_ = p_180932_;
-      this.f_139773_ = DedicatedServerProperties.m_180929_(p_180932_);
+   // CraftBukkit start
+   public DedicatedServerSettings(joptsimple.OptionSet optionset) {
+      this.f_139772_ = ((java.io.File) optionset.valueOf("config")).toPath();
+      this.f_139773_ = DedicatedServerProperties.fromFile(f_139772_, optionset);
+      // CraftBukkit end
    }
 
    public DedicatedServerProperties m_139777_() {
