--- a/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
@@ -21,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 
 public class MinecartCommandBlock extends AbstractMinecart {
-   static final EntityDataAccessor<String> DATA_ID_COMMAND_NAME = SynchedEntityData.defineId(MinecartCommandBlock.class, EntityDataSerializers.STRING);
+   public static final EntityDataAccessor<String> DATA_ID_COMMAND_NAME = SynchedEntityData.defineId(MinecartCommandBlock.class, EntityDataSerializers.STRING);
    static final EntityDataAccessor<Component> DATA_ID_LAST_OUTPUT = SynchedEntityData.defineId(MinecartCommandBlock.class, EntityDataSerializers.COMPONENT);
    private final BaseCommandBlock commandBlock = new MinecartCommandBlock.MinecartCommandBase();
    private static final int ACTIVATION_DELAY = 4;
@@ -78,6 +_,8 @@
    }
 
    public InteractionResult interact(Player p_38522_, InteractionHand p_38523_) {
+      InteractionResult ret = super.interact(p_38522_, p_38523_);
+      if (ret.consumesAction()) return ret;
       return this.commandBlock.usedBy(p_38522_);
    }
 
@@ -99,6 +_,13 @@
    }
 
    public class MinecartCommandBase extends BaseCommandBlock {
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return (org.bukkit.craftbukkit.entity.CraftMinecartCommand) MinecartCommandBlock.this.getBukkitEntity();
+      }
+      // CraftBukkit end
+
       public ServerLevel getLevel() {
          return (ServerLevel)MinecartCommandBlock.this.level();
       }
