--- a/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/MinecartCommandBlock.java
@@ -78,6 +_,8 @@
    }
 
    public InteractionResult interact(Player p_38522_, InteractionHand p_38523_) {
+      InteractionResult ret = super.interact(p_38522_, p_38523_);
+      if (ret.consumesAction()) return ret;
       return this.commandBlock.usedBy(p_38522_);
    }
 
@@ -99,6 +_,13 @@
    }
 
    public class MinecartCommandBase extends BaseCommandBlock {
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return (org.bukkit.craftbukkit.v1_20_R1.entity.CraftMinecartCommand) MinecartCommandBlock.this.getBukkitEntity();
+      }
+      // CraftBukkit end
+
       public ServerLevel getLevel() {
          return (ServerLevel)MinecartCommandBlock.this.level();
       }
