--- a/net/minecraft/world/entity/vehicle/VehicleEntity.java
+++ b/net/minecraft/world/entity/vehicle/VehicleEntity.java
@@ -27,6 +_,18 @@
          if (this.isInvulnerableTo(p_310829_)) {
             return false;
          } else {
+            // CraftBukkit start
+            org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+            org.bukkit.entity.Entity attacker = (p_310829_.getEntity() == null) ? null : p_310829_.getEntity().getBukkitEntity();
+
+            org.bukkit.event.vehicle.VehicleDamageEvent event = new org.bukkit.event.vehicle.VehicleDamageEvent(vehicle, attacker, (double) p_310313_);
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return false;
+            }
+            p_310313_ = (float) event.getDamage();
+            // CraftBukkit end
             this.setHurtDir(-this.getHurtDir());
             this.setHurtTime(10);
             this.markHurt();
@@ -35,9 +_,27 @@
             boolean flag = p_310829_.getEntity() instanceof Player && ((Player)p_310829_.getEntity()).getAbilities().instabuild;
             if ((flag || !(this.getDamage() > 40.0F)) && !this.shouldSourceDestroy(p_310829_)) {
                if (flag) {
+                  // CraftBukkit start
+                  org.bukkit.event.vehicle.VehicleDestroyEvent destroyEvent = new org.bukkit.event.vehicle.VehicleDestroyEvent(vehicle, attacker);
+                  this.level().getCraftServer().getPluginManager().callEvent(destroyEvent);
+
+                  if (destroyEvent.isCancelled()) {
+                      this.setDamage(40.0F); // Maximize damage so this doesn't get triggered again right away
+                      return true;
+                  }
+                  // CraftBukkit end
                   this.discard();
                }
             } else {
+               // CraftBukkit start
+               org.bukkit.event.vehicle.VehicleDestroyEvent destroyEvent = new org.bukkit.event.vehicle.VehicleDestroyEvent(vehicle, attacker);
+               this.level().getCraftServer().getPluginManager().callEvent(destroyEvent);
+
+               if (destroyEvent.isCancelled()) {
+                   this.setDamage(40.0F); // Maximize damage so this doesn't get triggered again right away
+                   return true;
+               }
+               // CraftBukkit end
                this.destroy(p_310829_);
             }
 
