--- a/net/minecraft/world/entity/ai/behavior/GoToWantedItem.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToWantedItem.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
 import net.minecraft.world.entity.item.ItemEntity;
+import org.bukkit.craftbukkit.v1_20_R2.entity.CraftEntity;
 
 public class GoToWantedItem {
    public static BehaviorControl<LivingEntity> create(float p_260027_, boolean p_259769_, int p_259671_) {
@@ -23,7 +_,22 @@
             return (p_258380_, p_258381_, p_258382_) -> {
                ItemEntity itementity = p_258371_.get(p_258389_);
                if (p_258371_.tryGet(p_258390_).isEmpty() && p_259490_.test(p_258381_) && itementity.closerThan(p_258381_, (double)p_259054_) && p_258381_.level().getWorldBorder().isWithinBounds(itementity.blockPosition())) {
+                  // CraftBukkit start
+                  if (p_258381_ instanceof net.minecraft.world.entity.animal.allay.Allay) {
+                     org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityTargetEvent(p_258381_, itementity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+
+                     if (event.isCancelled()) {
+                        return false;
+                     }
+                     if (!(event.getTarget() instanceof ItemEntity)) {
+                        p_258389_.erase();
+                     }
+
+                     itementity = (ItemEntity) ((CraftEntity) event.getTarget()).getHandle();
+                  }
+                  // CraftBukkit end
                   WalkTarget walktarget = new WalkTarget(new EntityTracker(itementity, false), p_260346_, 0);
+
                   p_258387_.set(new EntityTracker(itementity, true));
                   p_258388_.set(walktarget);
                   return true;
