--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -40,7 +_,7 @@
    }
 
    protected boolean checkExtraStartConditions(ServerLevel p_23174_, Villager p_23175_) {
-      if (!p_23174_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_23174_, p_23175_)) {
          return false;
       } else if (p_23175_.getVillagerData().getProfession() != VillagerProfession.FARMER) {
          return false;
@@ -98,7 +_,9 @@
             Block block = blockstate.getBlock();
             Block block1 = p_23196_.getBlockState(this.aboveFarmlandPos.below()).getBlock();
             if (block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockstate)) {
-               p_23196_.destroyBlock(this.aboveFarmlandPos, true, p_23197_);
+               if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.aboveFarmlandPos, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState())) { // CraftBukkit
+                  p_23196_.destroyBlock(this.aboveFarmlandPos, true, p_23197_);
+               } // CraftBukkit
             }
 
             if (blockstate.isAir() && block1 instanceof FarmBlock && p_23197_.hasFarmSeeds()) {
@@ -112,9 +_,23 @@
                      if ($$11 instanceof BlockItem) {
                         BlockItem blockitem = (BlockItem)$$11;
                         BlockState blockstate1 = blockitem.getBlock().defaultBlockState();
-                        p_23196_.setBlockAndUpdate(this.aboveFarmlandPos, blockstate1);
-                        p_23196_.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(p_23197_, blockstate1));
-                        flag = true;
+
+                        if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.aboveFarmlandPos, blockstate1)) { // CraftBukkit
+                           p_23196_.setBlockAndUpdate(this.aboveFarmlandPos, blockstate1);
+                           p_23196_.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(p_23197_, blockstate1));
+                           flag = true;
+                        } // CraftBukkit
+                     } else if (itemstack.getItem() instanceof net.minecraftforge.common.IPlantable plantable) { //Ketting - variable name
+                        if (plantable.getPlantType(p_23196_, aboveFarmlandPos) == net.minecraftforge.common.PlantType.CROP) {
+                           //Ketting start - fire bukkit event
+                           BlockState blockstate2 = plantable.getPlant(p_23196_, aboveFarmlandPos);
+
+                           if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.aboveFarmlandPos, blockstate2)) { // CraftBukkit
+                              p_23196_.setBlock(aboveFarmlandPos, blockstate2, 3);
+                              flag = true;
+                           }
+                           //Ketting end
+                        }
                      }
                   }
 
