--- a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
@@ -56,6 +_,13 @@
             return this.f_147718_.m_26885_(p_147737_, p_147789_);
          });
       }).ifPresent((p_147778_) -> {
+         // CraftBukkit start
+         org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityTargetLivingEvent(p_147737_, p_147778_, (p_147778_ instanceof net.minecraft.server.level.ServerPlayer) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY);
+         if (event.isCancelled() || event.getTarget() == null) {
+            return;
+         }
+         p_147778_ = ((org.bukkit.craftbukkit.v1_20_R2.entity.CraftLivingEntity) event.getTarget()).getHandle();
+         // CraftBukkit end
          this.m_147765_(p_147737_, p_147778_);
       });
    }
