--- a/net/minecraft/world/entity/ai/behavior/ResetProfession.java
+++ b/net/minecraft/world/entity/ai/behavior/ResetProfession.java
@@ -13,7 +_,14 @@
             return (p_260244_, p_260084_, p_259597_) -> {
                VillagerData villagerdata = p_260084_.getVillagerData();
                if (villagerdata.getProfession() != VillagerProfession.NONE && villagerdata.getProfession() != VillagerProfession.NITWIT && p_260084_.getVillagerXp() == 0 && villagerdata.getLevel() <= 1) {
-                  p_260084_.setVillagerData(p_260084_.getVillagerData().setProfession(VillagerProfession.NONE));
+                  // CraftBukkit start
+                  org.bukkit.event.entity.VillagerCareerChangeEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callVillagerCareerChangeEvent(p_260084_, org.bukkit.craftbukkit.v1_20_R1.entity.CraftVillager.nmsToBukkitProfession(VillagerProfession.NONE), org.bukkit.event.entity.VillagerCareerChangeEvent.ChangeReason.LOSING_JOB);
+                  if (event.isCancelled()) {
+                     return false;
+                  }
+
+                  p_260084_.setVillagerData(p_260084_.getVillagerData().setProfession(org.bukkit.craftbukkit.v1_20_R1.entity.CraftVillager.bukkitToNmsProfession(event.getProfession())));
+                  // CraftBukkit end
                   p_260084_.refreshBrain(p_260244_);
                   return true;
                } else {
