--- a/net/minecraft/world/entity/ai/behavior/StartAttacking.java
+++ b/net/minecraft/world/entity/ai/behavior/StartAttacking.java
@@ -30,7 +_,22 @@
                      if (!p_258774_.m_6779_(livingentity)) {
                         return false;
                      } else {
-                        p_258778_.m_257512_(livingentity);
+                        net.minecraftforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.minecraftforge.common.ForgeHooks.onLivingChangeTarget(p_258774_, livingentity, net.minecraftforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.BEHAVIOR_TARGET);
+                        if (changeTargetEvent.isCanceled())
+                           return false;
+
+                        // CraftBukkit start
+                        org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTargetLivingEvent(p_258774_, changeTargetEvent.getNewTarget(), (changeTargetEvent.getNewTarget() instanceof net.minecraft.server.level.ServerPlayer) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY); //Ketting - forge compat
+                        if (event.isCancelled()) {
+                           return false;
+                        }
+                        if (event.getTarget() == null) {
+                           p_258779_.m_257971_();
+                           return true;
+                        }
+                        livingentity = ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) event.getTarget()).getHandle();
+                        // CraftBukkit end
+                        p_258778_.m_257512_(livingentity); //Ketting
                         p_258779_.m_257971_();
                         return true;
                      }
