--- a/net/minecraft/world/entity/ai/behavior/StartAttacking.java
+++ b/net/minecraft/world/entity/ai/behavior/StartAttacking.java
@@ -3,10 +_,10 @@
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.Predicate;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.behavior.declarative.BehaviorBuilder;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 
 public class StartAttacking {
    public static <E extends Mob> BehaviorControl<E> create(Function<E, Optional<? extends LivingEntity>> p_259868_) {
@@ -30,7 +_,22 @@
                      if (!p_258774_.canAttack(livingentity)) {
                         return false;
                      } else {
-                        p_258778_.set(livingentity);
+                        net.minecraftforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.minecraftforge.common.ForgeHooks.onLivingChangeTarget(p_258774_, livingentity, net.minecraftforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.BEHAVIOR_TARGET);
+                        if (changeTargetEvent.isCanceled())
+                           return false;
+
+                        // CraftBukkit start
+                        org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityTargetLivingEvent(p_258774_, changeTargetEvent.getNewTarget(), (changeTargetEvent.getNewTarget() instanceof net.minecraft.server.level.ServerPlayer) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY); //Ketting - forge compat
+                        if (event.isCancelled()) {
+                           return false;
+                        }
+                        if (event.getTarget() == null) {
+                           p_258779_.erase();
+                           return true;
+                        }
+                        livingentity = ((org.bukkit.craftbukkit.v1_20_R2.entity.CraftLivingEntity) event.getTarget()).getHandle();
+                        // CraftBukkit end
+                        p_258778_.set(livingentity); //Ketting
                         p_258779_.erase();
                         return true;
                      }
