--- a/net/minecraft/client/multiplayer/ClientHandshakePacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientHandshakePacketListenerImpl.java
@@ -129,6 +_,7 @@
       this.updateStatus.accept(Component.translatable("connect.joining"));
       this.localGameProfile = p_104547_.getGameProfile();
       this.connection.setProtocol(ConnectionProtocol.PLAY);
+      net.minecraftforge.network.NetworkHooks.handleClientLoginSuccess(this.connection);
       this.connection.setListener(new ClientPacketListener(this.minecraft, this.parent, this.connection, this.serverData, this.localGameProfile, this.minecraft.getTelemetryManager().createWorldSessionManager(this.newWorld, this.worldLoadDuration, this.minigameName)));
    }
 
@@ -136,7 +_,7 @@
       if (this.parent != null && this.parent instanceof RealmsScreen) {
          this.minecraft.setScreen(new DisconnectedRealmsScreen(this.parent, CommonComponents.CONNECT_FAILED, p_104543_));
       } else {
-         this.minecraft.setScreen(new DisconnectedScreen(this.parent, CommonComponents.CONNECT_FAILED, p_104543_));
+         this.minecraft.setScreen(net.minecraftforge.network.NetworkHooks.getModMismatchData(connection) != null ? new net.minecraftforge.client.gui.ModMismatchDisconnectedScreen(this.parent, CommonComponents.CONNECT_FAILED, p_104543_, net.minecraftforge.network.NetworkHooks.getModMismatchData(connection)) : new DisconnectedScreen(this.parent, CommonComponents.CONNECT_FAILED, p_104543_));
       }
 
    }
@@ -157,6 +_,7 @@
    }
 
    public void handleCustomQuery(ClientboundCustomQueryPacket p_104545_) {
+      if (net.minecraftforge.network.NetworkHooks.onCustomPayload(p_104545_, this.connection)) return;
       this.updateStatus.accept(Component.translatable("connect.negotiating"));
       this.connection.send(new ServerboundCustomQueryPacket(p_104545_.getTransactionId(), (FriendlyByteBuf)null));
    }
