--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -412,6 +_,7 @@
 
       this.minecraft.debugRenderer.clear();
       this.minecraft.player.resetPos();
+      net.minecraftforge.client.ForgeHooksClient.firePlayerLogin(this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().connection);
       int i = p_105030_.playerId();
       this.minecraft.player.setId(i);
       this.level.addPlayer(i, this.minecraft.player);
@@ -425,6 +_,7 @@
       this.minecraft.player.setPortalCooldown(p_105030_.portalCooldown());
       this.minecraft.gameMode.setLocalMode(p_105030_.gameType(), p_105030_.previousGameType());
       this.minecraft.options.setServerRenderDistance(p_105030_.chunkRadius());
+      net.minecraftforge.network.NetworkHooks.sendMCRegistryPackets(connection, "PLAY_TO_SERVER");
       this.minecraft.options.broadcastOptions();
       this.connection.send(new ServerboundCustomPayloadPacket(ServerboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf(ClientBrandRetriever.getClientModName())));
       this.chatSession = null;
@@ -817,7 +_,7 @@
 
    public void handleSystemChat(ClientboundSystemChatPacket p_233708_) {
       PacketUtils.ensureRunningOnSameThread(p_233708_, this, this.minecraft);
-      this.minecraft.getChatListener().handleSystemMessage(p_233708_.content(), p_233708_.overlay());
+     // this.minecraft.getChatListener().handleSystemMessage(p_233708_.content(), p_233708_.overlay());
    }
 
    public void handlePlayerChat(ClientboundPlayerChatPacket p_233702_) {
@@ -1069,8 +_,10 @@
          localplayer1.getAttributes().assignValues(localplayer.getAttributes());
       }
 
+      localplayer1.updateSyncFields(localplayer); // Forge: fix MC-10657
       localplayer1.resetPos();
       localplayer1.setServerBrand(s);
+      net.minecraftforge.client.ForgeHooksClient.firePlayerRespawn(this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.connection);
       this.level.addPlayer(i, localplayer1);
       localplayer1.setYRot(-180.0F);
       localplayer1.input = new KeyboardInput(this.minecraft.options);
@@ -1180,10 +_,7 @@
       PacketUtils.ensureRunningOnSameThread(p_104976_, this, this.minecraft);
       BlockPos blockpos = p_104976_.getPos();
       this.minecraft.level.getBlockEntity(blockpos, p_104976_.getType()).ifPresent((p_205557_) -> {
-         CompoundTag compoundtag = p_104976_.getTag();
-         if (compoundtag != null) {
-            p_205557_.load(compoundtag);
-         }
+         p_205557_.onDataPacket(connection, p_104976_);
 
          if (p_205557_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
             ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
@@ -1329,7 +_,9 @@
 
    public void handleCommands(ClientboundCommandsPacket p_104990_) {
       PacketUtils.ensureRunningOnSameThread(p_104990_, this, this.minecraft);
-      this.commands = new CommandDispatcher<>(p_104990_.getRoot(CommandBuildContext.simple(this.registryAccess.compositeAccess(), this.enabledFeatures)));
+      var context = CommandBuildContext.simple(this.registryAccess.compositeAccess(), this.enabledFeatures);
+      this.commands = new CommandDispatcher<>(p_104990_.getRoot(context));
+      this.commands = net.minecraftforge.client.ClientCommandHandler.mergeServerCommands(this.commands, context);
    }
 
    public void handleStopSoundEvent(ClientboundStopSoundPacket p_105116_) {
@@ -1348,6 +_,7 @@
       ClientRecipeBook clientrecipebook = this.minecraft.player.getRecipeBook();
       clientrecipebook.setupCollections(this.recipeManager.getRecipes(), this.minecraft.level.registryAccess());
       this.minecraft.populateSearchTree(SearchRegistry.RECIPE_COLLECTIONS, clientrecipebook.getCollections());
+      net.minecraftforge.client.ForgeHooksClient.onRecipesUpdated(this.recipeManager);
    }
 
    public void handleLookAt(ClientboundPlayerLookAtPacket p_105054_) {
@@ -1443,7 +_,8 @@
          Blocks.rebuildCache();
       }
 
-      CreativeModeTabs.searchTab().rebuildSearchTree();
+      CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(net.minecraft.world.item.CreativeModeTab::rebuildSearchTree);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent(this.registryAccess.compositeAccess(), true, connection.isMemoryConnection()));
    }
 
    public void handleEnabledFeatures(ClientboundUpdateEnabledFeaturesPacket p_251591_) {
@@ -1855,6 +_,7 @@
    }
 
    public void handleCustomPayload(ClientboundCustomPayloadPacket p_105004_) {
+      if (net.minecraftforge.network.NetworkHooks.onCustomPayload(p_105004_, this.connection)) return;
       PacketUtils.ensureRunningOnSameThread(p_105004_, this, this.minecraft);
       ResourceLocation resourcelocation = p_105004_.getIdentifier();
       FriendlyByteBuf friendlybytebuf = null;
@@ -2052,7 +_,7 @@
             int j5 = friendlybytebuf.readVarInt();
             this.minecraft.debugRenderer.gameEventListenerRenderer.trackListener(positionsource, j5);
          } else {
-            LOGGER.warn("Unknown custom packed identifier: {}", (Object)resourcelocation);
+            LOGGER.warn("Unknown custom packet identifier: {}", (Object)resourcelocation);
          }
       } finally {
          if (friendlybytebuf != null) {
@@ -2399,6 +_,8 @@
    }
 
    public void sendChat(String p_249888_) {
+      p_249888_ = net.minecraftforge.client.ForgeHooksClient.onClientSendMessage(p_249888_);
+      if (p_249888_.isEmpty()) return;
       Instant instant = Instant.now();
       long i = Crypt.SaltSupplier.getLong();
       LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
@@ -2407,6 +_,7 @@
    }
 
    public void sendCommand(String p_250092_) {
+      if (net.minecraftforge.client.ClientCommandHandler.runCommand(p_250092_)) return;
       Instant instant = Instant.now();
       long i = Crypt.SaltSupplier.getLong();
       LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
