--- a/net/minecraft/world/effect/PoisonMobEffect.java
+++ b/net/minecraft/world/effect/PoisonMobEffect.java
@@ -10,7 +_,7 @@
    public void m_6742_(LivingEntity p_299064_, int p_299074_) {
       super.m_6742_(p_299064_, p_299074_);
       if (p_299064_.m_21223_() > 1.0F) {
-         p_299064_.m_6469_(p_299064_.m_269291_().m_269425_(), 1.0F);
+         p_299064_.m_6469_(p_299064_.m_269291_().poison, 1.0F); // CraftBukkit - DamageSource.MAGIC -> CraftEventFactory.POISON
       }
 
    }
