--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -22,10 +_,14 @@
       if (resourcekey == null) {
          return false;
       } else {
-         Holder<ConfiguredFeature<?, ?>> holder = p_222905_.m_9598_().m_175515_(Registries.f_256911_).m_203636_(resourcekey).orElse((Holder.Reference<ConfiguredFeature<?, ?>>)null);
+         Holder<ConfiguredFeature<?, ?>> holder = p_222905_.m_9598_().m_175515_(Registries.f_256911_).m_203636_(resourcekey).orElse(null);
+         var event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_222905_, p_222909_, p_222907_, holder);
+         holder = event.getFeature();
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.DENY) return false;
          if (holder == null) {
             return false;
          } else {
+            setTreeType(holder); // CraftBukkit
             ConfiguredFeature<?, ?> configuredfeature = holder.m_203334_();
             BlockState blockstate = p_222905_.m_6425_(p_222907_).m_76188_();
             p_222905_.m_7731_(p_222907_, blockstate, 4);
@@ -52,4 +_,58 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void setTreeType(Holder<ConfiguredFeature<?, ?>> holder) {
+      ResourceKey<ConfiguredFeature<?, ?>> worldgentreeabstract = holder.m_203543_().get();
+      if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195123_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195142_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195122_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195121_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195129_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195131_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195128_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195127_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195126_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195125_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195108_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195135_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195137_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195130_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195111_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195138_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195124_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195133_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195134_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195132_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195139_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_236762_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MANGROVE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_236763_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_MANGROVE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_271469_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_271485_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CHERRY;
+      } else {
+         //Ketting start
+         // throw new IllegalArgumentException("Unknown tree generator " + worldgentreeabstract);
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CUSTOM;
+         //Ketting end
+      }
+   }
+   // CraftBukkit end
 }
