--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -22,10 +_,14 @@
       if (resourcekey == null) {
          return false;
       } else {
-         Holder<ConfiguredFeature<?, ?>> holder = p_222905_.registryAccess().registryOrThrow(Registries.CONFIGURED_FEATURE).getHolder(resourcekey).orElse((Holder.Reference<ConfiguredFeature<?, ?>>)null);
+         Holder<ConfiguredFeature<?, ?>> holder = p_222905_.registryAccess().registryOrThrow(Registries.CONFIGURED_FEATURE).getHolder(resourcekey).orElse(null);
+         var event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_222905_, p_222909_, p_222907_, holder);
+         holder = event.getFeature();
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.DENY) return false;
          if (holder == null) {
             return false;
          } else {
+            setTreeType(holder); // CraftBukkit
             ConfiguredFeature<?, ?> configuredfeature = holder.value();
             BlockState blockstate = p_222905_.getFluidState(p_222907_).createLegacyBlock();
             p_222905_.setBlock(p_222907_, blockstate, 4);
@@ -52,4 +_,58 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void setTreeType(Holder<ConfiguredFeature<?, ?>> holder) {
+      ResourceKey<ConfiguredFeature<?, ?>> worldgentreeabstract = holder.unwrapKey().get();
+      if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.OAK || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.OAK_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.HUGE_RED_MUSHROOM) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.HUGE_BROWN_MUSHROOM) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.JUNGLE_TREE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.JUNGLE_TREE_NO_VINE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.PINE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.SPRUCE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.ACACIA) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.BIRCH || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.BIRCH_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.SUPER_BIRCH_BEES_0002) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.SWAMP_OAK) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.FANCY_OAK || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.FANCY_OAK_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.JUNGLE_BUSH) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.DARK_OAK) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MEGA_SPRUCE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MEGA_PINE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MEGA_JUNGLE_TREE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.AZALEA_TREE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MANGROVE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MANGROVE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.TALL_MANGROVE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_MANGROVE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.CHERRY || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.CHERRY_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CHERRY;
+      } else {
+         //Ketting start
+         // throw new IllegalArgumentException("Unknown tree generator " + worldgentreeabstract);
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CUSTOM;
+         //Ketting end
+      }
+   }
+   // CraftBukkit end
 }
