--- a/net/minecraft/world/level/block/grower/TreeGrower.java
+++ b/net/minecraft/world/level/block/grower/TreeGrower.java
@@ -18,9 +_,9 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
 
 public final class TreeGrower {
    private static final Map<String, TreeGrower> GROWERS = new Object2ObjectArrayMap<>();
@@ -85,7 +_,11 @@
       ResourceKey<ConfiguredFeature<?, ?>> resourcekey = this.getConfiguredMegaFeature(p_309951_);
       if (resourcekey != null) {
          Holder<ConfiguredFeature<?, ?>> holder = p_309830_.registryAccess().registryOrThrow(Registries.CONFIGURED_FEATURE).getHolder(resourcekey).orElse((Holder.Reference<ConfiguredFeature<?, ?>>)null);
+         var event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_309830_, p_309951_, p_310327_, holder);
+         holder = event.getFeature();
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.DENY) return false;
          if (holder != null) {
+            setTreeType(holder); // CraftBukkit
             for(int i = 0; i >= -1; --i) {
                for(int j = 0; j >= -1; --j) {
                   if (isTwoByTwoSapling(p_312382_, p_309830_, p_310327_, i, j)) {
@@ -118,6 +_,7 @@
          if (holder1 == null) {
             return false;
          } else {
+            setTreeType(holder1); // CraftBukkit
             ConfiguredFeature<?, ?> configuredfeature1 = holder1.value();
             BlockState blockstate1 = p_309830_.getFluidState(p_310327_).createLegacyBlock();
             p_309830_.setBlock(p_310327_, blockstate1, 4);
@@ -134,6 +_,56 @@
          }
       }
    }
+   // CraftBukkit start
+   private void setTreeType(Holder<ConfiguredFeature<?, ?>> holder) {
+       ResourceKey<ConfiguredFeature<?, ?>> worldgentreeabstract = holder.unwrapKey().get();
+       if (worldgentreeabstract == TreeFeatures.OAK || worldgentreeabstract == TreeFeatures.OAK_BEES_005) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+       } else if (worldgentreeabstract == TreeFeatures.HUGE_RED_MUSHROOM) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+       } else if (worldgentreeabstract == TreeFeatures.HUGE_BROWN_MUSHROOM) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+       } else if (worldgentreeabstract == TreeFeatures.JUNGLE_TREE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+       } else if (worldgentreeabstract == TreeFeatures.JUNGLE_TREE_NO_VINE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+       } else if (worldgentreeabstract == TreeFeatures.PINE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+       } else if (worldgentreeabstract == TreeFeatures.SPRUCE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+       } else if (worldgentreeabstract == TreeFeatures.ACACIA) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+       } else if (worldgentreeabstract == TreeFeatures.BIRCH || worldgentreeabstract == TreeFeatures.BIRCH_BEES_005) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+       } else if (worldgentreeabstract == TreeFeatures.SUPER_BIRCH_BEES_0002) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+       } else if (worldgentreeabstract == TreeFeatures.SWAMP_OAK) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+       } else if (worldgentreeabstract == TreeFeatures.FANCY_OAK || worldgentreeabstract == TreeFeatures.FANCY_OAK_BEES_005) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+       } else if (worldgentreeabstract == TreeFeatures.JUNGLE_BUSH) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+       } else if (worldgentreeabstract == TreeFeatures.DARK_OAK) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+       } else if (worldgentreeabstract == TreeFeatures.MEGA_SPRUCE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+       } else if (worldgentreeabstract == TreeFeatures.MEGA_PINE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+       } else if (worldgentreeabstract == TreeFeatures.MEGA_JUNGLE_TREE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+       } else if (worldgentreeabstract == TreeFeatures.AZALEA_TREE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+       } else if (worldgentreeabstract == TreeFeatures.MANGROVE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MANGROVE;
+       } else if (worldgentreeabstract == TreeFeatures.TALL_MANGROVE) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_MANGROVE;
+       } else if (worldgentreeabstract == TreeFeatures.CHERRY || worldgentreeabstract == TreeFeatures.CHERRY_BEES_005) {
+           net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CHERRY;
+       } else {
+           throw new IllegalArgumentException("Unknown tree generator " + worldgentreeabstract);
+       }
+   }
+   // CraftBukkit end
 
    private static boolean isTwoByTwoSapling(BlockState p_310256_, BlockGetter p_311754_, BlockPos p_312442_, int p_310725_, int p_310118_) {
       Block block = p_310256_.getBlock();
