--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -118,11 +_,18 @@
       return i;
    }
 
-   static class Tracer implements ServerFunctionManager.TraceCallbacks, CommandSource {
+   public static class Tracer implements ServerFunctionManager.TraceCallbacks, CommandSource {
       public static final int INDENT_OFFSET = 1;
       private final PrintWriter output;
       private int lastIndent;
       private boolean waitingForResult;
+
+      //Ketting start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return new org.bukkit.craftbukkit.v1_20_R2.command.DebugCommandCommandSender(wrapper, Tracer.this);
+      }
+      //Ketting end
 
       Tracer(PrintWriter p_180079_) {
          this.output = p_180079_;
