--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -117,6 +_,13 @@
       private int f_180076_;
       private boolean f_180077_;
 
+      //Ketting start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return new org.bukkit.craftbukkit.v1_20_R2.command.DebugCommandCommandSender(wrapper, Tracer.this);
+      }
+      //Ketting end
+
       Tracer(PrintWriter p_180079_) {
          this.f_180075_ = p_180079_;
       }
