--- a/net/minecraft/server/commands/PlaceCommand.java
+++ b/net/minecraft/server/commands/PlaceCommand.java
@@ -121,6 +_,7 @@
       if (!structurestart.m_73603_()) {
          throw f_214532_.create();
       } else {
+         structurestart.generationEventCause = org.bukkit.event.world.AsyncStructureGenerateEvent.Cause.COMMAND; // CraftBukkit - set AsyncStructureGenerateEvent.Cause.COMMAND as generation cause
          BoundingBox boundingbox = structurestart.m_73601_();
          ChunkPos chunkpos = new ChunkPos(SectionPos.m_123171_(boundingbox.m_162395_()), SectionPos.m_123171_(boundingbox.m_162398_()));
          ChunkPos chunkpos1 = new ChunkPos(SectionPos.m_123171_(boundingbox.m_162399_()), SectionPos.m_123171_(boundingbox.m_162401_()));
