--- a/net/minecraft/server/commands/ReloadCommand.java
+++ b/net/minecraft/server/commands/ReloadCommand.java
@@ -37,6 +_,16 @@
       return collection;
    }
 
+   // CraftBukkit start
+   public static void reload(MinecraftServer minecraftserver) {
+      PackRepository resourcepackrepository = minecraftserver.m_129891_();
+      WorldData savedata = minecraftserver.m_129910_();
+      Collection<String> collection = resourcepackrepository.m_10523_();
+      Collection<String> collection1 = m_138222_(resourcepackrepository, savedata, collection);
+      minecraftserver.m_129861_(collection1);
+   }
+   // CraftBukkit end
+
    public static void m_138226_(CommandDispatcher<CommandSourceStack> p_138227_) {
       p_138227_.register(Commands.m_82127_("reload").requires((p_138231_) -> {
          return p_138231_.m_6761_(2);
