--- a/net/minecraft/server/commands/TriggerCommand.java
+++ b/net/minecraft/server/commands/TriggerCommand.java
@@ -85,7 +_,7 @@
       if (p_139140_.getCriteria() != ObjectiveCriteria.TRIGGER) {
          throw ERROR_INVALID_OBJECTIVE.create();
       } else {
-         Scoreboard scoreboard = p_139139_.getScoreboard();
+         Scoreboard scoreboard = p_139139_.getServer().getScoreboard(); // CraftBukkit - SPIGOT-6917: use main scoreboard
          String s = p_139139_.getScoreboardName();
          if (!scoreboard.hasPlayerScore(s, p_139140_)) {
             throw ERROR_NOT_PRIMED.create();
