--- a/net/minecraft/world/level/border/WorldBorder.java
+++ b/net/minecraft/world/level/border/WorldBorder.java
@@ -27,6 +_,7 @@
    int absoluteMaxSize = 29999984;
    private WorldBorder.BorderExtent extent = new WorldBorder.StaticBorderExtent((double)5.999997E7F);
    public static final WorldBorder.Settings DEFAULT_SETTINGS = new WorldBorder.Settings(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, (double)5.999997E7F, 0L, 0.0D);
+   public net.minecraft.world.level.Level world; // CraftBukkit
 
    public boolean isWithinBounds(BlockPos p_61938_) {
       return (double)(p_61938_.getX() + 1) > this.getMinX() && (double)p_61938_.getX() < this.getMaxX() && (double)(p_61938_.getZ() + 1) > this.getMinZ() && (double)p_61938_.getZ() < this.getMaxZ();
@@ -149,6 +_,7 @@
    }
 
    public void addListener(BorderChangeListener p_61930_) {
+      if (listeners.contains(p_61930_)) return; // CraftBukkit
       this.listeners.add(p_61930_);
    }
 
