--- a/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -204,6 +_,19 @@
          if (this.isInvulnerableTo(p_38117_)) {
             return false;
          } else {
+            // CraftBukkit start - fire VehicleDamageEvent
+            Vehicle vehicle = (Vehicle) this.getBukkitEntity();
+            org.bukkit.entity.Entity passenger = (p_38117_.getEntity() == null) ? null : p_38117_.getEntity().getBukkitEntity();
+
+            VehicleDamageEvent event = new VehicleDamageEvent(vehicle, passenger, p_38118_);
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               return false;
+            }
+
+            p_38118_ = (float) event.getDamage();
+            // CraftBukkit end
             this.setHurtDir(-this.getHurtDir());
             this.setHurtTime(10);
             this.markHurt();
@@ -211,6 +_,15 @@
             this.gameEvent(GameEvent.ENTITY_DAMAGE, p_38117_.getEntity());
             boolean flag = p_38117_.getEntity() instanceof Player && ((Player)p_38117_.getEntity()).getAbilities().instabuild;
             if (flag || this.getDamage() > 40.0F) {
+               // CraftBukkit start
+               VehicleDestroyEvent destroyEvent = new VehicleDestroyEvent(vehicle, passenger);
+               this.level().getCraftServer().getPluginManager().callEvent(destroyEvent);
+
+               if (destroyEvent.isCancelled()) {
+                  this.setDamage(40); // Maximize damage so this doesn't get triggered again right away
+                  return true;
+               }
+               // CraftBukkit end
                this.ejectPassengers();
                if (flag && !this.hasCustomName()) {
                   this.discard();
