--- a/net/minecraft/world/inventory/CartographyTableMenu.java
+++ b/net/minecraft/world/inventory/CartographyTableMenu.java
@@ -13,6 +_,22 @@
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
 
 public class CartographyTableMenu extends AbstractContainerMenu {
+
+   // CraftBukkit start
+   private org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryCartography inventory = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryCartography(this.f_39135_, this.f_39138_);
+      bukkitEntity = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
    public static final int f_150501_ = 0;
    public static final int f_150502_ = 1;
    public static final int f_150503_ = 2;
@@ -22,18 +_,8 @@
    private static final int f_150507_ = 39;
    private final ContainerLevelAccess f_39136_;
    long f_39137_;
-   public final Container f_39135_ = new SimpleContainer(2) {
-      public void m_6596_() {
-         CartographyTableMenu.this.m_6199_(this);
-         super.m_6596_();
-      }
-   };
-   private final ResultContainer f_39138_ = new ResultContainer() {
-      public void m_6596_() {
-         CartographyTableMenu.this.m_6199_(this);
-         super.m_6596_();
-      }
-   };
+   public final Container f_39135_;
+   private final ResultContainer f_39138_;
 
    public CartographyTableMenu(int p_39140_, Inventory p_39141_) {
       this(p_39140_, p_39141_, ContainerLevelAccess.f_39287_);
@@ -41,6 +_,32 @@
 
    public CartographyTableMenu(int p_39143_, Inventory p_39144_, final ContainerLevelAccess p_39145_) {
       super(MenuType.f_39979_, p_39143_);
+      this.f_39135_ = new SimpleContainer(2) {
+         public void m_6596_() {
+            CartographyTableMenu.this.m_6199_(this);
+            super.m_6596_();
+         }
+
+         // CraftBukkit start
+         @Override
+         public org.bukkit.Location getLocation() {
+            return p_39145_.getLocation();
+         }
+         // CraftBukkit end
+      };
+      this.f_39138_ = new ResultContainer() {
+         public void m_6596_() {
+            CartographyTableMenu.this.m_6199_(this);
+            super.m_6596_();
+         }
+
+         // CraftBukkit start
+         @Override
+         public org.bukkit.Location getLocation() {
+            return p_39145_.getLocation();
+         }
+         // CraftBukkit end
+      };
       this.f_39136_ = p_39145_;
       this.m_38897_(new Slot(this.f_39135_, 0, 15, 15) {
          public boolean m_5857_(ItemStack p_39194_) {
@@ -83,9 +_,11 @@
          this.m_38897_(new Slot(p_39144_, k, 8 + k * 18, 142));
       }
 
+      player = (org.bukkit.entity.Player) p_39144_.f_35978_.getBukkitEntity(); // CraftBukkit
    }
 
    public boolean m_6875_(Player p_39149_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return m_38889_(this.f_39136_, p_39149_, Blocks.f_50621_);
    }
 
