--- a/net/minecraft/world/inventory/ContainerLevelAccess.java
+++ b/net/minecraft/world/inventory/ContainerLevelAccess.java
@@ -7,6 +_,21 @@
 import net.minecraft.world.level.Level;
 
 public interface ContainerLevelAccess {
+
+   // CraftBukkit start
+   default Level getWorld() {
+      throw new UnsupportedOperationException("Not supported yet.");
+   }
+
+   default BlockPos getPosition() {
+      throw new UnsupportedOperationException("Not supported yet.");
+   }
+
+   default org.bukkit.Location getLocation() {
+      return new org.bukkit.Location(getWorld().getWorld(), getPosition().getX(), getPosition().getY(), getPosition().getZ());
+   }
+   // CraftBukkit end
+
    ContainerLevelAccess NULL = new ContainerLevelAccess() {
       public <T> Optional<T> evaluate(BiFunction<Level, BlockPos, T> p_39304_) {
          return Optional.empty();
@@ -15,6 +_,18 @@
 
    static ContainerLevelAccess create(final Level p_39290_, final BlockPos p_39291_) {
       return new ContainerLevelAccess() {
+         // CraftBukkit start
+         @Override
+         public Level getWorld() {
+            return p_39290_;
+         }
+
+         @Override
+         public BlockPos getPosition() {
+            return p_39291_;
+         }
+         // CraftBukkit end
+
          public <T> Optional<T> evaluate(BiFunction<Level, BlockPos, T> p_39311_) {
             return Optional.of(p_39311_.apply(p_39290_, p_39291_));
          }
