--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -43,10 +_,13 @@
       if (player instanceof ServerPlayer serverplayer) {
          Container container = this.container;
          if (container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
+            abstractfurnaceblockentity.ketting$awardUsedRecipesAndPopExperience$itemstack.set(p_39558_);
+            abstractfurnaceblockentity.ketting$awardUsedRecipesAndPopExperience$amount.set(this.removeCount);
             abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer);
          }
       }
 
       this.removeCount = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.player, p_39558_);
    }
 }
