--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -43,10 +_,11 @@
       if (player instanceof ServerPlayer serverplayer) {
          Container container = this.container;
          if (container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
-            abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer);
+            abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer, p_39558_, this.removeCount); // CraftBukkit
          }
       }
 
       this.removeCount = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.player, p_39558_);
    }
 }
