--- a/net/minecraft/world/inventory/InventoryMenu.java
+++ b/net/minecraft/world/inventory/InventoryMenu.java
@@ -33,15 +_,25 @@
    public static final ResourceLocation EMPTY_ARMOR_SLOT_SHIELD = new ResourceLocation("item/empty_armor_slot_shield");
    static final ResourceLocation[] TEXTURE_EMPTY_SLOTS = new ResourceLocation[]{EMPTY_ARMOR_SLOT_BOOTS, EMPTY_ARMOR_SLOT_LEGGINGS, EMPTY_ARMOR_SLOT_CHESTPLATE, EMPTY_ARMOR_SLOT_HELMET};
    private static final EquipmentSlot[] SLOT_IDS = new EquipmentSlot[]{EquipmentSlot.HEAD, EquipmentSlot.CHEST, EquipmentSlot.LEGS, EquipmentSlot.FEET};
-   private final CraftingContainer craftSlots = new TransientCraftingContainer(this, 2, 2);
+   public final CraftingContainer craftSlots = new TransientCraftingContainer(this, 3, 3);
    private final ResultContainer resultSlots = new ResultContainer();
    public final boolean active;
    private final Player owner;
+   // CraftBukkit start
+   private org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+   // CraftBukkit end
 
    public InventoryMenu(Inventory p_39706_, boolean p_39707_, final Player p_39708_) {
       super((MenuType<?>)null, 0);
       this.active = p_39707_;
       this.owner = p_39708_;
+      // CraftBukkit start
+      ((TransientCraftingContainer) this.craftSlots).setOwner(p_39708_); // CraftBukkit - pass player
+      ((TransientCraftingContainer) this.craftSlots).resultInventory = this.resultSlots; // CraftBukkit - let InventoryCrafting know about its result slot
+      this.player = p_39706_; // CraftBukkit - save player
+      setTitle(net.minecraft.network.chat.Component.translatable("container.crafting")); // SPIGOT-4722: Allocate title for player inventory
+      // CraftBukkit end
       this.addSlot(new ResultSlot(p_39706_.player, this.craftSlots, this.resultSlots, 0, 154, 28));
 
       for(int i = 0; i < 2; ++i) {
@@ -63,7 +_,7 @@
             }
 
             public boolean mayPlace(ItemStack p_39746_) {
-               return equipmentslot == Mob.getEquipmentSlotForItem(p_39746_);
+               return p_39746_.canEquip(equipmentslot, owner);
             }
 
             public boolean mayPickup(Player p_39744_) {
@@ -232,4 +_,17 @@
    public boolean shouldMoveToInventory(int p_150591_) {
       return p_150591_ != this.getResultSlotIndex();
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryCrafting inventory = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryCrafting(this.craftSlots, this.resultSlots);
+      bukkitEntity = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
