--- a/net/minecraft/world/inventory/LecternMenu.java
+++ b/net/minecraft/world/inventory/LecternMenu.java
@@ -6,6 +_,24 @@
 import net.minecraft.world.item.ItemStack;
 
 public class LecternMenu extends AbstractContainerMenu {
+
+   // CraftBukkit start
+   private org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView bukkitEntity = null;
+   public org.bukkit.entity.Player player;
+
+   @Override
+   public org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      if (this.player == null) this.player = org.kettingpowered.ketting.utils.InventoryViewHelper.getCraftContainerOwner();
+
+      org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryLectern inventory = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryLectern(this.lectern);
+      bukkitEntity = new org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
    private static final int DATA_COUNT = 1;
    private static final int SLOT_COUNT = 1;
    public static final int BUTTON_PREV_PAGE = 1;
@@ -20,6 +_,7 @@
    }
 
    public LecternMenu(int p_39824_, Container p_39825_, ContainerData p_39826_) {
+      // CraftBukkit end
       super(MenuType.LECTERN, p_39824_);
       checkContainerSize(p_39825_, 1);
       checkContainerDataCount(p_39826_, 1);
@@ -54,6 +_,13 @@
                   return false;
                }
 
+               // CraftBukkit start - Event for taking the book
+               org.bukkit.event.player.PlayerTakeLecternBookEvent event = new org.bukkit.event.player.PlayerTakeLecternBookEvent(player, ((org.bukkit.craftbukkit.v1_20_R1.inventory.CraftInventoryLectern) getBukkitView().getTopInventory()).getHolder());
+               org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  return false;
+               }
+               // CraftBukkit end
                ItemStack itemstack = this.lectern.removeItemNoUpdate(0);
                this.lectern.setChanged();
                if (!p_39833_.getInventory().add(itemstack)) {
@@ -77,6 +_,8 @@
    }
 
    public boolean stillValid(Player p_39831_) {
+      if (lectern instanceof net.minecraft.world.level.block.entity.LecternBlockEntity.LecternInventory && !((net.minecraft.world.level.block.entity.LecternBlockEntity.LecternInventory) lectern).getLectern().hasBook()) return false; // CraftBukkit
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.lectern.stillValid(p_39831_);
    }
 
