--- a/net/minecraft/world/inventory/MerchantContainer.java
+++ b/net/minecraft/world/inventory/MerchantContainer.java
@@ -18,6 +_,37 @@
    public int selectionHint;
    private int futureXp;
 
+   // CraftBukkit start - add fields and methods
+   private int maxStack = MAX_STACK;
+
+   public java.util.List<ItemStack> getContents() {
+      return this.itemStacks;
+   }
+
+   public void onClose(org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity who) {
+      Container.super.onClose(who); //Ketting
+      merchant.setTradingPlayer(null); // SPIGOT-4860
+   }
+
+   @Override
+   public int getMaxStackSize() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int i) {
+      maxStack = i;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return (merchant instanceof net.minecraft.world.entity.npc.AbstractVillager) ? (org.bukkit.craftbukkit.v1_20_R1.entity.CraftAbstractVillager) ((net.minecraft.world.entity.npc.AbstractVillager) this.merchant).getBukkitEntity() : null;
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return (merchant instanceof net.minecraft.world.entity.npc.Villager) ? ((net.minecraft.world.entity.npc.Villager) this.merchant).getBukkitEntity().getLocation() : null;
+   }
+   // CraftBukkit end
+
    public MerchantContainer(Merchant p_40003_) {
       this.merchant = p_40003_;
    }
