--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -11,9 +_,26 @@
 public class PlayerEnderChestContainer extends SimpleContainer {
    @Nullable
    private EnderChestBlockEntity f_40101_;
+   // CraftBukkit start
+   private final Player owner;
+
+   public org.bukkit.inventory.InventoryHolder getBukkitOwner() {
+      if (owner == null) return org.kettingpowered.ketting.craftbukkit.CraftStubInventoryHolder.INSTANCE;
+      return owner.getBukkitEntity();
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return this.f_40101_ != null ? org.bukkit.craftbukkit.v1_20_R1.util.CraftLocation.toBukkit(this.f_40101_.m_58899_(), this.f_40101_.m_58904_().getWorld()) : null;
+   }
 
    public PlayerEnderChestContainer() {
+      this(null);//Ketting - keep compat
+   }
+   public PlayerEnderChestContainer(Player owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void m_40105_(EnderChestBlockEntity p_40106_) {
