--- a/net/minecraft/world/inventory/TransientCraftingContainer.java
+++ b/net/minecraft/world/inventory/TransientCraftingContainer.java
@@ -13,6 +_,68 @@
    private final int f_286972_;
    private final AbstractContainerMenu f_286998_;
 
+   // CraftBukkit start - add fields
+   public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+   private net.minecraft.world.item.crafting.RecipeHolder<?> currentRecipe;
+   public net.minecraft.world.Container resultInventory;
+   private Player owner;
+   private int maxStack = MAX_STACK;
+
+   public List<ItemStack> getContents() {
+      return this.f_286951_;
+   }
+
+   public void onOpen(org.bukkit.craftbukkit.v1_20_R2.entity.CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public org.bukkit.event.inventory.InventoryType getInvType() {
+      return f_286951_.size() == 4 ? org.bukkit.event.inventory.InventoryType.CRAFTING : org.bukkit.event.inventory.InventoryType.WORKBENCH;
+   }
+
+   public void onClose(org.bukkit.craftbukkit.v1_20_R2.entity.CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public List<org.bukkit.entity.HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return (owner == null) ? null : owner.getBukkitEntity();
+   }
+
+   @Override
+   public int m_6893_() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+      resultInventory.setMaxStackSize(size);
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return f_286998_ instanceof CraftingMenu ? ((CraftingMenu) f_286998_).f_39350_.getLocation() : owner.getBukkitEntity().getLocation();
+   }
+
+   @Override
+   public @javax.annotation.Nullable net.minecraft.world.item.crafting.RecipeHolder<?> getCurrentRecipe() {
+      return currentRecipe;
+   }
+
+   @Override
+   public void setCurrentRecipe(net.minecraft.world.item.crafting.RecipeHolder<?> currentRecipe) {
+      this.currentRecipe = currentRecipe;
+   }
+
+   public TransientCraftingContainer(AbstractContainerMenu container, int i, int j, Player player) {
+      this(container, i, j);
+      this.owner = player;
+   }
+   // CraftBukkit end
+
    public TransientCraftingContainer(AbstractContainerMenu p_287684_, int p_287629_, int p_287593_) {
       this(p_287684_, p_287629_, p_287593_, NonNullList.m_122780_(p_287629_ * p_287593_, ItemStack.f_41583_));
    }
