--- a/net/minecraft/advancements/AdvancementTree.java
+++ b/net/minecraft/advancements/AdvancementTree.java
@@ -64,7 +_,7 @@
          }
       }
 
-      LOGGER.info("Loaded {} advancements", (int)this.nodes.size());
+      // LOGGER.info("Loaded {} advancements", (int)this.nodes.size()); // CraftBukkit - moved to AdvancementDataWorld#reload
    }
 
    private boolean tryInsert(AdvancementHolder p_300067_) {
