--- a/net/minecraft/advancements/AdvancementTree.java
+++ b/net/minecraft/advancements/AdvancementTree.java
@@ -64,7 +_,7 @@
          }
       }
 
-      f_291893_.info("Loaded {} advancements", (int)this.f_290621_.size());
+      // LOGGER.info("Loaded {} advancements", (int)this.nodes.size()); // CraftBukkit - moved to AdvancementDataWorld#reload
    }
 
    private boolean m_295692_(AdvancementHolder p_300067_) {
