--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -46,6 +_,7 @@
 
    public void setEntityId(EntityType<?> p_253682_, @Nullable Level p_254041_, RandomSource p_254221_, BlockPos p_254050_) {
       this.getOrCreateNextSpawnData(p_254041_, p_254221_, p_254050_).getEntityToSpawn().putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(p_253682_).toString());
+      this.spawnPotentials = SimpleWeightedRandomList.empty(); // CraftBukkit - SPIGOT-3496, MC-92282
    }
 
    private boolean isNearPlayer(Level p_151344_, BlockPos p_151345_) {
@@ -131,16 +_,29 @@
                   entity.moveTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                   if (entity instanceof Mob) {
                      Mob mob = (Mob)entity;
-                     if (spawndata.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(p_151312_, MobSpawnType.SPAWNER) || !mob.checkSpawnObstruction(p_151312_)) {
+                     if (!net.minecraftforge.event.ForgeEventFactory.checkSpawnPositionSpawner(mob, p_151312_, MobSpawnType.SPAWNER, spawndata, this)) {
                         continue;
                      }
 
-                     if (spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().contains("id", 8)) {
-                        ((Mob)entity).finalizeSpawn(p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.SPAWNER, (SpawnGroupData)null, (CompoundTag)null);
-                     }
+                     // Forge: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla normally would trigger it.
+                     var event = net.minecraftforge.event.ForgeEventFactory.onFinalizeSpawnSpawner(mob, p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), null, compoundtag, this);
+                     if (event != null && spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().contains("id", 8)) {
+                        ((Mob)entity).finalizeSpawn(p_151312_, event.getDifficulty(), event.getSpawnType(), event.getSpawnData(), event.getSpawnTag());
+                     }
+                     // Spigot Start
+                     if ( mob.level().spigotConfig.nerfSpawnerMobs )
+                     {
+                        mob.aware = false;
+                     }
+                     // Spigot End
                   }
 
-                  if (!p_151312_.tryAddFreshEntityWithPassengers(entity)) {
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callSpawnerSpawnEvent(entity, p_151313_).isCancelled()) {
+                     continue;
+                  }
+                  if (!p_151312_.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER)) {
+                     // CraftBukkit end
                      this.delay(p_151312_, p_151313_);
                      return;
                   }
@@ -283,4 +_,12 @@
    public double getoSpin() {
       return this.oSpin;
    }
+
+   @Nullable
+   public Entity getSpawnerEntity() {
+      return null;
+   }
+
+   @Nullable
+   public net.minecraft.world.level.block.entity.BlockEntity getSpawnerBlockEntity(){ return null; }
 }
