--- a/net/minecraft/world/level/GameRules.java
+++ b/net/minecraft/world/level/GameRules.java
@@ -210,7 +_,7 @@
          return Boolean.toString(this.value);
       }
 
-      protected void deserialize(String p_46234_) {
+      public void deserialize(String p_46234_) {
          this.value = Boolean.parseBoolean(p_46234_);
       }
 
@@ -299,7 +_,7 @@
          return Integer.toString(this.value);
       }
 
-      protected void deserialize(String p_46307_) {
+      public void deserialize(String p_46307_) {
          this.value = safeParse(p_46307_);
       }
 
@@ -420,14 +_,14 @@
          this.onChanged(p_46371_.getSource().getServer());
       }
 
-      protected void onChanged(@Nullable MinecraftServer p_46369_) {
+      public void onChanged(@Nullable MinecraftServer p_46369_) {
          if (p_46369_ != null) {
             this.type.callback.accept(p_46369_, this.getSelf());
          }
 
       }
 
-      protected abstract void deserialize(String p_46367_);
+      public abstract void deserialize(String p_46367_); // PAIL - private->public
 
       public abstract String serialize();
 
