--- a/net/minecraft/world/level/LevelWriter.java
+++ b/net/minecraft/world/level/LevelWriter.java
@@ -24,6 +_,22 @@
 
    boolean destroyBlock(BlockPos p_46957_, boolean p_46958_, @Nullable Entity p_46959_, int p_46960_);
 
+   //Ketting start
+   // CraftBukkit start
+   /**
+    * Ketting - use addFreshEntity instead with prepareAddFreshEntity
+    */
+   @Deprecated
+   default boolean addFreshEntity(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      prepareAddFreshEntity(reason);
+      return addFreshEntity(entity);
+   }
+   // CraftBukkit end
+   default void prepareAddFreshEntity(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {}
+
+   default void disableSpawnFor(Entity entity) {}
+   //Ketting end
+
    default boolean addFreshEntity(Entity p_46964_) {
       return false;
    }
