--- a/net/minecraft/world/level/ServerLevelAccessor.java
+++ b/net/minecraft/world/level/ServerLevelAccessor.java
@@ -6,7 +_,28 @@
 public interface ServerLevelAccessor extends LevelAccessor {
    ServerLevel getLevel();
 
+   //Ketting start
+   class Reason {
+      private static org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT;
+   }
+
+   default void prepareAddFreshEntityWithPassengers(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      Reason.reason = reason;
+   }
+
    default void addFreshEntityWithPassengers(Entity p_47206_) {
-      p_47206_.getSelfAndPassengers().forEach(this::addFreshEntity);
-   }
+      p_47206_.getSelfAndPassengers().forEach(e -> {
+         this.prepareAddFreshEntity(Reason.reason);
+         this.addFreshEntity(e);
+      });
+      Reason.reason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT;
+      //Ketting end
+   }
+
+   // CraftBukkit start
+   @Override
+   default ServerLevel getMinecraftWorld() {
+      return getLevel();
+   }
+   // CraftBukkit end
 }
