--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -14,12 +_,23 @@
 public class DerivedLevelData implements ServerLevelData {
    private final WorldData f_78076_;
    private final ServerLevelData f_78077_;
+   //Ketting start
+   private net.minecraft.resources.ResourceKey<net.minecraft.world.level.dimension.LevelStem> typeKey;
+
+   public void setTypeKey(net.minecraft.resources.ResourceKey<net.minecraft.world.level.dimension.LevelStem> typeKey) {
+      this.typeKey = typeKey;
+   }
+   //Ketting end
 
    public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_) {
       this.f_78076_ = p_78079_;
       this.f_78077_ = p_78080_;
    }
 
+   public ServerLevelData getDelegate() {
+      return f_78077_;
+   }
+
    public int m_6789_() {
       return this.f_78077_.m_6789_();
    }
@@ -45,7 +_,20 @@
    }
 
    public String m_5462_() {
-      return this.f_78076_.m_5462_();
+      if (typeKey == null || typeKey == net.minecraft.world.level.dimension.LevelStem.f_63971_) {
+         return this.f_78077_.m_5462_();
+      } else {
+         String worldName = this.f_78077_.m_5462_() + "/";
+         String suffix;
+         if (typeKey == net.minecraft.world.level.dimension.LevelStem.f_63973_) {
+            suffix = "DIM1";
+         } else if (typeKey == net.minecraft.world.level.dimension.LevelStem.f_63972_) {
+            suffix = "DIM-1";
+         } else {
+            suffix = typeKey.m_135782_().m_135827_() + "/" + typeKey.m_135782_().m_135815_();
+         }
+         return worldName + suffix;
+      }
    }
 
    public int m_6537_() {
