--- a/net/minecraft/network/protocol/login/ClientboundCustomQueryPacket.java
+++ b/net/minecraft/network/protocol/login/ClientboundCustomQueryPacket.java
@@ -4,7 +_,7 @@
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 
-public class ClientboundCustomQueryPacket implements Packet<ClientLoginPacketListener> {
+public class ClientboundCustomQueryPacket implements Packet<ClientLoginPacketListener>, net.minecraftforge.network.ICustomPacket<ClientboundCustomQueryPacket> {
    private static final int MAX_PAYLOAD_SIZE = 1048576;
    private final int transactionId;
    private final ResourceLocation identifier;
@@ -30,7 +_,7 @@
    public void write(FriendlyByteBuf p_134757_) {
       p_134757_.writeVarInt(this.transactionId);
       p_134757_.writeResourceLocation(this.identifier);
-      p_134757_.writeBytes(this.data.copy());
+      p_134757_.writeBytes(this.data.slice()); // Use Slice instead of copy, to not update the read index, allowing packet to be sent multiple times.
    }
 
    public void handle(ClientLoginPacketListener p_134754_) {
@@ -48,4 +_,8 @@
    public FriendlyByteBuf getData() {
       return this.data;
    }
+
+   @Override public int getIndex() { return getTransactionId(); }
+   @Override public ResourceLocation getName() { return getIdentifier(); }
+   @org.jetbrains.annotations.Nullable @Override public FriendlyByteBuf getInternalData() { return getData(); }
 }
