--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -66,7 +_,7 @@
       SharedConstants.m_142977_();
       OptionParser optionparser = new OptionParser();
       OptionSpec<Void> optionspec = optionparser.accepts("nogui");
-      OptionSpec<Void> optionspec1 = optionparser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
+      OptionSpec<Void> optionspec1 = optionparser.accepts("initSettings", "Only create configuration files and then exit");
       OptionSpec<Void> optionspec2 = optionparser.accepts("demo");
       OptionSpec<Void> optionspec3 = optionparser.accepts("bonusChest");
       OptionSpec<Void> optionspec4 = optionparser.accepts("forceUpgrade");
@@ -80,13 +_,51 @@
       OptionSpec<Void> optionspec13 = optionparser.accepts("jfrProfile");
       OptionSpec<Path> optionspec14 = optionparser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
       OptionSpec<String> optionspec15 = optionparser.nonOptions();
+      optionparser.accepts("allowUpdates").withRequiredArg().ofType(Boolean.class).defaultsTo(Boolean.TRUE); // Forge: allow mod updates to proceed
+      optionparser.accepts("gameDir").withRequiredArg().ofType(File.class).defaultsTo(new File(".")); //Forge: Consume this argument, we use it in the launcher, and the client side.
+      final OptionSpec<net.minecraft.core.BlockPos> spawnPosOpt;
+      boolean gametestEnabled = Boolean.getBoolean("forge.gameTestServer");
+      if (gametestEnabled) {
+         spawnPosOpt = optionparser.accepts("spawnPos").withRequiredArg().withValuesConvertedBy(new net.minecraftforge.gametest.BlockPosValueConverter()).defaultsTo(new net.minecraft.core.BlockPos(0, 60, 0));
+      } else {
+          spawnPosOpt = null;
+      }
+
+      //Ketting start - Bukkit Options
+      acceptBukkitSettings(optionparser);
+      //Ketting end
+
+      //Ketting start - Initialize Ketting
+      org.kettingpowered.ketting.core.Ketting ketting = org.kettingpowered.ketting.core.Ketting.init(SharedConstants.m_183709_().m_132493_());
+      ketting.registerAdapter(new org.kettingpowered.ketting.adapters.Forge_1_20_2_Adapter(), new org.kettingpowered.ketting.adapters.Bukkit_1_20_2_Adapter());
+      //Ketting end
 
       try {
          OptionSet optionset = optionparser.parse(p_129699_);
+         //Ketting start - Bukkit Options
+         bukkitSettings(optionset);
+         //Ketting end
          if (optionset.has(optionspec7)) {
             optionparser.printHelpOn(System.err);
             return;
          }
+         Path path2 = Paths.get("eula.txt");
+         Eula eula = new Eula(path2);
+
+         // Spigot Start
+         boolean eulaAgreed = Boolean.getBoolean( "com.mojang.eula.agree" );
+         if ( eulaAgreed )
+         {
+            System.err.println( "You have used the Spigot command line EULA agreement flag." );
+            System.err.println( "By using this setting you are indicating your agreement to Mojang's EULA (https://account.mojang.com/documents/minecraft_eula)." );
+            System.err.println( "If you do not agree to the above EULA please stop your server and remove this flag immediately." );
+         }
+         // Spigot End
+
+         if (!eula.m_135944_() && !eulaAgreed) { // Spigot
+            f_129670_.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            return;
+         }
 
          Path path = optionset.valueOf(optionspec14);
          if (path != null) {
@@ -98,29 +_,52 @@
             JvmProfiler.f_185340_.m_183425_(Environment.SERVER);
          }
 
-         Bootstrap.m_135870_();
+         try {
+            Bootstrap.m_135870_();
+         } catch (org.spongepowered.asm.mixin.throwables.MixinError mixin) {
+            org.kettingpowered.ketting.core.injectprotect.InjectProtect.onBootErrorCaught(mixin);
+         }
+
          Bootstrap.m_135889_();
+
+         //Ketting start - init dimension registry after bootstrap
+         ketting.registerDimensionRegistry(new org.kettingpowered.ketting.adapters.DimensionRegistry_1_20_2_Adapter());
+         //Ketting end
+
          Util.m_137584_();
          Path path1 = Paths.get("server.properties");
-         DedicatedServerSettings dedicatedserversettings = new DedicatedServerSettings(path1);
+         if (!optionset.has(optionspec1)) net.minecraftforge.server.loading.ServerModLoader.load(); // Load mods before we load almost anything else anymore. Single spot now. Only loads if they haven't passed the initserver param
+         DedicatedServerSettings dedicatedserversettings = new DedicatedServerSettings(optionset); // CraftBukkit - CLI argument support
          dedicatedserversettings.m_139780_();
-         Path path2 = Paths.get("eula.txt");
-         Eula eula = new Eula(path2);
+         org.kettingpowered.ketting.config.KettingConfig.getInstance(); // Ketting
          if (optionset.has(optionspec1)) {
+            // CraftBukkit start - SPIGOT-5761: Create bukkit.yml and commands.yml if not present
+            File configFile = (File) optionset.valueOf("bukkit-settings");
+            org.bukkit.configuration.file.YamlConfiguration configuration = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(configFile);
+            configuration.options().copyDefaults(true);
+            configuration.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/bukkit.yml"), com.google.common.base.Charsets.UTF_8)));
+            configuration.save(configFile);
+
+            File commandFile = (File) optionset.valueOf("commands-settings");
+            org.bukkit.configuration.file.YamlConfiguration commandsConfiguration = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(commandFile);
+            commandsConfiguration.options().copyDefaults(true);
+            commandsConfiguration.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/commands.yml"), com.google.common.base.Charsets.UTF_8)));
+            commandsConfiguration.save(commandFile);
+            // CraftBukkit end
             f_129670_.info("Initialized '{}' and '{}'", path1.toAbsolutePath(), path2.toAbsolutePath());
             return;
          }
 
-         if (!eula.m_135944_()) {
-            f_129670_.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-            return;
-         }
-
          File file1 = new File(optionset.valueOf(optionspec9));
          Services services = Services.m_214344_(new YggdrasilAuthenticationService(Proxy.NO_PROXY), file1);
          String s = Optional.ofNullable(optionset.valueOf(optionspec10)).orElse(dedicatedserversettings.m_139777_().f_139741_);
+         if (s == null || s.isEmpty() || new File(file1, s).getAbsolutePath().equals(new File(s).getAbsolutePath())) {
+            f_129670_.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! " + s);
+            return;
+         }
          LevelStorageSource levelstoragesource = LevelStorageSource.m_78242_(file1.toPath());
-         LevelStorageSource.LevelStorageAccess levelstoragesource$levelstorageaccess = levelstoragesource.m_289864_(s);
+         LevelStorageSource.LevelStorageAccess levelstoragesource$levelstorageaccess = levelstoragesource.validateAndCreateAccess(s, LevelStem.f_63971_);
+         levelstoragesource$levelstorageaccess.readAdditionalLevelSaveData();
          LevelSummary levelsummary = levelstoragesource$levelstorageaccess.m_78308_();
          if (levelsummary != null) {
             if (levelsummary.m_193020_()) {
@@ -141,11 +_,31 @@
 
          PackRepository packrepository = ServerPacksSource.m_247728_(levelstoragesource$levelstorageaccess);
 
+         // CraftBukkit start
+         File bukkitDataPackFolder = new File(levelstoragesource$levelstorageaccess.m_78283_(net.minecraft.world.level.storage.LevelResource.f_78180_).toFile(), "bukkit");
+         if (!bukkitDataPackFolder.exists()) {
+            bukkitDataPackFolder.mkdirs();
+         }
+         File mcMeta = new File(bukkitDataPackFolder, "pack.mcmeta");
+         try {
+            com.google.common.io.Files.write("{\n"
+                    + "    \"pack\": {\n"
+                    + "        \"description\": \"Data pack for resources provided by Bukkit plugins\",\n"
+                    + "        \"pack_format\": " + SharedConstants.m_183709_().m_264084_(net.minecraft.server.packs.PackType.SERVER_DATA) + "\n"
+                    + "    }\n"
+                    + "}\n", mcMeta, com.google.common.base.Charsets.UTF_8);
+         } catch (java.io.IOException ex) {
+            throw new RuntimeException("Could not initialize Bukkit datapack", ex);
+         }
+         java.util.concurrent.atomic.AtomicReference<WorldLoader.DataLoadContext> worldLoader = new java.util.concurrent.atomic.AtomicReference<>();
+         // CraftBukkit end
+
          WorldStem worldstem;
          try {
             WorldLoader.InitConfig worldloader$initconfig = m_245383_(dedicatedserversettings.m_139777_(), levelstoragesource$levelstorageaccess, flag, packrepository);
             worldstem = Util.m_214679_((p_248086_) -> {
                return WorldLoader.m_214362_(worldloader$initconfig, (p_248079_) -> {
+                  worldLoader.set(p_248079_); // CraftBukkit
                   Registry<LevelStem> registry = p_248079_.f_243759_().m_175515_(Registries.f_256862_);
                   DynamicOps<Tag> dynamicops = RegistryOps.m_255058_(NbtOps.f_128958_, p_248079_.f_244104_());
                   Pair<WorldData, WorldDimensions.Complete> pair = levelstoragesource$levelstorageaccess.m_246049_(dynamicops, p_248079_.f_244127_(), registry, p_248079_.f_244104_().m_211816_());
@@ -166,6 +_,8 @@
                         worlddimensions = dedicatedserverproperties.m_246483_(p_248079_.f_244104_());
                      }
 
+                     //Forge: Do a write-read-cycle to inject modded dimensions on first start of a dedicated server into its generated world dimensions list.
+                     worlddimensions = WorldDimensions.f_243810_.encoder().encodeStart(dynamicops, worlddimensions).flatMap((writtenPayloadWithModdedDimensions) -> WorldDimensions.f_243810_.decoder().parse(dynamicops, writtenPayloadWithModdedDimensions)).resultOrPartial(f_129670_::error).orElse(worlddimensions);
                      WorldDimensions.Complete worlddimensions$complete = worlddimensions.m_245300_(registry);
                      Lifecycle lifecycle = worlddimensions$complete.m_245945_().add(p_248079_.f_244104_().m_211816_());
                      return new WorldLoader.DataLoadOutput<>(new PrimaryLevelData(levelsettings, worldoptions, worlddimensions$complete.f_244634_(), lifecycle), worlddimensions$complete.m_245593_());
@@ -190,21 +_,38 @@
 
          WorldData worlddata = worldstem.f_206895_();
          levelstoragesource$levelstorageaccess.m_78287_(registryaccess$frozen, worlddata);
-         final DedicatedServer dedicatedserver = MinecraftServer.m_129872_((p_236710_) -> {
-            DedicatedServer dedicatedserver1 = new DedicatedServer(p_236710_, levelstoragesource$levelstorageaccess, packrepository, worldstem, dedicatedserversettings, DataFixers.m_14512_(), services, LoggerChunkProgressListener::new);
+         final MinecraftServer dedicatedserver = MinecraftServer.m_129872_((p_129697_) -> {
+            MinecraftServer dedicatedserver1;
+            if (gametestEnabled) {
+               net.minecraftforge.gametest.ForgeGameTestHooks.registerGametests();
+               java.util.Collection<net.minecraft.gametest.framework.GameTestBatch> testBatches = net.minecraft.gametest.framework.GameTestRunner.m_127724_(net.minecraft.gametest.framework.GameTestRegistry.m_127658_());
+               net.minecraft.core.BlockPos spawnPos = optionset.valueOf(spawnPosOpt);
+               dedicatedserver1 = new net.minecraft.gametest.framework.GameTestServer(p_129697_, levelstoragesource$levelstorageaccess, packrepository, worldstem, testBatches, spawnPos);
+            } else {
+               dedicatedserver1 = new DedicatedServer(p_129697_, levelstoragesource$levelstorageaccess, packrepository, worldstem, dedicatedserversettings, DataFixers.m_14512_(), services, LoggerChunkProgressListener::new);
+            }
+            //Ketting start - add optionset and worldloader
+            dedicatedserver1.setBukkitOptions(optionset, worldLoader.get());
+            //Ketting end
             dedicatedserver1.m_236740_(optionset.has(optionspec8) ? new GameProfile((UUID)null, optionset.valueOf(optionspec8)) : null);
             dedicatedserver1.m_129801_(optionset.valueOf(optionspec11));
             dedicatedserver1.m_129948_(optionset.valueOf(optionspec12));
             boolean flag1 = !optionset.has(optionspec) && !optionset.valuesOf(optionspec15).contains("nogui");
-            if (flag1 && !GraphicsEnvironment.isHeadless()) {
-               dedicatedserver1.m_139667_();
+            if (dedicatedserver1 instanceof DedicatedServer dedicatedServer && flag1 && !GraphicsEnvironment.isHeadless()) {
+               dedicatedServer.m_139667_();
             }
 
             return dedicatedserver1;
          });
          Thread thread = new Thread("Server Shutdown Thread") {
             public void run() {
+               //Ketting start - InjectProtect
+               org.kettingpowered.ketting.core.injectprotect.InjectProtect.shutdownCalled();
+               //Ketting end
+               // FORGE: Halting as GameTestServer will cause issues as it always calls System#exit on both crash and normal exit, so skip it
+               if (!(dedicatedserver instanceof net.minecraft.gametest.framework.GameTestServer))
                dedicatedserver.m_7570_(true);
+               org.apache.logging.log4j.LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
             }
          };
          thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(f_129670_));
@@ -210,6 +_,64 @@
       }
 
    }
+
+   //Ketting start - Bukkit Options
+   private static void acceptBukkitSettings(OptionParser parser) {
+      parser.acceptsAll(java.util.Arrays.asList("nojline"), "Disables jline and emulates the vanilla console");
+      parser.acceptsAll(java.util.Arrays.asList("noconsole"), "Disables the console");
+      parser.acceptsAll(java.util.Arrays.asList("b", "bukkit-settings"), "File for bukkit settings")
+              .withRequiredArg()
+              .ofType(File.class)
+              .defaultsTo(new File("bukkit.yml"))
+              .describedAs("Yml file");
+      parser.acceptsAll(java.util.Arrays.asList("C", "commands-settings"), "File for command settings")
+              .withRequiredArg()
+              .ofType(File.class)
+              .defaultsTo(new File("commands.yml"))
+              .describedAs("Yml file");
+      // Spigot Start
+      parser.acceptsAll(java.util.Arrays.asList("S", "spigot-settings"), "File for spigot settings")
+              .withRequiredArg()
+              .ofType(File.class)
+              .defaultsTo(new File("spigot.yml"))
+              .describedAs("Yml file");
+      // Spigot End
+      parser.acceptsAll(java.util.Arrays.asList("c", "config"), "Properties file to use")
+              .withRequiredArg()
+              .ofType(File.class)
+              .defaultsTo(new File("server.properties"))
+              .describedAs("Properties file");
+      parser.acceptsAll(java.util.Arrays.asList("P", "plugins"), "Plugin directory to use")
+              .withRequiredArg()
+              .ofType(File.class)
+              .defaultsTo(new File("plugins"))
+              .describedAs("Plugin directory");
+   }
+
+   private static void bukkitSettings(OptionSet options) {
+      // This trick bypasses Maven Shade's clever rewriting of our getProperty call when using String literals
+      String jline_UnsupportedTerminal = new String(new char[]{'j', 'l', 'i', 'n', 'e', '.', 'U', 'n', 's', 'u', 'p', 'p', 'o', 'r', 't', 'e', 'd', 'T', 'e', 'r', 'm', 'i', 'n', 'a', 'l'});
+      String jline_terminal = new String(new char[]{'j', 'l', 'i', 'n', 'e', '.', 't', 'e', 'r', 'm', 'i', 'n', 'a', 'l'});
+
+      org.bukkit.craftbukkit.Main.useJline = !(jline_UnsupportedTerminal).equals(System.getProperty(jline_terminal));
+
+      if (options.has("nojline")) {
+         System.setProperty("user.language", "en");
+         org.bukkit.craftbukkit.Main.useJline = false;
+      }
+
+      if (org.bukkit.craftbukkit.Main.useJline) {
+         org.fusesource.jansi.AnsiConsole.systemInstall();
+      } else {
+         // This ensures the terminal literal will always match the jline implementation
+         System.setProperty(jline.TerminalFactory.JLINE_TERMINAL, jline.UnsupportedTerminal.class.getName());
+      }
+
+      if (options.has("noconsole")) {
+         org.bukkit.craftbukkit.Main.useConsole = false;
+      }
+   }
+   //Ketting end
 
    private static void m_269308_(Path p_270192_) {
       try {
