--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -244,11 +_,30 @@
    private final StructureTemplateManager f_236720_;
    protected final WorldData f_129749_;
    private volatile boolean f_195494_;
+   // CraftBukkit start
+   public WorldLoader.DataLoadContext worldLoader; //Ketting - not final
+   public org.bukkit.craftbukkit.v1_20_R1.CraftServer server;
+   public joptsimple.OptionSet options;
+   public org.bukkit.command.ConsoleCommandSender console;
+   public jline.console.ConsoleReader reader;
+   public static int currentTick = (int) (System.currentTimeMillis() / 50);
+   public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
+   public int autosavePeriod;
+   public Commands vanillaCommandDispatcher;
+   private boolean forceTicks;
+   // CraftBukkit end
+   // Spigot start
+   public static final int TPS = 20;
+   public static final int TICK_TIME = 1000000000 / TPS;
+   private static final int SAMPLE_INTERVAL = 100;
+   public final double[] recentTps = new double[ 3 ];
+   // Spigot end
 
    public static <S extends MinecraftServer> S m_129872_(Function<Thread, S> p_129873_) {
       AtomicReference<S> atomicreference = new AtomicReference<>();
-      Thread thread = new Thread(() -> {
+      Thread thread = new Thread(net.minecraftforge.fml.util.thread.SidedThreadGroups.SERVER, () -> {
          atomicreference.get().m_130011_();
+         Runtime.getRuntime().halt(0); //Ketting - fix server hanging on exit
       }, "Server thread");
       thread.setUncaughtExceptionHandler((p_177909_, p_177910_) -> {
          f_129750_.error("Uncaught exception in server thread", p_177910_);
@@ -267,7 +_,7 @@
       super("Server");
       this.f_244176_ = p_236726_.f_244542_();
       this.f_129749_ = p_236726_.f_206895_();
-      if (!this.f_244176_.m_247579_().m_175515_(Registries.f_256862_).m_142003_(LevelStem.f_63971_)) {
+      if (false && !this.f_244176_.m_247579_().m_175515_(Registries.f_256862_).m_142003_(LevelStem.f_63971_)) { // CraftBukkit - initialised later
          throw new IllegalStateException("Missing Overworld dimension data");
       } else {
          this.f_129747_ = p_236727_;
@@ -278,7 +_,7 @@
             p_236729_.f_214336_().m_143974_(this);
          }
 
-         this.f_129755_ = new ServerConnectionListener(this);
+         // this.connection = new ServerConnectionListener(this); // Spigot
          this.f_129756_ = p_236730_;
          this.f_129744_ = p_236724_;
          this.f_129745_ = p_236724_.m_78301_();
@@ -289,6 +_,31 @@
          this.f_129725_ = p_236723_;
          this.f_129738_ = Util.m_183991_();
       }
+      // CraftBukkit start
+      this.vanillaCommandDispatcher = p_236726_.f_206893_().f_206847_; // CraftBukkit
+      // Try to see if we're actually running in a terminal, disable jline if not
+      if (System.console() == null && System.getProperty("jline.terminal") == null) {
+         System.setProperty("jline.terminal", "jline.UnsupportedTerminal");
+         org.bukkit.craftbukkit.Main.useJline = false;
+      }
+
+      try {
+         reader = new jline.console.ConsoleReader(System.in, System.out);
+         reader.setExpandEvents(false); // Avoid parsing exceptions for uncommonly used event designators
+      } catch (Throwable e) {
+         try {
+            // Try again with jline disabled for Windows users without C++ 2008 Redistributable
+            System.setProperty("jline.terminal", "jline.UnsupportedTerminal");
+            System.setProperty("user.language", "en");
+            org.bukkit.craftbukkit.Main.useJline = false;
+            reader = new jline.console.ConsoleReader(System.in, System.out);
+            reader.setExpandEvents(false);
+         } catch (IOException ex) {
+            f_129750_.warn((String) null, ex);
+         }
+      }
+      Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.v1_20_R1.util.ServerShutdownThread(this));
+      // CraftBukkit end
    }
 
    private void m_129841_(DimensionDataStorage p_129842_) {
@@ -297,7 +_,19 @@
 
    protected abstract boolean m_7038_() throws IOException;
 
-   protected void m_130006_() {
+   //Ketting start
+   private String loadLevel$worldName;
+   protected void prepareLoadLevel(String worldName) {
+      this.loadLevel$worldName = worldName;
+   }
+
+   public void setBukkitOptions(joptsimple.OptionSet options, WorldLoader.DataLoadContext dataLoadContext) {
+      this.options = options;
+      this.worldLoader = dataLoadContext;
+   }
+   //Ketting end
+
+   protected void m_130006_() { // CraftBukkit
       if (!JvmProfiler.f_185340_.m_183608_()) {
       }
 
@@ -308,6 +_,13 @@
       this.m_129815_(chunkprogresslistener);
       this.m_7044_();
       this.m_129940_(chunkprogresslistener);
+
+      //Ketting start
+      this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+      this.server.getPluginManager().callEvent(new org.bukkit.event.server.ServerLoadEvent(org.bukkit.event.server.ServerLoadEvent.LoadType.STARTUP));
+      this.f_129755_.acceptConnections();
+      //Ketting end
+
       if (profiledduration != null) {
          profiledduration.m_185413_();
       }
@@ -334,12 +_,16 @@
       long j = BiomeManager.m_47877_(i);
       List<CustomSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(serverleveldata));
       LevelStem levelstem = registry.m_6246_(LevelStem.f_63971_);
-      ServerLevel serverlevel = new ServerLevel(this, this.f_129738_, this.f_129744_, serverleveldata, Level.f_46428_, levelstem, p_129816_, flag, j, list, true, (RandomSequences)null);
+      ServerLevel serverlevel = new ServerLevel(this, this.f_129738_, this.f_129744_, serverleveldata, Level.f_46428_, levelstem, p_129816_, flag, j, list, true, (RandomSequences)null, org.bukkit.World.Environment.NORMAL, this.server.getGenerator(loadLevel$worldName), this.server.getBiomeProvider(loadLevel$worldName));
+      org.kettingpowered.ketting.core.Ketting.getInstance().getDimensionRegistry().get().register(this.m_206579_().m_175515_(Registries.f_256862_)); //Ketting
       this.f_129762_.put(Level.f_46428_, serverlevel);
       DimensionDataStorage dimensiondatastorage = serverlevel.m_8895_();
+      this.server.scoreboardManager = new org.bukkit.craftbukkit.v1_20_R1.scoreboard.CraftScoreboardManager(this, serverlevel.m_6188_()); // CraftBukkit
       this.m_129841_(dimensiondatastorage);
       this.f_129732_ = new CommandStorage(dimensiondatastorage);
       WorldBorder worldborder = serverlevel.m_6857_();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Load(f_129762_.get(Level.f_46428_)));
+      this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(f_129762_.get(Level.f_46428_).getWorld())); // CraftBukkit - SPIGOT-5569: Call WorldInitEvent before any chunks are generated
       if (!serverleveldata.m_6535_()) {
          try {
             m_177896_(serverlevel, serverleveldata, worldoptions.m_245100_(), flag);
@@ -373,9 +_,49 @@
          if (resourcekey != LevelStem.f_63971_) {
             ResourceKey<Level> resourcekey1 = ResourceKey.m_135785_(Registries.f_256858_, resourcekey.m_135782_());
             DerivedLevelData derivedleveldata = new DerivedLevelData(this.f_129749_, serverleveldata);
-            ServerLevel serverlevel1 = new ServerLevel(this, this.f_129738_, this.f_129744_, derivedleveldata, resourcekey1, entry.getValue(), p_129816_, flag, j, ImmutableList.of(), false, randomsequences);
+            derivedleveldata.setTypeKey(resourcekey); //Ketting
+
+            int dimension = 0;
+
+            if (resourcekey == LevelStem.f_63972_) {
+               if (m_7079_()) {
+                  dimension = -1;
+               } else {
+                  continue;
+               }
+            } else if (resourcekey == LevelStem.f_63973_) {
+               if (server.getAllowEnd()) {
+                  dimension = 1;
+               } else {
+                  continue;
+               }
+            } else {
+               dimension = -999;
+            }
+
+            String worldType = (dimension == -999) ? resourcekey.m_135782_().m_135827_() + "_" + resourcekey.m_135782_().m_135815_() : org.bukkit.World.Environment.getEnvironment(dimension).toString().toLowerCase();
+            String name = loadLevel$worldName + "_" + worldType;
+
+            org.bukkit.generator.ChunkGenerator gen = this.server.getGenerator(name);
+            org.bukkit.generator.BiomeProvider biomeProvider = this.server.getBiomeProvider(name);
+
+            LevelStorageSource.LevelStorageAccess source;
+            try {
+               source = LevelStorageSource.m_78242_(server.getWorldContainer().toPath()).validateAndCreateAccess(name, resourcekey);
+            } catch (IOException | net.minecraft.world.level.validation.ContentValidationException e) {
+               throw new RuntimeException(e);
+            }
+
+            ServerLevel serverlevel1 = new ServerLevel(this, this.f_129738_, source, derivedleveldata, resourcekey1, entry.getValue(), p_129816_, flag, j, ImmutableList.of(), false, randomsequences, (org.bukkit.World.Environment) org.kettingpowered.ketting.core.Ketting.getInstance().getDimensionRegistry().get().getKey(resourcekey), gen, biomeProvider);
+
+            if (serverlevel1.generator != null) {
+               serverlevel1.getWorld().getPopulators().addAll(serverlevel1.generator.getDefaultPopulators(serverlevel1.getWorld()));
+            }
+
             worldborder.m_61929_(new BorderChangeListener.DelegateBorderChangeListener(serverlevel1.m_6857_()));
             this.f_129762_.put(resourcekey1, serverlevel1);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Load(f_129762_.get(resourcekey)));
+            this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(serverlevel1.getWorld())); // CraftBukkit - SPIGOT-5569: Call WorldInitEvent before any chunks are generated
          }
       }
 
@@ -387,7 +_,23 @@
          p_177898_.m_7250_(BlockPos.f_121853_.m_6630_(80), 0.0F);
       } else {
          ServerChunkCache serverchunkcache = p_177897_.m_7726_();
+         if (net.minecraftforge.event.ForgeEventFactory.onCreateWorldSpawn(p_177897_, p_177898_)) return;
          ChunkPos chunkpos = new ChunkPos(serverchunkcache.m_214994_().m_224579_().m_183230_());
+         // CraftBukkit start
+         if (p_177897_.generator != null) {
+            java.util.Random rand = new java.util.Random(p_177897_.m_7328_());
+            org.bukkit.Location spawn = p_177897_.generator.getFixedSpawnLocation(p_177897_.getWorld(), rand);
+
+            if (spawn != null) {
+               if (spawn.getWorld() != p_177897_.getWorld()) {
+                  throw new IllegalStateException("Cannot set spawn point for " + p_177898_.m_5462_() + " to be in another world (" + spawn.getWorld().getName() + ")");
+               } else {
+                  p_177898_.m_7250_(new BlockPos(spawn.getBlockX(), spawn.getBlockY(), spawn.getBlockZ()), spawn.getYaw());
+                  return;
+               }
+            }
+         }
+         // CraftBukkit end
          int i = serverchunkcache.m_8481_().m_142051_(p_177897_);
          if (i < p_177897_.m_141937_()) {
             BlockPos blockpos = chunkpos.m_45615_();
@@ -442,8 +_,23 @@
       serverleveldata.m_5458_(GameType.SPECTATOR);
    }
 
+   //Ketting Start
+   private @Nullable ServerLevel worldserver = null;
+   public void prepareLevels(ChunkProgressListener p_129941_, ServerLevel worldserver){
+      this.worldserver = worldserver;
+      m_129940_(p_129941_);
+   }
+   //Ketting End
+
    private void m_129940_(ChunkProgressListener p_129941_) {
+      this.forceTicks = true; // CraftBukkit
       ServerLevel serverlevel = this.m_129783_();
+      //Ketting Start
+      if (worldserver != null) {
+         serverlevel = worldserver;
+         worldserver = null;
+      }
+      //Ketting End
       f_129750_.info("Preparing start region for dimension {}", (Object)serverlevel.m_46472_().m_135782_());
       BlockPos blockpos = serverlevel.m_220360_();
       p_129941_.m_7647_(new ChunkPos(blockpos));
@@ -451,13 +_,11 @@
       this.f_129726_ = Util.m_137550_();
       serverchunkcache.m_8387_(TicketType.f_9442_, new ChunkPos(blockpos), 11, Unit.INSTANCE);
 
-      while(serverchunkcache.m_8427_() != 441) {
-         this.f_129726_ = Util.m_137550_() + 10L;
-         this.m_130012_();
+      while (serverchunkcache.m_8427_() < 441) {
+         this.executeModerately();
       }
 
-      this.f_129726_ = Util.m_137550_() + 10L;
-      this.m_130012_();
+      this.executeModerately();
 
       for(ServerLevel serverlevel1 : this.f_129762_.values()) {
          ForcedChunksSavedData forcedchunkssaveddata = serverlevel1.m_8895_().m_164858_(ForcedChunksSavedData::m_151483_, "chunks");
@@ -469,13 +_,19 @@
                ChunkPos chunkpos = new ChunkPos(i);
                serverlevel1.m_7726_().m_6692_(chunkpos, true);
             }
+            net.minecraftforge.common.world.ForgeChunkManager.reinstatePersistentChunks(serverlevel1, forcedchunkssaveddata);
          }
+
+         // CraftBukkit start
+         serverlevel1.f_143244_.m_157506_(); // SPIGOT-6526: Load pending entities so they are available to the API
+         this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldLoadEvent(serverlevel1.getWorld()));
+         // CraftBukkit end
       }
 
-      this.f_129726_ = Util.m_137550_() + 10L;
-      this.m_130012_();
+      this.executeModerately();
       p_129941_.m_7646_();
       this.m_129962_();
+      this.forceTicks = false; // CraftBukkit
    }
 
    public GameType m_130008_() {
@@ -504,11 +_,15 @@
          flag = true;
       }
 
-      ServerLevel serverlevel2 = this.m_129783_();
-      ServerLevelData serverleveldata = this.f_129749_.m_5996_();
-      serverleveldata.m_7831_(serverlevel2.m_6857_().m_61970_());
-      this.f_129749_.m_5917_(this.m_129901_().m_136307_());
-      this.f_129744_.m_78290_(this.m_206579_(), this.f_129749_, this.m_6846_().m_6960_());
+      // CraftBukkit start - moved to WorldServer.save
+      /*
+      ServerLevel serverlevel2 = this.overworld();
+      ServerLevelData serverleveldata = this.worldData.overworldData();
+      serverleveldata.setWorldBorder(serverlevel2.getWorldBorder().createSettings());
+      this.worldData.setCustomBossEvents(this.getCustomBossEvents().save());
+      this.storageSource.saveDataTag(this.registryAccess(), this.worldData, this.getPlayerList().getSingleplayerData());
+      */
+      // CraftBukkit end
       if (p_129887_) {
          for(ServerLevel serverlevel1 : this.m_129785_()) {
             f_129750_.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", (Object)serverlevel1.m_7726_().f_8325_.m_182285_());
@@ -537,12 +_,33 @@
       this.m_7041_();
    }
 
+   // CraftBukkit start
+   private boolean hasStopped = false;
+   private final Object stopLock = new Object();
+   public final boolean hasStopped() {
+      synchronized (stopLock) {
+         return hasStopped;
+      }
+   }
+   // CraftBukkit end
+
    public void m_7041_() {
+      // CraftBukkit start - prevent double stopping on multiple threads
+      synchronized(stopLock) {
+         if (hasStopped) return;
+         hasStopped = true;
+      }
+      // CraftBukkit end
       if (this.f_177872_.m_142763_()) {
          this.m_236737_();
       }
 
       f_129750_.info("Stopping server");
+      // CraftBukkit start
+      if (this.server != null) {
+         this.server.disablePlugins();
+      }
+      // CraftBukkit end
       if (this.m_129919_() != null) {
          this.m_129919_().m_9718_();
       }
@@ -552,6 +_,7 @@
          f_129750_.info("Saving players");
          this.f_129763_.m_11302_();
          this.f_129763_.m_11313_();
+         try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
       }
 
       f_129750_.info("Saving worlds");
@@ -582,6 +_,7 @@
       for(ServerLevel serverlevel2 : this.m_129785_()) {
          if (serverlevel2 != null) {
             try {
+               net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.level.LevelEvent.Unload(serverlevel2));
                serverlevel2.close();
             } catch (IOException ioexception1) {
                f_129750_.error("Exception closing the level", (Throwable)ioexception1);
@@ -598,6 +_,12 @@
          f_129750_.error("Failed to unlock level {}", this.f_129744_.m_78277_(), ioexception);
       }
 
+      // Spigot start
+      if (org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) {
+         f_129750_.info("Saving usercache.json");
+         this.m_129927_().m_11006_();
+      }
+      // Spigot end
    }
 
    public String m_130009_() {
@@ -624,30 +_,56 @@
 
    }
 
+   // Spigot Start
+   private static double calcTps(double avg, double exp, double tps)
+   {
+      return ( avg * exp ) + ( tps * ( 1 - exp ) );
+   }
+   // Spigot End
+
    protected void m_130011_() {
       try {
          if (!this.m_7038_()) {
             throw new IllegalStateException("Failed to initialize server");
          }
 
+         net.minecraftforge.server.ServerLifecycleHooks.handleServerStarted(this);
          this.f_129726_ = Util.m_137550_();
-         this.f_271173_ = this.m_272273_().orElse((ServerStatus.Favicon)null);
+         this.f_271173_ = this.m_272273_().orElse((ServerStatus.Favicon) null);
          this.f_129757_ = this.m_271988_();
-
-         while(this.f_129764_) {
-            long i = Util.m_137550_() - this.f_129726_;
-            if (i > 2000L && this.f_129726_ - this.f_129718_ >= 15000L) {
+         resetStatusCache(f_129757_);
+
+         // Spigot start
+         Arrays.fill( recentTps, 20 );
+         long curTime, tickSection = Util.m_137550_(), tickCount = 1;
+         while (this.f_129764_) {
+            long i = (curTime = Util.m_137550_()) - this.f_129726_;
+
+            if (i > 5000L && this.f_129726_ - this.f_129718_ >= 30000L) { // CraftBukkit
                long j = i / 50L;
-               f_129750_.warn("Can't keep up! Is the server overloaded? Running {}ms or {} ticks behind", i, j);
+
+               if (server.getWarnOnOverload())
+                  f_129750_.warn("Can't keep up! Is the server overloaded? Running {}ms or {} ticks behind", i, j);
                this.f_129726_ += j * 50L;
                this.f_129718_ = this.f_129726_;
             }
 
+            if ( tickCount++ % SAMPLE_INTERVAL == 0 )
+            {
+               double currentTps = 1E3 / ( curTime - tickSection ) * SAMPLE_INTERVAL;
+               recentTps[0] = calcTps( recentTps[0], 0.92, currentTps ); // 1/exp(5sec/1min)
+               recentTps[1] = calcTps( recentTps[1], 0.9835, currentTps ); // 1/exp(5sec/5min)
+               recentTps[2] = calcTps( recentTps[2], 0.9945, currentTps ); // 1/exp(5sec/15min)
+               tickSection = curTime;
+            }
+            // Spigot end
+
             if (this.f_177877_) {
                this.f_177877_ = false;
                this.f_177876_ = new MinecraftServer.TimeProfiler(Util.m_137569_(), this.f_129766_);
             }
 
+            MinecraftServer.currentTick = (int) (System.currentTimeMillis() / 50); // CraftBukkit
             this.f_129726_ += 50L;
             this.m_177945_();
             this.f_129754_.m_6180_("tick");
@@ -661,8 +_,18 @@
             this.f_129717_ = true;
             JvmProfiler.f_185340_.m_183597_(this.f_129737_);
          }
+         net.minecraftforge.server.ServerLifecycleHooks.handleServerStopping(this);
+         net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
+      } catch (org.spongepowered.asm.mixin.throwables.MixinError mixin) {
+         org.kettingpowered.ketting.core.injectprotect.InjectProtect.mixinInjectCaught(null, mixin);
       } catch (Throwable throwable1) {
          f_129750_.error("Encountered an unexpected exception", throwable1);
+         // Spigot Start
+         if ( throwable1.getCause() != null )
+         {
+            MinecraftServer.f_129750_.error( "\tCause of unexpected exception was", throwable1.getCause() );
+         }
+         // Spigot End
          CrashReport crashreport = m_206568_(throwable1);
          this.m_177935_(crashreport.m_178626_());
          File file1 = new File(new File(this.m_6237_(), "crash-reports"), "crash-" + Util.m_241986_() + "-server.txt");
@@ -672,6 +_,7 @@
             f_129750_.error("We were unable to save this crash report to disk.");
          }
 
+         net.minecraftforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
          this.m_7268_(crashreport);
       } finally {
          try {
@@ -684,6 +_,15 @@
                this.f_236721_.f_214336_().m_196559_();
             }
 
+            net.minecraftforge.server.ServerLifecycleHooks.handleServerStopped(this);
+
+            org.spigotmc.WatchdogThread.doStop(); // Spigot
+            // CraftBukkit start - Restore terminal to original settings
+            try {
+               reader.getTerminal().restore();
+            } catch (Exception ignored) {
+            }
+            // CraftBukkit end
             this.m_6988_();
          }
 
@@ -714,7 +_,7 @@
    }
 
    private boolean m_129960_() {
-      return this.m_18767_() || Util.m_137550_() < (this.f_129728_ ? this.f_129727_ : this.f_129726_);
+      return this.forceTicks || this.m_18767_() || Util.m_137550_() < (this.f_129728_ ? this.f_129727_ : this.f_129726_); // CraftBukkit
    }
 
    protected void m_130012_() {
@@ -724,6 +_,18 @@
       });
    }
 
+   private void executeModerately() {
+      this.m_18699_();
+      this.drainQueuedTasks();
+      java.util.concurrent.locks.LockSupport.parkNanos("executing tasks", 1000L);
+   }
+
+   public void drainQueuedTasks() {
+      while (!processQueue.isEmpty()) {
+         processQueue.remove().run();
+      }
+   }
+
    public TickTask m_6681_(Runnable p_129852_) {
       return new TickTask(this.f_129766_, p_129852_);
    }
@@ -797,20 +_,26 @@
    }
 
    public void m_5705_(BooleanSupplier p_129871_) {
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.serverTickTimer.startTiming(); // Spigot
       long i = Util.m_137569_();
+
+      net.minecraftforge.event.ForgeEventFactory.onPreServerTick(p_129871_, this);
       ++this.f_129766_;
       this.m_5703_(p_129871_);
       if (i - this.f_129724_ >= 5000000000L) {
          this.f_129724_ = i;
          this.f_129757_ = this.m_271988_();
+         resetStatusCache(f_129757_);
       }
 
-      if (this.f_129766_ % 6000 == 0) {
+      if (autosavePeriod > 0 && this.f_129766_ % autosavePeriod == 0) { // CraftBukkit
+         org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.worldSaveTimer.startTiming(); // Spigot
          f_129750_.debug("Autosave started");
          this.f_129754_.m_6180_("save");
          this.m_195514_(true, false, false);
          this.f_129754_.m_7238_();
          f_129750_.debug("Autosave finished");
+         org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.worldSaveTimer.stopTiming(); // Spigot
       }
 
       this.f_129754_.m_6180_("tallying");
@@ -819,11 +_,25 @@
       long k = Util.m_137569_();
       this.f_129735_.m_13755_(k - i);
       this.f_129754_.m_7238_();
+      net.minecraftforge.event.ForgeEventFactory.onPostServerTick(p_129871_, this);
+      org.spigotmc.WatchdogThread.tick(); // Spigot
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.serverTickTimer.stopTiming(); // Spigot
+      org.spigotmc.CustomTimingsHandler.tick(); // Spigot
+   }
+
+   private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+   private String cachedServerStatus; // FORGE: cache the server status json in case a client spams requests
+   private void resetStatusCache(ServerStatus status) {
+      this.cachedServerStatus = GSON.toJson(ServerStatus.f_271163_.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status)
+              .result().orElseThrow());
+   }
+   public String getStatusJson() {
+      return cachedServerStatus;
    }
 
    private ServerStatus m_271988_() {
       ServerStatus.Players serverstatus$players = this.m_271961_();
-      return new ServerStatus(Component.m_130674_(this.f_129709_), Optional.of(serverstatus$players), Optional.of(ServerStatus.Version.m_272202_()), Optional.ofNullable(this.f_271173_), this.m_214005_());
+      return new ServerStatus(Component.m_130674_(this.f_129709_), Optional.of(serverstatus$players), Optional.of(ServerStatus.Version.m_272202_()), Optional.ofNullable(this.f_271173_), this.m_214005_(), Optional.of(new net.minecraftforge.network.ServerStatusPing()));
    }
 
    private ServerStatus.Players m_271961_() {
@@ -847,47 +_,92 @@
    }
 
    public void m_5703_(BooleanSupplier p_129954_) {
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.schedulerTimer.startTiming(); // Spigot
+      this.server.getScheduler().mainThreadHeartbeat(this.f_129766_); // CraftBukkit
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.schedulerTimer.stopTiming(); // Spigot
       this.f_129754_.m_6180_("commandFunctions");
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.commandFunctionsTimer.startTiming(); // Spigot
       this.m_129890_().m_136128_();
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.commandFunctionsTimer.stopTiming(); // Spigot
       this.f_129754_.m_6182_("levels");
 
-      for(ServerLevel serverlevel : this.m_129785_()) {
+      // CraftBukkit start
+      // Run tasks that are waiting on processing
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.processQueueTimer.startTiming(); // Spigot
+      while (!processQueue.isEmpty()) {
+         processQueue.remove().run();
+      }
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.processQueueTimer.stopTiming(); // Spigot
+
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.timeUpdateTimer.startTiming(); // Spigot
+      // Send time updates to everyone, it will get the right time from the world the player is in.
+      if (this.f_129766_ % 20 == 0) {
+         for (int i = 0; i < this.m_6846_().f_11196_.size(); ++i) {
+            ServerPlayer serverplayer = (ServerPlayer) this.m_6846_().f_11196_.get(i);
+            serverplayer.f_8906_.m_9829_(new ClientboundSetTimePacket(serverplayer.m_9236_().m_46467_(), serverplayer.getPlayerTime(), serverplayer.m_9236_().m_46469_().m_46207_(GameRules.f_46140_))); // Add support for per player time
+         }
+      }
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.timeUpdateTimer.stopTiming(); // Spigot
+
+      for(ServerLevel serverlevel : this.getWorldArray()) {
+         long tickStart = Util.m_137569_();
          this.f_129754_.m_6521_(() -> {
             return serverlevel + " " + serverlevel.m_46472_().m_135782_();
          });
-         if (this.f_129766_ % 20 == 0) {
-            this.f_129754_.m_6180_("timeSync");
-            this.m_276346_(serverlevel);
-            this.f_129754_.m_7238_();
+         /* Drop global time updates
+         if (this.tickCount % 20 == 0) {
+            this.profiler.push("timeSync");
+            this.synchronizeTime(serverlevel);
+            this.profiler.pop();
          }
+         // CraftBukkit end */
 
          this.f_129754_.m_6180_("tick");
+         net.minecraftforge.event.ForgeEventFactory.onPreLevelTick(serverlevel, p_129954_);
 
          try {
+            serverlevel.timings.doTick.startTiming(); // Spigot
             serverlevel.m_8793_(p_129954_);
+            serverlevel.timings.doTick.stopTiming(); // Spigot
          } catch (Throwable throwable) {
-            CrashReport crashreport = CrashReport.m_127521_(throwable, "Exception ticking world");
+            // Spigot Start
+            CrashReport crashreport;
+            try {
+               crashreport = CrashReport.m_127521_(throwable, "Exception ticking world");
+            } catch (Throwable t) {
+               throw new RuntimeException("Error generating crash report", t);
+            }
+            // Spigot End
+
             serverlevel.m_6026_(crashreport);
             throw new ReportedException(crashreport);
          }
+         net.minecraftforge.event.ForgeEventFactory.onPostLevelTick(serverlevel, p_129954_);
 
          this.f_129754_.m_7238_();
          this.f_129754_.m_7238_();
+         perWorldTickTimes.computeIfAbsent(serverlevel.m_46472_(), k -> new long[100])[this.f_129766_ % 100] = Util.m_137569_() - tickStart;
       }
 
       this.f_129754_.m_6182_("connection");
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.connectionTimer.startTiming(); // Spigot
       this.m_129919_().m_9721_();
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.connectionTimer.stopTiming(); // Spigot
       this.f_129754_.m_6182_("players");
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.playerListTimer.startTiming(); // Spigot
       this.f_129763_.m_11288_();
-      if (SharedConstants.f_136183_) {
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.playerListTimer.stopTiming(); // Spigot
+      if (net.minecraftforge.gametest.ForgeGameTestHooks.isGametestEnabled()) {
          GameTestTicker.f_177648_.m_127790_();
       }
 
       this.f_129754_.m_6182_("server gui refresh");
 
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.tickablesTimer.startTiming(); // Spigot
       for(int i = 0; i < this.f_129752_.size(); ++i) {
          this.f_129752_.get(i).run();
       }
+      org.bukkit.craftbukkit.v1_20_R1.SpigotTimings.tickablesTimer.stopTiming(); // Spigot
 
       this.f_129754_.m_7238_();
    }
@@ -935,6 +_,22 @@
       return this.f_129762_.get(p_129881_);
    }
 
+   // CraftBukkit start
+   public void addLevel(ServerLevel level) {
+      Map<ResourceKey<Level>, ServerLevel> oldLevels = this.f_129762_;
+      Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+      newLevels.put(level.m_46472_(), level);
+      this.f_129762_ = Collections.unmodifiableMap(newLevels);
+   }
+
+   public void removeLevel(ServerLevel level) {
+      Map<ResourceKey<Level>, ServerLevel> oldLevels = this.f_129762_;
+      Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+      newLevels.remove(level.m_46472_());
+      this.f_129762_ = Collections.unmodifiableMap(newLevels);
+   }
+   // CraftBukkit end
+
    public Set<ResourceKey<Level>> m_129784_() {
       return this.f_129762_.keySet();
    }
@@ -961,7 +_,7 @@
 
    @DontObfuscate
    public String getServerModName() {
-      return "vanilla";
+      return org.kettingpowered.ketting.internal.KettingConstants.NAME; //net.minecraftforge.internal.BrandingControl.getServerBranding(); //Ketting
    }
 
    public SystemReport m_177935_(SystemReport p_177936_) {
@@ -1164,7 +_,7 @@
 
    @Nullable
    public ServerConnectionListener m_129919_() {
-      return this.f_129755_;
+      return this.f_129755_ == null ? this.f_129755_ = new ServerConnectionListener(this) : this.f_129755_; // Spigot
    }
 
    public boolean m_129920_() {
@@ -1303,6 +_,13 @@
       }).thenAcceptAsync((p_248090_) -> {
          this.f_129740_.close();
          this.f_129740_ = p_248090_;
+         //Ketting start
+         this.server.doSync(p_248090_.f_206585_().f_206847_);
+         if (this instanceof net.minecraft.server.dedicated.DedicatedServer ded && ded.betterGui != null)
+            ded.betterGui.reload();
+         org.kettingpowered.ketting.core.Ketting.getInstance().reload();
+         org.kettingpowered.ketting.metrics.MetricManager.reload();
+         //Ketting end
          this.f_129730_.m_10509_(p_129862_);
          WorldDataConfiguration worlddataconfiguration = new WorldDataConfiguration(m_129817_(this.f_129730_), this.f_129749_.m_247623_());
          this.f_129749_.m_245843_(worlddataconfiguration);
@@ -1311,6 +_,7 @@
          this.m_6846_().m_11315_();
          this.f_129734_.m_136120_(this.f_129740_.f_206585_.m_206860_());
          this.f_236720_.m_230370_(this.f_129740_.f_206584_);
+         this.m_6846_().m_11314_().forEach(this.m_6846_()::m_11289_); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
       }, this);
       if (this.m_18695_()) {
          this.m_18701_(completablefuture::isDone);
@@ -1320,10 +_,13 @@
    }
 
    public static WorldDataConfiguration m_246048_(PackRepository p_248681_, DataPackConfig p_248920_, boolean p_249869_, FeatureFlagSet p_251243_) {
+      net.minecraftforge.resource.ResourcePackLoader.loadResourcePacks(p_248681_, net.minecraftforge.server.ServerLifecycleHooks::buildPackFinder);
       p_248681_.m_10506_();
+      DataPackConfig.f_45842_.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
+      p_248920_.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
       if (p_249869_) {
-         p_248681_.m_10509_(Collections.singleton("vanilla"));
-         return WorldDataConfiguration.f_244649_;
+         p_248681_.m_10509_(net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla());
+         return new WorldDataConfiguration(new DataPackConfig(net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla(), ImmutableList.of()), FeatureFlags.f_244332_);
       } else {
          Set<String> set = Sets.newLinkedHashSet();
 
@@ -1481,6 +_,31 @@
 
    public abstract boolean m_7779_(GameProfile p_129840_);
 
+   private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+   @Nullable
+   public long[] getTickTime(ResourceKey<Level> dim) {
+      return perWorldTickTimes.get(dim);
+   }
+
+   @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+   public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+      return this.f_129762_;
+   }
+   private int worldArrayMarker = 0;
+   private int worldArrayLast = -1;
+   private ServerLevel[] worldArray;
+   @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+   public synchronized void markWorldsDirty() {
+      worldArrayMarker++;
+   }
+   private ServerLevel[] getWorldArray() {
+      if (worldArrayMarker == worldArrayLast && worldArray != null)
+         return worldArray;
+      worldArray = this.f_129762_.values().stream().toArray(x -> new ServerLevel[x]);
+      worldArrayLast = worldArrayMarker;
+      return worldArray;
+   }
+
    public void m_142116_(Path p_177911_) throws IOException {
    }
 
@@ -1583,6 +_,22 @@
       }
    }
 
+   // CraftBukkit start
+   @Override
+   public boolean m_18695_() {
+      return super.m_18695_() || this.m_129918_(); // CraftBukkit - MC-142590
+   }
+
+   public boolean isDebugging() {
+      return false;
+   }
+
+   @Deprecated
+   public static MinecraftServer getServer() {
+      return (org.bukkit.Bukkit.getServer() instanceof org.bukkit.craftbukkit.v1_20_R1.CraftServer) ? ((org.bukkit.craftbukkit.v1_20_R1.CraftServer) org.bukkit.Bukkit.getServer()).getServer() : null;
+   }
+   // CraftBukkit end
+
    private void m_177945_() {
       if (this.f_177875_) {
          this.f_177872_ = ActiveMetricsRecorder.m_146132_(new ServerMetricsSamplersProvider(Util.f_137440_, this.m_6982_()), Util.f_137440_, Util.m_183992_(), new MetricsPersister("server"), this.f_177873_, (p_212927_) -> {
@@ -1646,6 +_,10 @@
       return this.f_129749_;
    }
 
+   public MinecraftServer.ReloadableResources getServerResources() {
+       return f_129740_;
+   }
+
    public RegistryAccess.Frozen m_206579_() {
       return this.f_244176_.m_247579_();
    }
@@ -1706,6 +_,11 @@
       }
 
    }
+
+   // CraftBukkit start
+   public final java.util.concurrent.ExecutorService chatExecutor = java.util.concurrent.Executors.newCachedThreadPool(
+           new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Chat Thread - #%d").build());
+   // CraftBukkit end
 
    public ChatDecorator m_236742_() {
       return ChatDecorator.f_236947_;
