--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -77,7 +_,7 @@
 
 public abstract class ChunkGenerator {
    public static final Codec<ChunkGenerator> f_62136_ = BuiltInRegistries.f_256914_.m_194605_().dispatchStable(ChunkGenerator::m_6909_, Function.identity());
-   public BiomeSource f_62137_;
+   protected BiomeSource f_62137_; //Ketting - not final
    private final Supplier<List<FeatureSorter.StepFeatureData>> f_223020_;
    public final Function<Holder<Biome>, BiomeGenerationSettings> f_223021_;
 
@@ -97,10 +_,16 @@
       });
    }
 
+   //Ketting start - make biomeSource configurable
+   public void setBiomeSource(BiomeSource biomeSource) {
+       this.f_62137_ = biomeSource;
+   }
+   //Ketting end
+
    protected abstract Codec<? extends ChunkGenerator> m_6909_();
 
-   public ChunkGeneratorStructureState m_255169_(HolderLookup<StructureSet> p_256405_, RandomState p_256101_, long p_256018_) {
-      return ChunkGeneratorStructureState.m_255081_(p_256101_, p_256018_, this.f_62137_, p_256405_);
+   public ChunkGeneratorStructureState createState(HolderLookup<StructureSet> p_256405_, RandomState p_256101_, long p_256018_, org.spigotmc.SpigotWorldConfig conf) { // Spigot
+      return ChunkGeneratorStructureState.createForNormal(p_256101_, p_256018_, this.f_62137_, p_256405_, conf);
    }
 
    public Optional<ResourceKey<Codec<? extends ChunkGenerator>>> m_187743_() {
@@ -265,7 +_,7 @@
       }
    }
 
-   public void m_213609_(WorldGenLevel p_223087_, ChunkAccess p_223088_, StructureManager p_223089_) {
+   public void addVanillaDecorations(WorldGenLevel p_223087_, ChunkAccess p_223088_, StructureManager p_223089_) { // CraftBukkit
       ChunkPos chunkpos = p_223088_.m_7697_();
       if (!SharedConstants.m_183707_(chunkpos)) {
          SectionPos sectionpos = SectionPos.m_123196_(chunkpos, p_223087_.m_151560_());
@@ -365,6 +_,33 @@
       }
    }
 
+   // CraftBukkit start
+   public void m_213609_(WorldGenLevel p_223087_, ChunkAccess p_223088_, StructureManager p_223089_) {
+      applyBiomeDecoration(p_223087_, p_223088_, p_223089_, true);
+   }
+
+   public void applyBiomeDecoration(WorldGenLevel generatoraccessseed, ChunkAccess ichunkaccess, StructureManager structuremanager, boolean vanilla) {
+      if (vanilla) {
+         addVanillaDecorations(generatoraccessseed, ichunkaccess, structuremanager);
+      }
+
+      org.bukkit.World world = generatoraccessseed.getMinecraftWorld().getWorld();
+      // only call when a populator is present (prevents unnecessary entity conversion)
+      if (!world.getPopulators().isEmpty()) {
+         org.bukkit.craftbukkit.v1_20_R2.generator.CraftLimitedRegion limitedRegion = new org.bukkit.craftbukkit.v1_20_R2.generator.CraftLimitedRegion(generatoraccessseed, ichunkaccess.m_7697_());
+         int x = ichunkaccess.m_7697_().f_45578_;
+         int z = ichunkaccess.m_7697_().f_45579_;
+         for (org.bukkit.generator.BlockPopulator populator : world.getPopulators()) {
+            WorldgenRandom seededrandom = new WorldgenRandom(new net.minecraft.world.level.levelgen.LegacyRandomSource(generatoraccessseed.m_7328_()));
+            seededrandom.m_64690_(generatoraccessseed.m_7328_(), x, z);
+            populator.populate(world, new org.bukkit.craftbukkit.v1_20_R2.util.RandomSourceWrapper.RandomWrapper(seededrandom), x, z, limitedRegion);
+         }
+         limitedRegion.saveEntities();
+         limitedRegion.breakLink();
+      }
+   }
+   // CraftBukkit end
+
    private static BoundingBox m_187717_(ChunkAccess p_187718_) {
       ChunkPos chunkpos = p_187718_.m_7697_();
       int i = chunkpos.m_45604_();
@@ -480,6 +_,14 @@
       Predicate<Holder<Biome>> predicate = holderset::m_203333_;
       StructureStart structurestart = structure.m_226596_(p_223107_, this, this.f_62137_, p_223108_, p_223109_, p_223110_, p_223112_, i, p_223111_, predicate);
       if (structurestart.m_73603_()) {
+         // CraftBukkit start
+         BoundingBox box = structurestart.m_73601_();
+         org.bukkit.event.world.AsyncStructureSpawnEvent event = new org.bukkit.event.world.AsyncStructureSpawnEvent(p_223106_.f_220460_.getMinecraftWorld().getWorld(), org.bukkit.craftbukkit.v1_20_R2.generator.structure.CraftStructure.minecraftToBukkit(structure, p_223107_), new org.bukkit.util.BoundingBox(box.m_162395_(), box.m_162396_(), box.m_162398_(), box.m_162399_(), box.m_162400_(), box.m_162401_()), p_223112_.f_45578_, p_223112_.f_45579_);
+         org.bukkit.Bukkit.getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return true;
+         }
+         // CraftBukkit end
          p_223106_.m_220516_(p_223113_, structure, structurestart, p_223111_);
          return true;
       } else {
