--- a/net/minecraft/world/level/chunk/ChunkStatus.java
+++ b/net/minecraft/world/level/chunk/ChunkStatus.java
@@ -38,7 +_,7 @@
    public static final ChunkStatus f_62314_ = m_62414_("empty", (ChunkStatus)null, -1, f_62327_, ChunkStatus.ChunkType.PROTOCHUNK, (p_156307_, p_156308_, p_156309_, p_156310_, p_156311_) -> {
    });
    public static final ChunkStatus f_62315_ = m_280108_("structure_starts", f_62314_, 0, false, f_62327_, ChunkStatus.ChunkType.PROTOCHUNK, (p_296976_, p_296977_, p_296978_, p_296979_, p_296980_, p_296981_, p_296982_, p_296983_, p_296984_) -> {
-      if (p_296978_.m_7654_().m_129910_().m_246337_().m_247749_()) {
+      if (p_296978_.f_8549_.m_246337_().m_247749_()) { // CraftBukkit
          p_296979_.m_255037_(p_296978_.m_9598_(), p_296978_.m_7726_().m_255415_(), p_296978_.m_215010_(), p_296984_, p_296980_);
       }
 
