--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -149,6 +_,18 @@
       this.duration = p_19735_;
    }
 
+   // Spigot start - copied from below
+   @Override
+   public void inactiveTick() {
+      super.inactiveTick();
+
+      if (this.tickCount >= this.waitTime + this.duration) {
+         this.discard();
+         return;
+      }
+   }
+   // Spigot end
+
    public void tick() {
       super.tick();
       boolean flag = this.isWaiting();
@@ -238,19 +_,35 @@
             } else {
                List<LivingEntity> list1 = this.level().getEntitiesOfClass(LivingEntity.class, this.getBoundingBox());
                if (!list1.isEmpty()) {
+                  List<org.bukkit.entity.LivingEntity> entities = new java.util.ArrayList<org.bukkit.entity.LivingEntity>(); // CraftBukkit
                   for(LivingEntity livingentity : list1) {
                      if (!this.victims.containsKey(livingentity) && livingentity.isAffectedByPotions()) {
                         double d8 = livingentity.getX() - this.getX();
                         double d1 = livingentity.getZ() - this.getZ();
                         double d3 = d8 * d8 + d1 * d1;
                         if (d3 <= (double)(f * f)) {
-                           this.victims.put(livingentity, this.tickCount + this.reapplicationDelay);
-
-                           for(MobEffectInstance mobeffectinstance1 : list) {
-                              if (mobeffectinstance1.getEffect().isInstantenous()) {
-                                 mobeffectinstance1.getEffect().applyInstantenousEffect(this, this.getOwner(), livingentity, mobeffectinstance1.getAmplifier(), 0.5D);
+                           // CraftBukkit start
+                           entities.add((org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity());
+                        }
+                     }
+                  }
+
+                  org.bukkit.event.entity.AreaEffectCloudApplyEvent event = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callAreaEffectCloudApplyEvent(this, entities);
+                  if (!event.isCancelled()) {
+                     for (org.bukkit.entity.LivingEntity entity : event.getAffectedEntities()) {
+                        if (entity instanceof org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) {
+                           LivingEntity entityliving = ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) entity).getHandle();
+                           // CraftBukkit end
+                           this.victims.put(entityliving, this.tickCount + this.reapplicationDelay);
+                           java.util.Iterator iterator2 = list.iterator();
+
+                           while (iterator2.hasNext()) {
+                              MobEffectInstance mobeffect1 = (MobEffectInstance) iterator2.next();
+
+                              if (mobeffect1.getEffect().isInstantenous()) {
+                                 mobeffect1.getEffect().applyInstantenousEffect(this, this.getOwner(), entityliving, mobeffect1.getAmplifier(), 0.5D);
                               } else {
-                                 livingentity.addEffect(new MobEffectInstance(mobeffectinstance1), this);
+                                 entityliving.addEffect(new MobEffectInstance(mobeffect1), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AREA_EFFECT_CLOUD); // CraftBukkit
                               }
                            }
 
