--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -52,6 +_,14 @@
       return this.f_20837_.m_135370_(this.f_20838_);
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.f_20834_ = true;
+      this.f_20835_ = 0;
+      this.f_20837_.m_135381_(this.f_20838_, ticks);
+   }
+   // CraftBukkit end
+
    public void m_20847_(CompoundTag p_20848_) {
       p_20848_.m_128379_("Saddle", this.m_20851_());
    }
