--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -115,6 +_,11 @@
    private CompoundTag f_21359_;
    private BlockPos f_21360_ = BlockPos.f_121853_;
    private float f_21341_ = -1.0F;
+   @Nullable
+   private MobSpawnType spawnType;
+   private boolean spawnCancelled = false;
+
+   public boolean aware = true; // CraftBukkit
 
    protected Mob(EntityType<? extends Mob> p_21368_, Level p_21369_) {
       super(p_21368_, p_21369_);
@@ -134,6 +_,12 @@
 
    }
 
+   // CraftBukkit start
+   public void setPersistenceRequired(boolean persistenceRequired) {
+      this.f_21353_ = persistenceRequired;
+   }
+   // CraftBukkit end
+
    protected void m_8099_() {
    }
 
@@ -229,7 +_,43 @@
    }
 
    public void m_6710_(@Nullable LivingEntity p_21544_) {
-      this.f_21362_ = p_21544_;
+      // CraftBukkit start - fire event
+      setTarget(p_21544_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.UNKNOWN, true);
+   }
+
+   public boolean setTarget(@Nullable LivingEntity p_21544_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent) {
+      if (m_5448_() == p_21544_) return false;
+      if (fireEvent) {
+         if (reason == org.bukkit.event.entity.EntityTargetEvent.TargetReason.UNKNOWN && m_5448_() != null && p_21544_ == null) {
+            reason = m_5448_().m_6084_() ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET : org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED;
+         }
+         if (reason == org.bukkit.event.entity.EntityTargetEvent.TargetReason.UNKNOWN) {
+            this.m_9236_().getCraftServer().getLogger().log(java.util.logging.Level.WARNING, "Unknown target reason, please report on the issue tracker", new Exception());
+         }
+         org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity ctarget = null;
+         if (p_21544_ != null) {
+            ctarget = (org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) p_21544_.getBukkitEntity();
+         }
+         org.bukkit.event.entity.EntityTargetLivingEntityEvent event = new org.bukkit.event.entity.EntityTargetLivingEntityEvent(this.getBukkitEntity(), ctarget, reason);
+         this.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return false;
+         }
+
+         if (event.getTarget() != null) {
+            p_21544_ = ((org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) event.getTarget()).getHandle();
+         } else {
+            p_21544_ = null;
+         }
+      }
+
+      net.minecraftforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.minecraftforge.common.ForgeHooks.onLivingChangeTarget(this, p_21544_, net.minecraftforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.MOB_TARGET);
+      if(!changeTargetEvent.isCanceled()) {
+         this.f_21362_ = changeTargetEvent.getNewTarget();
+         return true;
+      }
+      return false;
+      // CraftBukkit end
    }
 
    public boolean m_6549_(EntityType<?> p_21399_) {
@@ -356,6 +_,12 @@
       return null;
    }
 
+   // CraftBukkit start - Add delegate method
+   public SoundEvent getAmbientSound0() {
+      return m_7515_();
+   }
+   // CraftBukkit end
+
    public void m_7380_(CompoundTag p_21484_) {
       super.m_7380_(p_21484_);
       p_21484_.m_128379_("CanPickUpLoot", this.m_21531_());
@@ -427,15 +_,29 @@
          p_21484_.m_128379_("NoAI", this.m_21525_());
       }
 
+      if (this.spawnType != null) {
+         p_21484_.m_128359_("forge:spawn_type", this.spawnType.name());
+      }
+
+      p_21484_.m_128379_("Bukkit.Aware", this.aware); // CraftBukkit
    }
 
    public void m_7378_(CompoundTag p_21450_) {
       super.m_7378_(p_21450_);
+
+      // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
       if (p_21450_.m_128425_("CanPickUpLoot", 1)) {
-         this.m_21553_(p_21450_.m_128471_("CanPickUpLoot"));
+         boolean data = p_21450_.m_128471_("CanPickUpLoot");
+         if (isLevelAtLeast(p_21450_, 1) || data) {
+            this.m_21553_(data);
+         }
       }
 
-      this.f_21353_ = p_21450_.m_128471_("PersistenceRequired");
+      boolean data = p_21450_.m_128471_("PersistenceRequired");
+      if (isLevelAtLeast(p_21450_, 1) || data) {
+         this.f_21353_ = data;
+      }
+      // CraftBukkit end
       if (p_21450_.m_128425_("ArmorItems", 9)) {
          ListTag listtag = p_21450_.m_128437_("ArmorItems", 10);
 
@@ -479,6 +_,20 @@
       }
 
       this.m_21557_(p_21450_.m_128471_("NoAI"));
+
+      // CraftBukkit start
+      if (p_21450_.m_128441_("Bukkit.Aware")) {
+         this.aware = p_21450_.m_128471_("Bukkit.Aware");
+      }
+      // CraftBukkit end
+
+      if (p_21450_.m_128441_("forge:spawn_type")) {
+         try {
+            this.spawnType = MobSpawnType.valueOf(p_21450_.m_128461_("forge:spawn_type"));
+         } catch (Exception ex) {
+            p_21450_.m_128473_("forge:spawn_type");
+         }
+      }
    }
 
    protected void m_7625_(DamageSource p_21389_, boolean p_21390_) {
@@ -518,9 +_,9 @@
    public void m_8107_() {
       super.m_8107_();
       this.m_9236_().m_46473_().m_6180_("looting");
-      if (!this.m_9236_().f_46443_ && this.m_21531_() && this.m_6084_() && !this.f_20890_ && this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
          Vec3i vec3i = this.m_213552_();
 
+      if (!this.m_9236_().f_46443_ && this.m_21531_() && this.m_6084_() && !this.f_20890_ && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.m_9236_(), this)) {
          for(ItemEntity itementity : this.m_9236_().m_45976_(ItemEntity.class, this.m_20191_().m_82377_((double)vec3i.m_123341_(), (double)vec3i.m_123342_(), (double)vec3i.m_123343_()))) {
             if (!itementity.m_213877_() && !itementity.m_32055_().m_41619_() && !itementity.m_32063_() && this.m_7243_(itementity.m_32055_())) {
                this.m_7581_(itementity);
@@ -537,7 +_,7 @@
 
    protected void m_7581_(ItemEntity p_21471_) {
       ItemStack itemstack = p_21471_.m_32055_();
-      ItemStack itemstack1 = this.m_255207_(itemstack.m_41777_());
+      ItemStack itemstack1 = this.equipItemIfPossible(itemstack.m_41777_(), p_21471_); // CraftBukkit - add item
       if (!itemstack1.m_41619_()) {
          this.m_21053_(p_21471_);
          this.m_7938_(p_21471_, itemstack1.m_41613_());
@@ -550,6 +_,12 @@
    }
 
    public ItemStack m_255207_(ItemStack p_255842_) {
+      // CraftBukkit start - add item
+      return this.equipItemIfPossible(p_255842_, null);
+   }
+
+   public ItemStack equipItemIfPossible(ItemStack p_255842_, ItemEntity entity) {
+      // CraftBukkit end
       EquipmentSlot equipmentslot = m_147233_(p_255842_);
       ItemStack itemstack = this.m_6844_(equipmentslot);
       boolean flag = this.m_7808_(p_255842_, itemstack);
@@ -559,10 +_,18 @@
          flag = itemstack.m_41619_();
       }
 
-      if (flag && this.m_7252_(p_255842_)) {
+      // CraftBukkit start
+      boolean canPickup = flag && this.m_7252_(p_255842_);
+      if (entity != null) {
+         canPickup = !org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityPickupItemEvent(this, entity, 0, !canPickup).isCancelled();
+      }
+      if (canPickup) {
+         // CraftBukkit end
          double d0 = (double)this.m_21519_(equipmentslot);
          if (!itemstack.m_41619_() && (double)Math.max(this.f_19796_.m_188501_() - 0.1F, 0.0F) < d0) {
+            this.forceDrops = true; // CraftBukkit
             this.m_19983_(itemstack);
+            this.forceDrops = false; // CraftBukkit
          }
 
          if (equipmentslot.m_254934_() && p_255842_.m_41613_() > 1) {
@@ -692,6 +_,14 @@
          this.m_146870_();
       } else if (!this.m_21532_() && !this.m_8023_()) {
          Entity entity = this.m_9236_().m_45930_(this, -1.0D);
+         net.minecraftforge.eventbus.api.Event.Result result = net.minecraftforge.event.ForgeEventFactory.canEntityDespawn(this, (ServerLevel) this.m_9236_());
+         if (result == net.minecraftforge.eventbus.api.Event.Result.DENY) {
+            f_20891_ = 0;
+            entity = null;
+         } else if (result == net.minecraftforge.eventbus.api.Event.Result.ALLOW) {
+            this.m_146870_();
+            entity = null;
+         }
          if (entity != null) {
             double d0 = entity.m_20280_(this);
             int i = this.m_6095_().m_20674_().m_21611_();
@@ -716,6 +_,7 @@
 
    protected final void m_6140_() {
       ++this.f_20891_;
+      if (!this.aware) return; // CraftBukkit
       this.m_9236_().m_46473_().m_6180_("sensing");
       this.f_21349_.m_26789_();
       this.m_9236_().m_46473_().m_7238_();
@@ -1025,6 +_,16 @@
 
    }
 
+   /**
+    * Forge: Override-Only, call via ForgeEventFactory.onFinalizeSpawn.<br>
+    * Overrides are allowed. Do not wrap super calls within override (as that will cause stack overflows).<br>
+    * Vanilla calls are replaced with a transformer, and are not visible in source.<br>
+    * <p>
+    * Be certain to either call super.finalizeSpawn or set the {@link #spawnType} field from within your override.
+    * @see {@link net.minecraftforge.event.ForgeEventFactory#onFinalizeSpawn onFinalizeSpawn} for additional documentation.
+    */
+   @Deprecated
+   @org.jetbrains.annotations.ApiStatus.OverrideOnly
    @Nullable
    public SpawnGroupData m_6518_(ServerLevelAccessor p_21434_, DifficultyInstance p_21435_, MobSpawnType p_21436_, @Nullable SpawnGroupData p_21437_, @Nullable CompoundTag p_21438_) {
       RandomSource randomsource = p_21434_.m_213780_();
@@ -1035,6 +_,7 @@
          this.m_21559_(false);
       }
 
+      this.spawnType = p_21436_;
       return p_21437_;
    }
 
@@ -1074,6 +_,12 @@
       if (!this.m_6084_()) {
          return InteractionResult.PASS;
       } else if (this.m_21524_() == p_21420_) {
+         // CraftBukkit start - fire PlayerUnleashEntityEvent
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerUnleashEntityEvent(this, p_21420_, p_21421_).isCancelled()) {
+            ((net.minecraft.server.level.ServerPlayer) p_21420_).f_8906_.m_9829_(new ClientboundSetEntityLinkPacket(this, this.m_21524_()));
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          this.m_21455_(true, !p_21420_.m_150110_().f_35937_);
          this.m_146852_(GameEvent.f_223708_, p_21420_);
          return InteractionResult.m_19078_(this.m_9236_().f_46443_);
@@ -1097,6 +_,12 @@
    private InteractionResult m_21499_(Player p_21500_, InteractionHand p_21501_) {
       ItemStack itemstack = p_21500_.m_21120_(p_21501_);
       if (itemstack.m_150930_(Items.f_42655_) && this.m_6573_(p_21500_)) {
+         // CraftBukkit start - fire PlayerLeashEntityEvent
+         if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerLeashEntityEvent(this, p_21500_, p_21500_, p_21501_).isCancelled()) {
+            ((net.minecraft.server.level.ServerPlayer) p_21500_).f_8906_.m_9829_(new ClientboundSetEntityLinkPacket(this, this.m_21524_()));
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          this.m_21463_(p_21500_, true);
          itemstack.m_41774_(1);
          return InteractionResult.m_19078_(this.m_9236_().f_46443_);
@@ -1165,8 +_,15 @@
       return this.f_21341_ != -1.0F;
    }
 
+   // CraftBukkit start
    @Nullable
    public <T extends Mob> T m_21406_(EntityType<T> p_21407_, boolean p_21408_) {
+      return this.convertTo(p_21407_, p_21408_, org.bukkit.event.entity.EntityTransformEvent.TransformReason.UNKNOWN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
+   @Nullable
+   public <T extends Mob> T convertTo(EntityType<T> p_21407_, boolean p_21408_, org.bukkit.event.entity.EntityTransformEvent.TransformReason transformReason, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason) {
+      // CraftBukkit end
       if (this.m_213877_()) {
          return (T)null;
       } else {
@@ -1199,7 +_,12 @@
                }
             }
 
-            this.m_9236_().m_7967_(t);
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callEntityTransformEvent(this, t, transformReason).isCancelled()) {
+               return null;
+            }
+            this.m_9236_().addFreshEntity(t, spawnReason);
+            // CraftBukkit end
             if (this.m_20159_()) {
                Entity entity = this.m_20202_();
                this.m_8127_();
@@ -1219,6 +_,7 @@
 
       if (this.f_21357_ != null) {
          if (!this.m_6084_() || !this.f_21357_.m_6084_()) {
+            this.m_9236_().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), (!this.m_6084_()) ? org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.PLAYER_UNLEASH : org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.HOLDER_GONE)); // CraftBukkit
             this.m_21455_(true, true);
          }
 
@@ -1230,7 +_,9 @@
          this.f_21357_ = null;
          this.f_21359_ = null;
          if (!this.m_9236_().f_46443_ && p_21457_) {
+            this.forceDrops = true; // CraftBukkit
             this.m_19998_(Items.f_42655_);
+            this.forceDrops = false; // CraftBukkit
          }
 
          if (!this.m_9236_().f_46443_ && p_21456_ && this.m_9236_() instanceof ServerLevel) {
@@ -1278,6 +_,7 @@
    public boolean m_7998_(Entity p_21396_, boolean p_21397_) {
       boolean flag = super.m_7998_(p_21396_, p_21397_);
       if (flag && this.m_21523_()) {
+         this.m_9236_().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.UNKNOWN)); // CraftBukkit
          this.m_21455_(true, true);
       }
 
@@ -1300,7 +_,9 @@
          }
 
          if (this.f_19797_ > 100) {
+            this.forceDrops = true; // CraftBukkit
             this.m_19998_(Items.f_42655_);
+            this.forceDrops = false; // CraftBukkit
             this.f_21359_ = null;
          }
       }
@@ -1368,6 +_,14 @@
 
       int i = EnchantmentHelper.m_44914_(this);
       if (i > 0) {
+         // CraftBukkit start - Call a combust event when somebody hits with a fire enchanted item
+         org.bukkit.event.entity.EntityCombustByEntityEvent combustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), p_21372_.getBukkitEntity(), i * 4);
+         org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+
+         if (!combustEvent.isCancelled()) {
+            p_21372_.setSecondsOnFire(combustEvent.getDuration(), false);
+         }
+         // CraftBukkit end
          p_21372_.m_20254_(i * 4);
       }
 
@@ -1414,15 +_,25 @@
       return false;
    }
 
+   @Deprecated // FORGE: use jumpInFluid instead
    protected void m_203347_(TagKey<Fluid> p_204045_) {
+      this.jumpInLiquidInternal(() -> super.m_203347_(p_204045_));
+   }
+
+   private void jumpInLiquidInternal(Runnable onSuper) {
       if (this.m_21573_().m_26576_()) {
-         super.m_203347_(p_204045_);
+         onSuper.run();
       } else {
          this.m_20256_(this.m_20184_().m_82520_(0.0D, 0.3D, 0.0D));
       }
 
    }
 
+   @Override
+   public void jumpInFluid(net.minecraftforge.fluids.FluidType type) {
+      this.jumpInLiquidInternal(() -> super.jumpInFluid(type));
+   }
+
    public void m_147272_() {
       this.m_262441_((p_262562_) -> {
          return true;
@@ -1436,6 +_,7 @@
 
    protected void m_6089_() {
       super.m_6089_();
+      this.m_9236_().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.UNKNOWN)); // CraftBukkit
       this.m_21455_(true, false);
       this.m_20158_().forEach((p_278936_) -> {
          if (!p_278936_.m_41619_()) {
@@ -1449,5 +_,40 @@
    public ItemStack m_142340_() {
       SpawnEggItem spawneggitem = SpawnEggItem.m_43213_(this.m_6095_());
       return spawneggitem == null ? null : new ItemStack(spawneggitem);
+   }
+
+   /**
+   * Returns the type of spawn that created this mob, if applicable.
+   * If it could not be determined, this will return null.
+   * <p>
+   * This is set via {@link Mob#finalizeSpawn}, so you should not call this from within that method, instead using the parameter.
+   */
+   @Nullable
+   public final MobSpawnType getSpawnType() {
+      return this.spawnType;
+   }
+
+   /**
+    * This method exists so that spawns can be cancelled from the {@link net.minecraftforge.event.entity.living.MobSpawnEvent.FinalizeSpawn FinalizeSpawnEvent}
+    * without needing to hook up an additional handler for the {@link net.minecraftforge.event.entity.EntityJoinLevelEvent EntityJoinLevelEvent}.
+    * @return if this mob will be blocked from spawning during {@link Level#addFreshEntity(Entity)}
+    * @apiNote Not public-facing API.
+    */
+   @org.jetbrains.annotations.ApiStatus.Internal
+   public final boolean isSpawnCancelled() {
+      return this.spawnCancelled;
+   }
+
+   /**
+    * Marks this mob as being disallowed to spawn during {@link Level#addFreshEntity(Entity)}.<p>
+    * @throws UnsupportedOperationException if this entity has already been {@link Entity#isAddedToWorld() added to the world}.
+    * @apiNote Not public-facing API.
+    */
+   @org.jetbrains.annotations.ApiStatus.Internal
+   public final void setSpawnCancelled(boolean cancel) {
+      if (this.isAddedToWorld()) {
+         throw new UnsupportedOperationException("Late invocations of Mob#setSpawnCancelled are not permitted.");
+      }
+      this.spawnCancelled = cancel;
    }
 }
