--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -62,7 +_,7 @@
                   }
                }
 
-               this.mob.setTarget(livingentity);
+               this.mob.setTarget(livingentity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit
                return true;
             }
          }
@@ -80,7 +_,7 @@
    }
 
    public void stop() {
-      this.mob.setTarget((LivingEntity)null);
+      this.mob.setTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.targetMob = null;
    }
 
