--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -62,7 +_,7 @@
                   }
                }
 
-               this.f_26135_.m_6710_(livingentity);
+               this.f_26135_.setTarget(livingentity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit
                return true;
             }
          }
@@ -80,7 +_,7 @@
    }
 
    public void m_8041_() {
-      this.f_26135_.m_6710_((LivingEntity)null);
+      this.f_26135_.setTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.f_26137_ = null;
    }
 
