--- a/net/minecraft/world/level/Explosion.java
+++ b/net/minecraft/world/level/Explosion.java
@@ -238,20 +_,78 @@
          boolean flag1 = this.getIndirectSourceEntity() instanceof Player;
          Util.shuffle(this.toBlow, this.level.random);
 
+         // CraftBukkit start
+         org.bukkit.World bworld = this.level.getWorld();
+         org.bukkit.entity.Entity explode = this.source == null ? null : this.source.getBukkitEntity();
+         Location location = new Location(bworld, this.x, this.y, this.z);
+
+         List<org.bukkit.block.Block> blockList = new ObjectArrayList<>();
+         for (int i1 = this.toBlow.size() - 1; i1 >= 0; i1--) {
+            BlockPos cpos = this.toBlow.get(i1);
+            org.bukkit.block.Block bblock = bworld.getBlockAt(cpos.getX(), cpos.getY(), cpos.getZ());
+            if (!bblock.getType().isAir()) {
+               blockList.add(bblock);
+            }
+         }
+
+         boolean cancelled;
+         List<org.bukkit.block.Block> bukkitBlocks;
+         float yield;
+
+         if (explode != null) {
+            EntityExplodeEvent event = new EntityExplodeEvent(explode, location, blockList, this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F);
+            this.level.getCraftServer().getPluginManager().callEvent(event);
+            cancelled = event.isCancelled();
+            bukkitBlocks = event.blockList();
+            yield = event.getYield();
+         } else {
+            BlockExplodeEvent event = new BlockExplodeEvent(location.getBlock(), blockList, this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F);
+            this.level.getCraftServer().getPluginManager().callEvent(event);
+            cancelled = event.isCancelled();
+            bukkitBlocks = event.blockList();
+            yield = event.getYield();
+         }
+
+         this.toBlow.clear();
+
+         for (org.bukkit.block.Block bblock : bukkitBlocks) {
+            BlockPos coords = new BlockPos(bblock.getX(), bblock.getY(), bblock.getZ());
+            toBlow.add(coords);
+         }
+
+         if (cancelled) {
+            this.wasCanceled = true;
+            return;
+         }
+         // CraftBukkit end
          for(BlockPos blockpos : this.toBlow) {
             BlockState blockstate = this.level.getBlockState(blockpos);
             Block block = blockstate.getBlock();
+            // CraftBukkit start - TNTPrimeEvent
+            if (block instanceof net.minecraft.world.level.block.TntBlock) {
+               Entity sourceEntity = source == null ? null : source;
+               BlockPos sourceBlock = sourceEntity == null ? BlockPos.containing(this.x, this.y, this.z) : null;
+               if (!CraftEventFactory.callTNTPrimeEvent(this.level, blockpos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.EXPLOSION, sourceEntity, sourceBlock)) {
+                  this.level.sendBlockUpdated(blockpos, Blocks.AIR.defaultBlockState(), blockstate, 3); // Update the block on the client
+                  continue;
+               }
+            }
+            // CraftBukkit end
+
             if (!blockstate.isAir()) {
                BlockPos blockpos1 = blockpos.immutable();
+
                this.level.getProfiler().push("explosion_blocks");
-               if (block.dropFromExplosion(this)) {
+               if (blockstate.canDropFromExplosion(this.level, blockpos, this)) {
                   Level $$9 = this.level;
+
                   if ($$9 instanceof ServerLevel) {
                      ServerLevel serverlevel = (ServerLevel)$$9;
                      BlockEntity blockentity = blockstate.hasBlockEntity() ? this.level.getBlockEntity(blockpos) : null;
                      LootParams.Builder lootparams$builder = (new LootParams.Builder(serverlevel)).withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(blockpos)).withParameter(LootContextParams.TOOL, ItemStack.EMPTY).withOptionalParameter(LootContextParams.BLOCK_ENTITY, blockentity).withOptionalParameter(LootContextParams.THIS_ENTITY, this.source);
-                     if (this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY) {
-                        lootparams$builder.withParameter(LootContextParams.EXPLOSION_RADIUS, this.radius);
+
+                     if (yield < 1.0F) { // CraftBukkit - add yield
+                        lootparams$builder.withParameter(LootContextParams.EXPLOSION_RADIUS, 1.0F / yield); // CraftBukkit - add yield
                      }
 
                      blockstate.spawnAfterBreak(serverlevel, blockpos, ItemStack.EMPTY, flag1);
@@ -261,8 +_,7 @@
                   }
                }
 
-               this.level.setBlock(blockpos, Blocks.AIR.defaultBlockState(), 3);
-               block.wasExploded(this.level, blockpos, this);
+               blockstate.onBlockExploded(this.level, blockpos, this);
                this.level.getProfiler().pop();
             }
          }
