--- a/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
@@ -23,6 +_,24 @@
             this.f_31411_.m_7081_();
          }
 
+         // CraftBukkit start - Call EnderDragonChangePhaseEvent
+         org.bukkit.event.entity.EnderDragonChangePhaseEvent event = new org.bukkit.event.entity.EnderDragonChangePhaseEvent(
+                 (org.bukkit.craftbukkit.v1_20_R1.entity.CraftEnderDragon) this.f_31409_.getBukkitEntity(),
+                 (this.f_31411_ == null) ? null : org.bukkit.craftbukkit.v1_20_R1.entity.CraftEnderDragon.getBukkitPhase(this.f_31411_.m_7309_()),
+                 org.bukkit.craftbukkit.v1_20_R1.entity.CraftEnderDragon.getBukkitPhase(p_31417_)
+         );
+         this.f_31409_.m_9236_().getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return;
+         }
+
+         //Ketting start - check for modded phases
+         EnderDragonPhase<?> newPhase = org.bukkit.craftbukkit.v1_20_R1.entity.CraftEnderDragon.getMinecraftPhase(event.getNewPhase());
+         if (newPhase != null)
+            p_31417_ = newPhase;
+         //Ketting end
+         // CraftBukkit end
+
          this.f_31411_ = this.m_31418_(p_31417_);
          if (!this.f_31409_.m_9236_().f_46443_) {
             this.f_31409_.m_20088_().m_135381_(EnderDragon.f_31067_, p_31417_.m_31405_());
