--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -69,6 +_,11 @@
    private Vec3i size = Vec3i.ZERO;
    private String author = "?";
 
+   // CraftBukkit start - data containers
+   private static final org.bukkit.craftbukkit.v1_20_R1.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.v1_20_R1.persistence.CraftPersistentDataTypeRegistry();
+   public org.bukkit.craftbukkit.v1_20_R1.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.v1_20_R1.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
+   // CraftBukkit end
+
    public Vec3i getSize() {
       return this.size;
    }
@@ -201,6 +_,10 @@
       return transform(p_74565_, p_74564_.getMirror(), p_74564_.getRotation(), p_74564_.getRotationPivot());
    }
 
+   public static Vec3 transformedVec3d(StructurePlaceSettings placementIn, Vec3 pos) {
+      return transform(pos, placementIn.getMirror(), placementIn.getRotation(), placementIn.getRotationPivot());
+   }
+
    public boolean placeInWorld(ServerLevelAccessor p_230329_, BlockPos p_230330_, BlockPos p_230331_, StructurePlaceSettings p_230332_, RandomSource p_230333_, int p_230334_) {
       if (this.palettes.isEmpty()) {
          return false;
@@ -218,7 +_,7 @@
             int i1 = Integer.MIN_VALUE;
             int j1 = Integer.MIN_VALUE;
 
-            for(StructureTemplate.StructureBlockInfo structuretemplate$structureblockinfo : processBlockInfos(p_230329_, p_230330_, p_230331_, p_230332_, list)) {
+            for(StructureTemplate.StructureBlockInfo structuretemplate$structureblockinfo : processBlockInfos(p_230329_, p_230330_, p_230331_, p_230332_, list, this)) {
                BlockPos blockpos = structuretemplate$structureblockinfo.pos;
                if (boundingbox == null || boundingbox.isInside(blockpos)) {
                   FluidState fluidstate = p_230332_.shouldKeepLiquids() ? p_230329_.getFluidState(blockpos) : null;
@@ -359,7 +_,15 @@
       });
    }
 
+   /**
+    * @deprecated Forge: Use {@link #processBlockInfos(ServerLevelAccessor, BlockPos, BlockPos, StructurePlaceSettings, List, StructureTemplate)} instead.
+    */
+   @Deprecated
    public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(ServerLevelAccessor p_278297_, BlockPos p_74519_, BlockPos p_74520_, StructurePlaceSettings p_74521_, List<StructureTemplate.StructureBlockInfo> p_74522_) {
+      return processBlockInfos(p_278297_, p_74519_, p_74520_, p_74521_, p_74522_, null);
+   }
+
+   public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(ServerLevelAccessor p_278297_, BlockPos p_74519_, BlockPos p_74520_, StructurePlaceSettings p_74521_, List<StructureTemplate.StructureBlockInfo> p_74522_, @Nullable StructureTemplate template) {
       List<StructureTemplate.StructureBlockInfo> list = new ArrayList<>();
       List<StructureTemplate.StructureBlockInfo> list1 = new ArrayList<>();
 
@@ -367,7 +_,7 @@
          BlockPos blockpos = calculateRelativePosition(p_74521_, structuretemplate$structureblockinfo.pos).offset(p_74519_);
          StructureTemplate.StructureBlockInfo structuretemplate$structureblockinfo1 = new StructureTemplate.StructureBlockInfo(blockpos, structuretemplate$structureblockinfo.state, structuretemplate$structureblockinfo.nbt != null ? structuretemplate$structureblockinfo.nbt.copy() : null);
 
-         for(Iterator<StructureProcessor> iterator = p_74521_.getProcessors().iterator(); structuretemplate$structureblockinfo1 != null && iterator.hasNext(); structuretemplate$structureblockinfo1 = iterator.next().processBlock(p_278297_, p_74519_, p_74520_, structuretemplate$structureblockinfo, structuretemplate$structureblockinfo1, p_74521_)) {
+         for(Iterator<StructureProcessor> iterator = p_74521_.getProcessors().iterator(); structuretemplate$structureblockinfo1 != null && iterator.hasNext(); structuretemplate$structureblockinfo1 = iterator.next().process(p_278297_, p_74519_, p_74520_, structuretemplate$structureblockinfo, structuretemplate$structureblockinfo1, p_74521_, template)) {
          }
 
          if (structuretemplate$structureblockinfo1 != null) {
@@ -383,6 +_,23 @@
       return list1;
    }
 
+   public static List<StructureTemplate.StructureEntityInfo> processEntityInfos(@Nullable StructureTemplate template, LevelAccessor p_215387_0_, BlockPos p_215387_1_, StructurePlaceSettings p_215387_2_, List<StructureTemplate.StructureEntityInfo> p_215387_3_) {
+      List<StructureTemplate.StructureEntityInfo> list = Lists.newArrayList();
+      for(StructureTemplate.StructureEntityInfo entityInfo : p_215387_3_) {
+         Vec3 pos = transformedVec3d(p_215387_2_, entityInfo.pos).add(Vec3.atLowerCornerOf(p_215387_1_));
+         BlockPos blockpos = calculateRelativePosition(p_215387_2_, entityInfo.blockPos).offset(p_215387_1_);
+         StructureTemplate.StructureEntityInfo info = new StructureTemplate.StructureEntityInfo(pos, blockpos, entityInfo.nbt);
+         for (StructureProcessor proc : p_215387_2_.getProcessors()) {
+            info = proc.processEntity(p_215387_0_, p_215387_1_, entityInfo, info, p_215387_2_, template);
+            if (info == null)
+               break;
+         }
+         if (info != null)
+            list.add(info);
+      }
+      return list;
+   }
+
    private void placeEntities(ServerLevelAccessor p_74524_, BlockPos p_74525_, Mirror p_74526_, Rotation p_74527_, BlockPos p_74528_, @Nullable BoundingBox p_74529_, boolean p_74530_) {
       for(StructureTemplate.StructureEntityInfo structuretemplate$structureentityinfo : this.entityInfoList) {
          BlockPos blockpos = transform(structuretemplate$structureentityinfo.blockPos, p_74526_, p_74527_, p_74528_).offset(p_74525_);
@@ -412,11 +_,13 @@
    }
 
    private static Optional<Entity> createEntityIgnoreException(ServerLevelAccessor p_74544_, CompoundTag p_74545_) {
-      try {
+      // CraftBukkit start
+      // try {
          return EntityType.create(p_74545_, p_74544_.getLevel());
-      } catch (Exception exception) {
-         return Optional.empty();
-      }
+      // } catch (Exception exception) {
+      //    return Optional.empty();
+      // }
+      // CraftBukkit end
    }
 
    public Vec3i getSize(Rotation p_163809_) {
@@ -607,6 +_,11 @@
 
       p_74619_.put("entities", listtag);
       p_74619_.put("size", this.newIntegerList(this.size.getX(), this.size.getY(), this.size.getZ()));
+      // CraftBukkit start - PDC
+      if (!this.persistentDataContainer.isEmpty()) {
+         p_74619_.put("BukkitValues", this.persistentDataContainer.toTagCompound());
+      }
+      // CraftBukkit end
       return NbtUtils.addCurrentDataVersion(p_74619_);
    }
 
@@ -640,6 +_,12 @@
          }
       }
 
+      // CraftBukkit start - PDC
+      net.minecraft.nbt.Tag base = p_248574_.get("BukkitValues");
+      if (base instanceof CompoundTag) {
+         this.persistentDataContainer.putAll((CompoundTag) base);
+      }
+      // CraftBukkit end
    }
 
    private void loadPalette(HolderGetter<Block> p_256546_, ListTag p_251056_, ListTag p_251493_) {
