--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
@@ -227,7 +_,7 @@
       }
    }
 
-   private StructureTemplate m_230377_(InputStream p_230378_) throws IOException {
+   public StructureTemplate m_230377_(InputStream p_230378_) throws IOException {
       CompoundTag compoundtag = NbtIo.m_128939_(p_230378_);
       return this.m_230404_(compoundtag);
    }
