--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
@@ -227,7 +_,7 @@
       }
    }
 
-   private StructureTemplate readStructure(InputStream p_230378_) throws IOException {
+   public StructureTemplate readStructure(InputStream p_230378_) throws IOException {
       CompoundTag compoundtag = NbtIo.readCompressed(p_230378_);
       return this.readStructure(compoundtag);
    }
